---

title: Examples


keywords: fastai
sidebar: home_sidebar

summary: "Example workflows for using glycowork in your work"
description: "Example workflows for using glycowork in your work"
nb_path: "05_examples.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 05_examples.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In this file, you will access to different toy examples demonstrating how you can apply <code>glycowork</code> to your projects! 
Here is a summary of the different case studies available: <br></p>
<ol>
<li><a href="#example1">Investigation of <em>N</em>-linked and <em>O</em>-linked glycosylations in animals</a></li>
<li><a href="#example2">Glycan binding specificities of influenza viruses</a></li>
<li><a href="#example3">Glycan classification using machine learning models</a></li>
<li><a href="#example4">Deep Learning Code Snippet</a></li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Example-1:-Investigation-of-N-linked-and-O-linked-glycosylations-in-arthropods-">Example 1: Investigation of <em>N</em>-linked and <em>O</em>-linked glycosylations in arthropods <a class="anchor" id="example1" /><a class="anchor-link" href="#Example-1:-Investigation-of-N-linked-and-O-linked-glycosylations-in-arthropods-"> </a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Suppose that after investigating glycans from plants as presented in the <a href="/glycowork/core.html"><code>core</code></a> module, you are now wondering what happens for glycans from the Arthropoda phylum. To satisfy your curiosity, let's start with importing the <code>df_species</code> dataset from <code>glycowork.glycan_data.loader</code> and putting a filter for animal glycans.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df_animal</span> <span class="o">=</span> <span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Phylum</span> <span class="o">==</span> <span class="s1">&#39;Arthropoda&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To have a better overview of the glycans similarities in arthropods, you can plot these data using <code>plot_embeddings</code> from <code>glycowork.motif.analysis</code>. The best option here to visualize the data is to color the dots by taxonomic class.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">plot_embeddings</span><span class="p">(</span><span class="n">df_animal</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">label_list</span> <span class="o">=</span> <span class="n">df_animal</span><span class="o">.</span><span class="n">Class</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Interesting! First, the graph shows Arachnida clusters in addition to lots of Insecta clusters. <br></p>
<p>Now, let's focus on further analyses. We will first convert our data into a count table using <code>presence_to_matrix</code> from the <code>glycowork.motif.processing</code> module and then visualize the results. A heatmap can be generated using <code>make_heatmap</code> from <code>glycowork.motif.analysis</code> and the <code>yticklabels = True</code> option allows us to display all the labels.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df_map_animal</span> <span class="o">=</span> <span class="n">presence_to_matrix</span><span class="p">(</span><span class="n">df_animal</span><span class="p">,</span> <span class="n">label_col_name</span> <span class="o">=</span> <span class="s1">&#39;Class&#39;</span><span class="p">)</span>
<span class="n">make_heatmap</span><span class="p">(</span><span class="n">df_map_animal</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;motif&#39;</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;known&#39;</span><span class="p">,</span> <span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span> <span class="n">datatype</span> <span class="o">=</span> <span class="s1">&#39;presence&#39;</span><span class="p">,</span> <span class="n">yticklabels</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">estimate_speedup</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The first thing that catches the eye is the absence of a bunch of glycan motifs in Insecta that are absent in the other classes, in part presumably because of the scarcity of their known glycans relative to Insecta. Coming back to Arachnida, this heatmap shows for example a notable presence of difucosylated cores, with the corresponding distinguishing a1-3 Fuc motif. The other classes, Malacostraca and Merostomata, seem rather enriched in mannose-rich motifs in their <em>N</em>-glycans. <br></p>
<p>To visualize which dots from the embedding plot correspond to, for instance, GlcA-containing glycans, we can map them again but with the additional option <code>shape_feature</code>. This parameter allows you to transform dots into crosses if the corresponding glycan contains the specified monosaccharide or linkage.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">plot_embeddings</span><span class="p">(</span><span class="n">df_animal</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">label_list</span> <span class="o">=</span> <span class="n">df_animal</span><span class="o">.</span><span class="n">Class</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">shape_feature</span> <span class="o">=</span> <span class="s2">&quot;GlcA&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>As you can see, in <code>df_species</code>, GlcA-containing glycans (crosses) exist only in Insecta organisms (in the phylum Arthropoda). In this class, they even constitute isolated clusters, presumably corresponding to certain taxonomic groups. <br></p>
<p>We can also use <code>glycowork</code> to take a look at the overall monosaccharide distribution in animal glycans. The <code>characterize_monosaccharide</code> from <code>glycowork.motif.analysis</code> will be our best ally to learn what are the other monosaccharides to which, for instance, Gal variants are often connected to. <br></p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">characterize_monosaccharide</span><span class="p">(</span><span class="s1">&#39;Gal&#39;</span><span class="p">,</span> <span class="n">rank</span> <span class="o">=</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">,</span> <span class="n">focus</span> <span class="o">=</span> <span class="s1">&#39;Animalia&#39;</span><span class="p">,</span> <span class="n">modifications</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Running this analysis, it appears that Gal is found connected to GlcNAc as well as various other monosaccharides and that it does display several variants, including sulfurylated galactose. This is a completely different profile that what was obtained studying xylose in plant glycans! <br></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Example-2:-Glycan-binding-specificities-of-influenza-viruses-">Example 2: Glycan binding specificities of influenza viruses <a class="anchor" id="example2" /><a class="anchor-link" href="#Example-2:-Glycan-binding-specificities-of-influenza-viruses-"> </a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>For this second example, we will move from animals to viruses. Influenza viruses are well known as the biological agents responsible for the seasonal flu in humans. In general, viruses penetrate inside host cells thanks to a mechanism involving contacts between glycosylated proteins. Surface proteins surrounding viruses and eukaryotic cells interact together and allow the entry of viral particules inside the targeted cells. However, different influenza strains may recognize different glycosylations on proteins with more or less efficiency. <br></p>
<p>To measure these glycan binding specificities, one can add virus particules to glycan arrays presenting immobilized glycans. Such a protocol allows screening for specific binding. From such data, we can then ask what are the glycans that are the most efficiently recognized by a given influenza virus? <br></p>
<p>Fortunately, <code>glycowork</code> can be really helpful to answer such a question! The binding efficiency of different influenza virus strains for a vast range of glycans is available in the <code>influenza_binding</code> dataset. Let's start by importing these data.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>As you can see, this dataset contains numerical values (Z-scores) corresponding to the ability of different influenza strains to bind different glycans. From these data, we can compute a heatmap with <code>make_heatmap</code>. The results should teach us whether some specific strains bind specific glycans. As we work with quantitative values, the <code>datatype</code> parameter must be set to <code>'response'</code>. If you want to include literature-annotated motifs (e.g., VIM, Lewis X, Blood group A), just add 'known' to the list of the feature_set. Let's see what kind of heatmap we can generate using this dataset.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">make_heatmap</span><span class="p">(</span><span class="n">influenza_binding</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;motif&#39;</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span> <span class="n">datatype</span> <span class="o">=</span> <span class="s1">&#39;response&#39;</span><span class="p">,</span> <span class="n">yticklabels</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Due to the high number of viruses in this dataset, the heatmap is difficult to analyze in detail. However, we can clearly see that some specific influenza strains are particularly well suited to recognize specific glycan motifs. Interestingly, we can note that the recognition of a2-6 and Neu5Ac(a2-6)Gal-containing glycans on one hand (two first lines), and the recognition of Neu5Ac, a2-3 and Neu5Ac(a2-3)Gal-containing glycans on the other hand (three last lines) seems to be exclusive in most of the strains. <br></p>
<p>Another way to analyze the glycan binding capacities of influenza viruses is to use the <code>get_pvals_motifs</code> function from the <code>glycowork.motif.analysis</code> module. As we are working with data obtained from different strains, we must also set <code>multiple_samples</code> to <code>True</code>. These adjustments being made, the next step consists in executing the function!</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="n">get_pvals_motifs</span><span class="p">(</span><span class="n">influenza_binding</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span> <span class="n">multiple_samples</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>These results indicate, for each motif, a p-value and a corrected p-value. Based on these numbers, we can identify Neu5Ac(a2-3)Gal and Neu5Ac(a2-6)Gal as particularly well recognized motifs. This is consistent with the current knowledge as Neu5Ac(a2-3)Gal is the canonical binding motif for avian influenza and Neu5Ac(a2-6)Gal for mammalian influenza. Interestingly, other motifs also seem to be recognized by influenza strains, such as sulfurylated glycans.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Example-3:-Glycan-classification-using-machine-learning-models-">Example 3: Glycan classification using machine learning models <a class="anchor" id="example3" /><a class="anchor-link" href="#Example-3:-Glycan-classification-using-machine-learning-models-"> </a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Investigation of glycans from animals and viruses demonstrated how specific their presence in an organism can be. This particular characteristic can be exploited to predict the source of a given glycan among living organisms. By classifying which glycans stem from which species/taxonomic groups, machine learning models can be trained to learn which motifs are enriched in which species, allowing further insights. <br></p>
<p>Let's see how we can use <code>glycowork</code> to classify vertebrate versus invertebrate glycans! First, we need to annotate each known glycan using a binary code where <strong>1</strong> means <strong>vertebrate glycan</strong> and <strong>0</strong> means <strong>invertebrate glycan</strong>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">vertebrate</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s1">&#39;Chordata&#39;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span> <span class="o">==</span> <span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">Phylum</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now, we have to generate two sets of glycans that we will use as train set and test set. In addition, we also need their corresponding sets with classes labelled as <strong>0</strong> or <strong>1</strong>. One way to do it is to use the <code>general_split</code> function from the <code>glycowork.ml.train_test_split</code> module. This function assigns the data with a 80/20 ratio to train and test sets respectively.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">general_split</span><span class="p">(</span><span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span> <span class="o">==</span> <span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">vertebrate</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>It is now time to train a machine learning model on these data! Different machine learning algorithms are available in <code>glycowork</code> and here we can use the <code>train_ml_model</code> function from the <code>glycowork.ml.model_training</code> module. This classification method uses a standard machine learning model trained on short glycan motifs which may not be the most accurate, but can be sufficient for a preliminary exploration of our problem. Plus, it has the advantage of running faster than other more complex algorithms such as deep learning so... let's give it a try!</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">model_ft</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">X_test</span> <span class="o">=</span> <span class="n">train_ml_model</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span> <span class="n">feature_calc</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span>
                         <span class="n">return_features</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>As you can see, even this (quite) simple algorithm is enough to reach around 93% prediction accuracy on the test set! It also confirms that glycan composition is highly representative of their origin and allows efficiently discriminating vertebrate from invertebrate glycans! <br></p>
<p>One more step we can do is the analysis of the model we have generated. This is important, as we do not know yet what information is the most useful for this model to predict the origin of a glycan. Fortunately, the <code>analyze_ml_model</code> function can help us to understand on which basis the model predictions are made.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">analyze_ml_model</span><span class="p">(</span><span class="n">model_ft</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Interesting! It appears that the presence of the Gal(b1-4)GlcNAc motif (type II LacNAc) in glycans is quite discriminant of its belonging to the Chordata Phylum! <br></p>
<p>Next to knowing what is important for model prediction, it is often also crucial to know the limits of a model. In <code>glycowork</code>, you can use <code>get_mismatch</code> from the <code>glycowork.ml.model_training</code> module to get a sense of wrong classifications. This could for instance inform you about potential model biases.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_mismatch</span><span class="p">(</span><span class="n">model_ft</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_test</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here we can see that our model was especially confused by glycans which can occur in both phyla and therefore are ambiguous for classification ambiguous. Maybe in these scenarios a more powerful, deep learning-based model might be able to differentiate (at least a bit better)! <br></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Deep-Learning-Code-Snippet-">Deep Learning Code Snippet <a class="anchor" id="example4" /><a class="anchor-link" href="#Deep-Learning-Code-Snippet-"> </a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>For those of you that are blessed with GPU-access, we wrote up a minimal example for training a glycan-focused deep learning model. Even if you usually do not have access to a GPU, you can always go to Google Colab (click on the badge at the top of this site to open this notebook on Google Colab) and change the runtime type to GPU. Then you can paste the following code in a cell, execute, and see our model train!</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>First, we need to install torch_geometric for deep learning:</p>

<pre><code>!pip install torch-scatter -f https://pytorch-geometric.com/whl/torch-1.8.0+cu101.html
!pip install torch-sparse -f https://pytorch-geometric.com/whl/torch-1.8.0+cu101.html
!pip install torch-cluster -f https://pytorch-geometric.com/whl/torch-1.8.0+cu101.html
!pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.8.0+cu101.html
!pip install torch-geometric</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Then, we import the relevant <code>glycowork</code> modules:</p>

<pre><code>from glycowork.glycan_data.loader import df_species
from glycowork.ml.train_test_split import hierarchy_filter
from glycowork.ml.processing import split_data_to_train
from glycowork.ml import models
from glycowork.ml import model_training</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Finally, we can prepare the data and train a model to classify from which taxonomic kingdom a glycan came from, using our graph convolutional neural network, <code>SweetNet</code>.</p>

<pre><code>train_x, val_x, train_y, val_y, id_val, class_list, class_converter = hierarchy_filter(df_species,
                                                                                       rank = 'Kingdom')

dataloaders = split_data_to_train(train_x, val_x, train_y, val_y)
model = models.prep_model('SweetNet', len(class_list))
optimizer_ft, scheduler, criterion = model_training.training_setup(model, 100, 0.0005)
model_ft = model_training.train_model(model, dataloaders, criterion, optimizer_ft, scheduler,
                   num_epochs = 100)</code></pre>
<p>And that's it! You've just trained a state-of-the-art deep learning model with effectively five lines of code (excluding package imports of course).</p>

</div>
</div>
</div>
</div>
 

