---

title: ml


keywords: fastai
sidebar: home_sidebar

summary: "data processing and model training + analysis using machine learning"
description: "data processing and model training + analysis using machine learning"
nb_path: "03_ml.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 03_ml.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/glycowork/ml.html"><code>ml</code></a> contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan reprsentations. It currently contains the following modules:</p>
<ul>
<li><code>model_training</code> contains functions for training machine learning models</li>
<li><code>models</code> describes some examples for machine learning architectures applicable to glycans</li>
<li><code>processing</code> contains helper functions to prepare glycan data for model training</li>
<li><code>representation</code> can be used to analyze trained models and to obtain glycan representations</li>
<li><code>train_test_split</code> contains various data split functions to get appropriate training and test sets</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="model_training">model_training<a class="anchor-link" href="#model_training"> </a></h1><blockquote><p>contains functions for training machine learning models</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="EarlyStopping" class="doc_header"><code>class</code> <code>EarlyStopping</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L14" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>EarlyStopping</code>(<strong><code>patience</code></strong>=<em><code>7</code></em>, <strong><code>verbose</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Early stops the training if validation loss doesn't improve after a given patience.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="train_model" class="doc_header"><code>train_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L55" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>train_model</code>(<strong><code>model</code></strong>, <strong><code>dataloaders</code></strong>, <strong><code>criterion</code></strong>, <strong><code>optimizer</code></strong>, <strong><code>scheduler</code></strong>, <strong><code>num_epochs</code></strong>=<em><code>25</code></em>, <strong><code>patience</code></strong>=<em><code>50</code></em>, <strong><code>mode</code></strong>=<em><code>'classification'</code></em>)</p>
</blockquote>
<p>trains a deep learning model on predicting glycan properties</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr>
<td style="text-align:left">dataloaders (PyTorch object): dictionary of dataloader objects with keys 'train' and 'val'</td>
</tr>
<tr>
<td style="text-align:left">criterion (PyTorch object): PyTorch loss function</td>
</tr>
<tr>
<td style="text-align:left">optimizer (PyTorch object): PyTorch optimizer</td>
</tr>
<tr>
<td style="text-align:left">scheduler (PyTorch object): PyTorch learning rate decay</td>
</tr>
<tr>
<td style="text-align:left">num_epochs (int): number of epochs for training; default: 25</td>
</tr>
<tr>
<td style="text-align:left">patience (int): number of epochs without improvement until early stop; default: 50</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'classification' or 'regression'; default is binary classification</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the best model seen during training</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="training_setup" class="doc_header"><code>training_setup</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L172" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>training_setup</code>(<strong><code>model</code></strong>, <strong><code>epochs</code></strong>, <strong><code>lr</code></strong>, <strong><code>lr_decay_length</code></strong>=<em><code>0.5</code></em>, <strong><code>weight_decay</code></strong>=<em><code>0.001</code></em>, <strong><code>mode</code></strong>=<em><code>'multiclass'</code></em>)</p>
</blockquote>
<p>prepares optimizer, learning rate scheduler, and loss criterion for model training</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr>
<td style="text-align:left">epochs (int): number of epochs for training the model</td>
</tr>
<tr>
<td style="text-align:left">lr (float): learning rate</td>
</tr>
<tr>
<td style="text-align:left">lr_decay_length (float): proportion of epochs over which to decay the learning rate;default:0.5</td>
</tr>
<tr>
<td style="text-align:left">weight_decay (float): regularization parameter for the optimizer; default:0.001</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'multiclass': classification with multiple classes, 'binary':binary classification, 'regression': regression; default:'multiclass'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns optimizer, learning rate scheduler, and loss criterion objects</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="train_ml_model" class="doc_header"><code>train_ml_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L201" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>train_ml_model</code>(<strong><code>X_train</code></strong>, <strong><code>X_test</code></strong>, <strong><code>y_train</code></strong>, <strong><code>y_test</code></strong>, <strong><code>mode</code></strong>=<em><code>'classification'</code></em>, <strong><code>feature_calc</code></strong>=<em><code>False</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>return_features</code></strong>=<em><code>False</code></em>, <strong><code>feature_set</code></strong>=<em><code>['known', 'exhaustive']</code></em>, <strong><code>additional_features_train</code></strong>=<em><code>None</code></em>, <strong><code>additional_features_test</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>wrapper function to train standard machine learning models on glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">X_train, X_test (list or dataframe): either lists of glycans (needs feature_calc = True) or motif dataframes such as from annotate_dataset</td>
</tr>
<tr>
<td style="text-align:left">y_train, y_test (list): lists of labels</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'classification' or 'regression'; default:'classification'</td>
</tr>
<tr>
<td style="text-align:left">feature_calc (bool): set to True for calculating motifs from glycans; default:False</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our data; default:lib</td>
</tr>
<tr>
<td style="text-align:left">return_features (bool): whether to return calculated features; default:False</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default:['known','exhaustive']; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans) and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
</tbody>
</table>
<p>| additional_features_train (dataframe): additional features (apart from glycans) to be used for training. Has to be of the same length as X_train; default:None
| additional_features_test (dataframe): additional features (apart from glycans) to be used for evaluation. Has to be of the same length as X_test; default:None
| Returns:
| :-
| Returns trained model</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">vertebrate</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s1">&#39;Chordata&#39;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span><span class="o">==</span><span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">Phylum</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
<span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">general_split</span><span class="p">(</span><span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span><span class="o">==</span><span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">vertebrate</span><span class="p">)</span>
<span class="n">model_ft</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">X_test</span> <span class="o">=</span> <span class="n">train_ml_model</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span> <span class="n">feature_calc</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span>
                         <span class="n">return_features</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="analyze_ml_model" class="doc_header"><code>analyze_ml_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L262" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>analyze_ml_model</code>(<strong><code>model</code></strong>)</p>
</blockquote>
<p>plots relevant features for model prediction</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (model object): trained machine learning model from train_ml_model</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">analyze_ml_model</span><span class="p">(</span><span class="n">model_ft</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_mismatch" class="doc_header"><code>get_mismatch</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L283" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_mismatch</code>(<strong><code>model</code></strong>, <strong><code>X_test</code></strong>, <strong><code>y_test</code></strong>, <strong><code>n</code></strong>=<em><code>10</code></em>)</p>
</blockquote>
<p>analyzes misclassifications of trained machine learning model</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (model object): trained machine learning model from train_ml_model</td>
</tr>
<tr>
<td style="text-align:left">X_test (dataframe): motif dataframe used for validating model</td>
</tr>
<tr>
<td style="text-align:left">y_test (list): list of labels</td>
</tr>
<tr>
<td style="text-align:left">n (int): number of returned misclassifications; default:10</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns tuples of misclassifications and their predicted probability</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_mismatch</span><span class="p">(</span><span class="n">model_ft</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_test</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="models">models<a class="anchor-link" href="#models"> </a></h1><blockquote><p>describes some examples for machine learning architectures applicable to glycans</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(SweetNet)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(init_weights)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(prep_model)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="processing">processing<a class="anchor-link" href="#processing"> </a></h1><blockquote><p>contains helper functions to prepare glycan data for model training</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(dataset_to_graphs)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(dataset_to_dataloader)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(split_data_to_train)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="representation">representation<a class="anchor-link" href="#representation"> </a></h1><blockquote><p>can be used to analyze trained models and to obtain glycan representations</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(glycans_to_emb)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="train_test_split">train_test_split<a class="anchor-link" href="#train_test_split"> </a></h1><blockquote><p>contains various data split functions to get appropriate training and test sets</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hierarchy_filter" class="doc_header"><code>hierarchy_filter</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L37" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hierarchy_filter</code>(<strong><code>df_in</code></strong>, <strong><code>rank</code></strong>=<em><code>'Domain'</code></em>, <strong><code>min_seq</code></strong>=<em><code>5</code></em>, <strong><code>wildcard_seed</code></strong>=<em><code>False</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>None</code></em>, <strong><code>wildcard_name</code></strong>=<em><code>None</code></em>, <strong><code>r</code></strong>=<em><code>0.1</code></em>, <strong><code>col</code></strong>=<em><code>'target'</code></em>)</p>
</blockquote>
<p>stratified data split in train/test at the taxonomic level, removing duplicate glycans and infrequent classes</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df_in (dataframe): dataframe of glycan sequences and taxonomic labels</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which rank should be filtered; default is 'domain'</td>
</tr>
<tr>
<td style="text-align:left">min_seq (int): how many glycans need to be present in class to keep it; default is 5</td>
</tr>
<tr>
<td style="text-align:left">wildcard_seed (bool): set to True if you want to seed wildcard glycoletters; default is False</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list which glycoletters a wildcard encompasses</td>
</tr>
<tr>
<td style="text-align:left">wildcard_name (string): how the wildcard should be named in the IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">r (float): rate of replacement, default is 0.1 or 10%</td>
</tr>
<tr>
<td style="text-align:left">col (string): column name for glycan sequences; default: target</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns train_x, val_x (lists of glycans (strings) after stratified shuffle split)</td>
</tr>
<tr>
<td style="text-align:left">train_y, val_y (lists of taxonomic labels (mapped integers))</td>
</tr>
<tr>
<td style="text-align:left">id_val (taxonomic labels in text form (strings))</td>
</tr>
<tr>
<td style="text-align:left">class_list (list of unique taxonomic classes (strings))</td>
</tr>
<tr>
<td style="text-align:left">class_converter (dictionary to map mapped integers back to text labels)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_x</span><span class="p">,</span> <span class="n">val_x</span><span class="p">,</span> <span class="n">train_y</span><span class="p">,</span> <span class="n">val_y</span><span class="p">,</span> <span class="n">id_val</span><span class="p">,</span> <span class="n">class_list</span><span class="p">,</span> <span class="n">class_converter</span> <span class="o">=</span> <span class="n">hierarchy_filter</span><span class="p">(</span><span class="n">df_species</span><span class="p">,</span>
                                                                                       <span class="n">rank</span> <span class="o">=</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">train_x</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="seed_wildcard_hierarchy" class="doc_header"><code>seed_wildcard_hierarchy</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L9" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>seed_wildcard_hierarchy</code>(<strong><code>glycans</code></strong>, <strong><code>labels</code></strong>, <strong><code>wildcard_list</code></strong>, <strong><code>wildcard_name</code></strong>, <strong><code>r</code></strong>=<em><code>0.1</code></em>)</p>
</blockquote>
<p>adds dataframe rows in which glycan parts have been replaced with the appropriate wildcards</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr>
<td style="text-align:left">labels (list): list of labels used for prediction</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list which glycoletters a wildcard encompasses</td>
</tr>
<tr>
<td style="text-align:left">wildcard_name (string): how the wildcard should be named in the IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">r (float): rate of replacement, default is 0.1 or 10%</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of glycans (strings) and labels (flexible) where some glycan parts have been replaced with wildcard_name</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_x</span><span class="p">,</span> <span class="n">val_x</span><span class="p">,</span> <span class="n">train_y</span><span class="p">,</span> <span class="n">val_y</span><span class="p">,</span> <span class="n">id_val</span><span class="p">,</span> <span class="n">class_list</span><span class="p">,</span> <span class="n">class_converter</span> <span class="o">=</span> <span class="n">hierarchy_filter</span><span class="p">(</span><span class="n">df_species</span><span class="p">,</span>
                                                                                       <span class="n">rank</span> <span class="o">=</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">,</span>
                                                                                       <span class="n">wildcard_seed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                                                                                       <span class="n">wildcard_list</span> <span class="o">=</span> <span class="n">linkages</span><span class="p">,</span>
                                                                                       <span class="n">wildcard_name</span> <span class="o">=</span> <span class="s1">&#39;bond&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">train_x</span><span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">:])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="general_split" class="doc_header"><code>general_split</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L104" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>general_split</code>(<strong><code>glycans</code></strong>, <strong><code>labels</code></strong>, <strong><code>test_size</code></strong>=<em><code>0.2</code></em>)</p>
</blockquote>
<p>splits glycans and labels into train / test sets</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr>
<td style="text-align:left">labels (list): list of labels used for prediction</td>
</tr>
<tr>
<td style="text-align:left">test_size (float): % size of test set; default is 0.2 / 20%</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns X_train, X_test, y_train, y_test</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_x</span><span class="p">,</span> <span class="n">val_x</span><span class="p">,</span> <span class="n">train_y</span><span class="p">,</span> <span class="n">val_y</span> <span class="o">=</span> <span class="n">general_split</span><span class="p">(</span><span class="n">df_species</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                                              <span class="n">df_species</span><span class="o">.</span><span class="n">Species</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">train_x</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

