---

title: motif


keywords: fastai
sidebar: home_sidebar

summary: "Processing and analyzing glycan sequences on a motif-basis"
description: "Processing and analyzing glycan sequences on a motif-basis"
nb_path: "03_motif.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 03_motif.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/glycowork/motif.html"><code>motif</code></a> contains many functions to process glycans in various ways and use this processing to analyze glycans via curated motifs, graph features, and sequence features. It contains the following modules:</p>
<ul>
<li><code>analysis</code> contains functions for downstream analyses of important glycan motifs etc.</li>
<li><code>annotate</code> contains functions to extract curated motifs, graph features, and sequence features from glycan sequences</li>
<li><code>graph</code> is used to convert glycan sequences to graphs and contains helper functions to search for motifs / check whether two sequences describe the same sequence, etc.</li>
<li><code>processing</code> contains all the basic functions to process IUPACcondensed glycan sequences into glycoletters etc. Also contains <code>get_lib</code> to construct the glycoletter library</li>
<li><code>query</code> is used to interact with the databases contained in glycowork, delivering insights for sequences of interest</li>
<li><code>tokenization</code> has helper functions to map m/z--&gt;composition, composition--&gt;structure, structure--&gt;motif, and more</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="analysis">analysis<a class="anchor-link" href="#analysis"> </a></h2><blockquote><p>downstream analyses of important glycan motifs</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_pvals_motifs" class="doc_header"><code>get_pvals_motifs</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/analysis.py#L32" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_pvals_motifs</code>(<strong><code>df</code></strong>, <strong><code>glycan_col_name</code></strong>=<em><code>'glycan'</code></em>, <strong><code>label_col_name</code></strong>=<em><code>'target'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>thresh</code></strong>=<em><code>1.645</code></em>, <strong><code>sorting</code></strong>=<em><code>True</code></em>, <strong><code>feature_set</code></strong>=<em><code>['exhaustive']</code></em>, <strong><code>extra</code></strong>=<em><code>'termini'</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>, <strong><code>multiple_samples</code></strong>=<em><code>False</code></em>, <strong><code>motifs</code></strong>=<em><code>None</code></em>, <strong><code>estimate_speedup</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>returns enriched motifs based on label data or predicted data</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe containing glycan sequences and labels</td>
</tr>
<tr>
<td style="text-align:left">glycan_col_name (string): column name for glycan sequences; arbitrary if multiple_samples = True; default:'glycan'</td>
</tr>
<tr>
<td style="text-align:left">label_col_name (string): column name for labels; arbitrary if multiple_samples = True; default:'target'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
<tr>
<td style="text-align:left">thresh (float): threshold value to separate positive/negative; default is 1.645 for Z-scores</td>
</tr>
<tr>
<td style="text-align:left">sorting (bool): whether p-value dataframe should be sorted ascendingly; default: True</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default is 'exhaustive'; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans) and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
<tr>
<td style="text-align:left">extra (string): 'ignore' skips this, 'wildcards' allows for wildcard matching', and 'termini' allows for positional matching; default:'termini'</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcard names (such as '?1-?', 'Hex', 'HexNAc', 'Sia')</td>
</tr>
<tr>
<td style="text-align:left">multiple_samples (bool): set to True if you have multiple samples (rows) with glycan information (columns); default:False</td>
</tr>
<tr>
<td style="text-align:left">motifs (dataframe): can be used to pass a modified motif_list to the function; default:None</td>
</tr>
<tr>
<td style="text-align:left">estimate_speedup (bool): if True, pre-selects motifs for those which are present in glycans, not 100% exact; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe with p-values and corrected p-values for every glycan motif</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcOPN(b1-6)GlcOPN&#39;</span><span class="p">,</span>
          <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Glc(b1-3)Glc(b1-3)Glc&#39;</span><span class="p">]</span>
<span class="n">label</span> <span class="o">=</span> <span class="p">[</span><span class="mf">3.234</span><span class="p">,</span> <span class="mf">2.423</span><span class="p">,</span> <span class="mf">0.733</span><span class="p">,</span> <span class="mf">3.102</span><span class="p">,</span> <span class="mf">0.108</span><span class="p">]</span>
<span class="n">test_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;glycan&#39;</span><span class="p">:</span><span class="n">glycans</span><span class="p">,</span> <span class="s1">&#39;binding&#39;</span><span class="p">:</span><span class="n">label</span><span class="p">})</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Glyco-Motif enrichment p-value test&quot;</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">get_pvals_motifs</span><span class="p">(</span><span class="n">test_df</span><span class="p">,</span> <span class="s1">&#39;glycan&#39;</span><span class="p">,</span> <span class="s1">&#39;binding&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">10</span><span class="p">,:]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Glyco-Motif enrichment p-value test
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<style type="text/css">
#T_1a94b_row0_col0, #T_1a94b_row0_col1, #T_1a94b_row0_col2, #T_1a94b_row1_col0, #T_1a94b_row1_col1, #T_1a94b_row1_col2, #T_1a94b_row2_col0, #T_1a94b_row2_col1, #T_1a94b_row2_col2, #T_1a94b_row3_col0, #T_1a94b_row3_col1, #T_1a94b_row3_col2, #T_1a94b_row4_col0, #T_1a94b_row4_col1, #T_1a94b_row4_col2, #T_1a94b_row5_col0, #T_1a94b_row5_col1, #T_1a94b_row5_col2, #T_1a94b_row6_col0, #T_1a94b_row6_col1, #T_1a94b_row6_col2, #T_1a94b_row7_col0, #T_1a94b_row7_col1, #T_1a94b_row7_col2, #T_1a94b_row8_col0, #T_1a94b_row8_col1, #T_1a94b_row8_col2, #T_1a94b_row9_col0, #T_1a94b_row9_col1, #T_1a94b_row9_col2 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>
<table id="T_1a94b">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_1a94b_level0_col0" class="col_heading level0 col0" >motif</th>
      <th id="T_1a94b_level0_col1" class="col_heading level0 col1" >pval</th>
      <th id="T_1a94b_level0_col2" class="col_heading level0 col2" >corr_pval</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_1a94b_level0_row0" class="row_heading level0 row0" >4</th>
      <td id="T_1a94b_row0_col0" class="data row0 col0" >GlcNAc</td>
      <td id="T_1a94b_row0_col1" class="data row0 col1" >0.013469</td>
      <td id="T_1a94b_row0_col2" class="data row0 col2" >0.394527</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row1" class="row_heading level0 row1" >19</th>
      <td id="T_1a94b_row1_col0" class="data row1 col0" >b1-4</td>
      <td id="T_1a94b_row1_col1" class="data row1 col1" >0.013469</td>
      <td id="T_1a94b_row1_col2" class="data row1 col2" >0.394527</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row2" class="row_heading level0 row2" >8</th>
      <td id="T_1a94b_row2_col0" class="data row2 col0" >Man</td>
      <td id="T_1a94b_row2_col1" class="data row2 col1" >0.025198</td>
      <td id="T_1a94b_row2_col2" class="data row2 col2" >0.590671</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row3" class="row_heading level0 row3" >11</th>
      <td id="T_1a94b_row3_col0" class="data row3 col0" >a1-3</td>
      <td id="T_1a94b_row3_col1" class="data row3 col1" >0.025636</td>
      <td id="T_1a94b_row3_col2" class="data row3 col2" >0.590671</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row4" class="row_heading level0 row4" >13</th>
      <td id="T_1a94b_row4_col0" class="data row4 col0" >a1-6</td>
      <td id="T_1a94b_row4_col1" class="data row4 col1" >0.091752</td>
      <td id="T_1a94b_row4_col2" class="data row4 col2" >0.958241</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row5" class="row_heading level0 row5" >26</th>
      <td id="T_1a94b_row5_col0" class="data row5 col0" >GlcNAc(b1-4)GlcNAc</td>
      <td id="T_1a94b_row5_col1" class="data row5 col1" >0.091752</td>
      <td id="T_1a94b_row5_col2" class="data row5 col2" >0.958241</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row6" class="row_heading level0 row6" >33</th>
      <td id="T_1a94b_row6_col0" class="data row6 col0" >Man(a1-3)Man</td>
      <td id="T_1a94b_row6_col1" class="data row6 col1" >0.091752</td>
      <td id="T_1a94b_row6_col2" class="data row6 col2" >0.958241</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row7" class="row_heading level0 row7" >34</th>
      <td id="T_1a94b_row7_col0" class="data row7 col0" >Man(a1-6)Man</td>
      <td id="T_1a94b_row7_col1" class="data row7 col1" >0.091752</td>
      <td id="T_1a94b_row7_col2" class="data row7 col2" >0.958241</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row8" class="row_heading level0 row8" >35</th>
      <td id="T_1a94b_row8_col0" class="data row8 col0" >Man(b1-4)GlcNAc</td>
      <td id="T_1a94b_row8_col1" class="data row8 col1" >0.091752</td>
      <td id="T_1a94b_row8_col2" class="data row8 col2" >0.958241</td>
    </tr>
    <tr>
      <th id="T_1a94b_level0_row9" class="row_heading level0 row9" >10</th>
      <td id="T_1a94b_row9_col0" class="data row9 col0" >a1-2</td>
      <td id="T_1a94b_row9_col1" class="data row9 col1" >0.130826</td>
      <td id="T_1a94b_row9_col2" class="data row9 col2" >0.980276</td>
    </tr>
  </tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_representative_substructures" class="doc_header"><code>get_representative_substructures</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/analysis.py#L101" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_representative_substructures</code>(<strong><code>enrichment_df</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>builds minimal glycans that contain enriched motifs from get_pvals_motifs</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">enrichment_df (dataframe): output from get_pvals_motifs</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns up to 10 minimal glycans in a list</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="make_heatmap" class="doc_header"><code>make_heatmap</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/analysis.py#L148" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>make_heatmap</code>(<strong><code>df</code></strong>, <strong><code>mode</code></strong>=<em><code>'sequence'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>feature_set</code></strong>=<em><code>['known']</code></em>, <strong><code>extra</code></strong>=<em><code>'termini'</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>, <strong><code>datatype</code></strong>=<em><code>'response'</code></em>, <strong><code>rarity_filter</code></strong>=<em><code>0.05</code></em>, <strong><code>filepath</code></strong>=<em><code>''</code></em>, <strong><code>index_col</code></strong>=<em><code>'target'</code></em>, <strong><code>estimate_speedup</code></strong>=<em><code>False</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>clusters samples based on glycan data (for instance glycan binding etc.)</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycan data, rows are samples and columns are glycans</td>
</tr>
<tr>
<td style="text-align:left">mode (string): whether glycan 'sequence' or 'motif' should be used for clustering; default:sequence</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default is 'exhaustive'; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans) and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
<tr>
<td style="text-align:left">extra (string): 'ignore' skips this, 'wildcards' allows for wildcard matching', and 'termini' allows for positional matching; default:'termini'</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcard names (such as 'bond', 'Hex', 'HexNAc', 'Sia')</td>
</tr>
<tr>
<td style="text-align:left">datatype (string): whether df comes from a dataset with quantitative variable ('response') or from presence_to_matrix ('presence')</td>
</tr>
<tr>
<td style="text-align:left">rarity_filter (float): proportion of samples that need to have a non-zero value for a variable to be included; default:0.05</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): absolute path including full filename allows for saving the plot</td>
</tr>
<tr>
<td style="text-align:left">index_col (string): default column to convert to dataframe index; default:'target'</td>
</tr>
<tr>
<td style="text-align:left">estimate_speedup (bool): if True, pre-selects motifs for those which are present in glycans, not 100% exact; default:False</td>
</tr>
<tr>
<td style="text-align:left">**kwargs: keyword arguments that are directly passed on to seaborn clustermap</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Prints clustermap                         </td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Glc(b1-3)Glc(b1-3)Glc&#39;</span><span class="p">]</span>
<span class="n">label</span> <span class="o">=</span> <span class="p">[</span><span class="mf">3.234</span><span class="p">,</span> <span class="mf">2.423</span><span class="p">,</span> <span class="mf">0.733</span><span class="p">,</span> <span class="mf">3.102</span><span class="p">,</span> <span class="mf">0.108</span><span class="p">]</span>
<span class="n">label2</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.134</span><span class="p">,</span> <span class="mf">0.345</span><span class="p">,</span> <span class="mf">1.15</span><span class="p">,</span> <span class="mf">0.233</span><span class="p">,</span> <span class="mf">2.981</span><span class="p">]</span>
<span class="n">label3</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.334</span><span class="p">,</span> <span class="mf">0.245</span><span class="p">,</span> <span class="mf">1.55</span><span class="p">,</span> <span class="mf">0.133</span><span class="p">,</span> <span class="mf">2.581</span><span class="p">]</span>
<span class="n">test_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">label</span><span class="p">,</span> <span class="n">label2</span><span class="p">,</span> <span class="n">label3</span><span class="p">],</span> <span class="n">columns</span> <span class="o">=</span> <span class="n">glycans</span><span class="p">)</span>

<span class="n">make_heatmap</span><span class="p">(</span><span class="n">test_df</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;motif&#39;</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;known&#39;</span><span class="p">,</span> <span class="s1">&#39;exhaustive&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_png output_subarea ">
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA94AAAPeCAYAAAD6bcIrAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAA9hAAAPYQGoP6dpAACZ6ElEQVR4nOzdeXRN1///8ddNIjeRSCKmGIJGDDGHVA1VVCvGoi2lpohSVXNNaamphJYqqtqaoqVoTfVBVUtjpoIolZpJTVVTEJLI8PujP/fb21wkkuMano+1zlruPvvs8z5XP+v7fdn77GtKS0tLEwAAAAAAMISDvQsAAAAAAOBxRvAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAzklNGOty4cM7IOIFNy5PWzdwkAAAAAkCEZDt5KTTGwDAAAAAAAHk8ZD95pqQaWAQAAAADA4ynDwTstJdnIOgAAAAAAeCxlYqk5M94AAAAAAGQWS80BAAAAADBQxoN3yi0DywAAAAAA4PHEUnMAAAAAAAzE5moAAAAAABiId7zxWJs+fbqmT5+uEydOSJLKlSun999/X40aNbJvYciUsLAwJSQk2LsMAHiouLi4KDw83N5lAAAyIBNLzVMMLAMwRpEiRTRu3DiVLFlSaWlpmjt3rpo3b649e/aoXLly9i4PGZSQkKBJkybZuwwAeKj069fP3iUAADIoE5ursdQcj55mzZpZfR4zZoymT5+u7du3E7wBAAAAPBAsNccTIyUlRd99953i4+NVo0YNe5cDAAAA4AmRic3V+DkxPDwSExOVmJho1WY2m2U2m9P13bdvn2rUqKGEhAS5u7tr2bJlKlu27IMqFQAAAMATziHDPdNSOTgemiM8PFyenp5Wx502mCldurSio6O1Y8cOvfXWW+rUqZMOHDiQXf8bAgAAAIC74ne88UgKCwtT//79rdpszXZLkrOzs/z9/SVJVatW1c6dOzV58mR98cUXhtcJAAAAAJnYXI2l5nh43GlZeUakpqamW6YOAAAAAEZhczU81sLCwtSoUSMVLVpU165d0zfffKPIyEj9+OOP9i4NAAAAwBOCnxPDY+38+fPq2LGjzp49K09PT1WsWFE//vijXnzxRXuXBgAAAOAJwTveeKzNmjXL3iUAAAAAeMIRvAEAAAAAMBC/4w0AAAAAgIHYXA0AAAAAAAOxuRoAAAAAAAbiHW8AAAAAAAzEUnMAAAAAAAyU8eCdzFJzAAAAAAAyixlvAAAAAAAMxOZqAAAAAAAYiM3VAAAAAAAwEEvNAQAAAAAwEJurAQAAAABgoEzMeKcZWAYAAAAAAI8nZrwBAAAAADAQ73gDAAAAAGAgdjUHAAAAAMBAmfgd7xQDywAAAAAA4PHEjDcAAAAAAAbKcPBOY8YbAAAAAIBMY8YbAAAAAAADsas5AAAAAAAGysTveLPUHAAAAACAzGKpOQAAAAAABuLnxAAAAAAAMFAmZrzTDCwDAAAAAIDHE5urAQAAAABgoIz/jjebqwEAAAAAkGksNQcAAAAAwEBsrgYAAAAAgIH4OTEAAAAAAAzEUnMAAAAAAAzEUnMAAAAAAAyU8V3NWWoOAAAAAECmZXzGO5ngDQAAAABAZmU8eKcRvAEAAAAAyCw2VwMAAAAAwEAZf8ebpeYAAAAAAGQav+MNAAAAAICB2FwNAAAAAAADZXypeRrveAMAAAAAkFlsrgYAAAAAgIHYXA0AAAAAAAMx4w0AAAAAgIEyMeNN8AYAAAAAILOY8QYAAAAAwECZCN4GVgEAAAAAwGOKpeYAAAAAABgo48GbpeYAAAAAAGRaJma8jSwDMEZ4eLiWLl2qP/74Q66urqpZs6bGjx+v0qVL27s0AAAAAE8I3vHGY23Dhg16++239fTTTys5OVnvvvuuGjRooAMHDsjNzc3e5QEAAAB4AmRiqbmRZQDGWLNmjdXniIgI5c+fX7t27dJzzz1np6oAAAAAPElYao5HUmJiohITE63azGazzGbzXa+Li4uTJHl7extWGwAAAAD8m0NGO6alcnA8PEd4eLg8PT2tjvDw8Lv+N5yamqq+ffuqVq1aKl++fJb/xwMAAAAAGZHxGe8Uk5F1AJkSFham/v37W7Xda7b77bff1v79+7V582YjSwMAAAAAK7zjjUdSRpaV/1vPnj21cuVKbdy4UUWKFMnSva9evar169erdOnSCggIyNJYAAAAAB5/mQjezHjj0ZOWlqZevXpp2bJlioyM1FNPPZXpMVq3bq3nnntOPXv21M2bNxUUFKQTJ04oLS1NCxcu1CuvvGJA5QAAAAAeFxkO3qksNccj6O2339Y333yj77//Xrly5dK5c+ckSZ6ennJ1dc3QGBs3btR7770nSVq2bJnS0tJ05coVzZ07Vx988AHBGwAAAMBdsdQcj7Xp06dLkurWrWvVPmfOHIWEhGRojLi4OMsu6GvWrNErr7yinDlzqkmTJho4cGB2lgsAAADgMcSMNx5raWlpWR7D19dX27Ztk7e3t9asWaOFCxdKki5fviwXF5csjw8AAADg8cY73sA99O3bV+3atZO7u7uKFStmmT3fuHGjKlSoYN/iAAAAADz0CN6ADVevXpWHh4ckqUePHnrmmWcUGxurF198UQ4ODpIkPz8/ffDBB/YsEwAAAMAjgKXmgA25c+fW2bNnlT9/fj3//PNaunSpqlatatWnSZMmdqoOAAAAwKMk4zPeaQRvPDnc3d118eJF5c+fX5GRkbp165a9SwIAAADwiGJXc8CGF154QfXq1VNAQIAkqWXLlnJ2drbZd/369Q+yNAAAAACPmAwH75RUByPrAB4q8+bN09y5c3X06FFt2LBB5cqVU86cOe1dFgAAAIBHEJurATa4urqqe/fukqSoqCiNHz9eXl5e9i0KAAAAwCOJzdWAe/jll1/sXQIAAACAR1jGgzebq+EJ0r9/f40ePVpubm7q37//Xft+/PHHD6gqAAAAAI8idjUHbNizZ49lJ/Pdu3fLZOK/fwAAAAD3JxObqxE88OT49/LyyMhI+xUCAAAA4JGX4a3K09JMHBwPzfEghYaG6tq1a+na4+PjFRoa+kBrAQAAAPDoYcYbuIe5c+dq3LhxypUrl1X7zZs39dVXX2n27Nl2qgwAAADAo4B3vIE7uHr1qtLS0pSWlqZr167JxcXFci4lJUWrV69W/vz57VghAAAAgEcBu5oDd+Dl5SWTySSTyaRSpUqlO28ymTRy5Eg7VAYAAADgUZLxpeYEbzxhfvnlF6Wlpen555/XkiVL5O3tbTnn7OysYsWKqVChQnasEAAAAMCjgKXmwB3UqVNHknT8+HH5+vrKwSHDexECAAAAgEXGZ7xF8MaTqVixYrpy5YpmzZqlmJgYSVK5cuUUGhoqT09PO1cHAAAA4GGXiXe8jSwDeHhFRUUpODhYrq6uqlatmiTp448/1pgxY7R27VpVqVLFzhUCAAAAeJhlPHgz440nVL9+/fTSSy9pxowZcnL6538yycnJeuONN9S3b19t3LjRzhUCAAAAeJix1By4h6ioKKvQLUlOTk4aNGiQgoKC7FgZAAAAgEdBxjdXI3jjCeXh4aHY2FiVKVPGqv3PP/9Urly57FQVAAAAgEdFhoN3spFVAA+x1157TV26dNGECRNUs2ZNSdKWLVs0cOBAtW3b1s7VAQAAAHjYMeMN3MOECRNkMpnUsWNHJScnKy0tTc7Oznrrrbc0btw4e5cHAAAA4CGX8c3VyN14Qjk7O2vy5MkKDw/X0aNHJUklSpRQzpw57VwZAAAAgEcBm6sBdxAaGpqhfrNnzza4EgAAAACPskz8nBjwZImIiFCxYsUUGBiotDR+yB4AAADA/cn4jLeJGW88Wd566y0tWLBAx48fV+fOndW+fXt5e3vbuywAAAAAjxiHjHZM5eB4iI4HYdq0aTp79qwGDRqk//3vf/L19VXr1q31448/MgMOAAAAIMPYXA24C7PZrLZt26pt27Y6efKkIiIi1KNHDyUnJ+v333+Xu7u7vUsEAAAA8JBjczUggxwcHGQymZSWlqaUlBR7lwMAAADgEcGMN3AXiYmJWrp0qWbPnq3NmzeradOm+vTTT9WwYUM5OGT4TQ0AAAAAT7BMzHgDT5YePXpo4cKF8vX1VWhoqBYsWKC8efPauywAAAAAjxhmvIE7+Pzzz1W0aFH5+flpw4YN2rBhg81+S5cufcCVAQAAAHiU8DvewB107NhRJn5GDwAAAEAWZeJ3vI0sA3j4RERE2LsEAAAAAI8BZrwBAAAAADAQm6sBAAAAAGAgNlcDAAAAAMBALDUHAAAAAMBALDUHAAAAAMBALDUHAAAAAMBAzHgDAAAAAGCgTLzjnWZkHQAAAAAAPJbYXA0AAAAAAAOx1BwAAAAAAAOxuRoAAAAAAAbKxIw373jj0bRx40Z99NFH2rVrl86ePatly5apRYsW9i4LAAAAwBOCd7zx2IuPj1elSpUUGhqql19+2d7lAAAAAHjCsKs5HnuNGjVSo0aN7F0GAAAAgCcUm6sBAAAAAGAgZrzxSEpMTFRiYqJVm9lsltlstlNFAAAAAGCbQ0Y7pnBwPERHeHi4PD09rY7w8HABAAAAwMMmwzPeacx44yESFham/v37W7Ux2w0AAADgYcSu5ngksawcAAAAwKOC3/HGY+/69es6cuSI5fPx48cVHR0tb29vFS1a1I6VAQAAAHgSsLkaHntRUVGqV6+e5fPtJeqdOnVSRESEnaoCAAAA8KRgxhuPvbp16yotjf9+AQAAANgH73gDAAAAAGAgdjUHAAAAAMBAGQ7eySzVBQAAAAAg0zIx4w0AAAAAADIrE5ur8ZY3AAAAAACZxeZqAAAAAAAYiN/xBgAAAADAQPyONwAAAAAABsr45mrsag4AAAAAQKZl/OfEmPEGAAAAACDTMvFzYgRvAAAAAAAyi83VAAAAAAAwUMY3V+MdbwAAAAAAMo2l5gAAAAAAGCgTM96pRtYBAAAAAMBjiXe8AQAAAAAwEEvNAQAAAAAwEJurAQAAAABgIJaaAwAAAABgIDZXAwAAAADAQMx4AwAAAABgIDZXAwAACgsLU0JCgr3LQCacP39e/fr1s3cZyAQXFxeFh4fbuwwAdsBScwAAoISEBE2aNMneZQCPNf6hBHhyZXzGm13NAQAAAADItIzPeIsZbwAAAAAAMivjm6sx4w0AAAAAQKaxuRoAAAAAAAZiczUAAAAAAAzEUnMAAAAAAAzEUnMAAAAAAAzEUnMAAAAAAAyUid/xJngDAAAAAJBZzHgDAAAAAGCgjG+uxjveAAAAAABkWiaWmhO8AQAAAADILJaaAwAAAABgIGa8AQAAAAAwEDPeAAAAAAAYKOObqzHjDQAAAABAprHUHAAAAAAAA7HUHAAAAAAAA7HUHAAAAAAAAzHjDQAAAACAgXjHGwAAAAAAA2U8eIvgDQAAAABAZmX8He9UlpoDAAAAAJBZpjTWkAMAAAAAYBgHexcAAAAAAMDjjOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjg/RAzmUxavny5vcsAAAAAAGTBEx+8//77b7311lsqWrSozGazfHx8FBwcrC1btti7NAAAAADAY8DJ3gXY2yuvvKKkpCTNnTtXfn5++uuvv7Ru3TpdvHjR3qUBAAAAAB4DT/SM95UrV7Rp0yaNHz9e9erVU7FixVStWjWFhYXppZdekiR9/PHHqlChgtzc3OTr66sePXro+vXrljEiIiLk5eWllStXqnTp0sqZM6deffVV3bhxQ3PnzlXx4sWVO3du9e7dWykpKZbrihcvrtGjR6tt27Zyc3NT4cKFNW3atLvW++eff6p169by8vKSt7e3mjdvrhMnTljOR0ZGqlq1anJzc5OXl5dq1aqlkydPZu+XBgAAAADIlCc6eLu7u8vd3V3Lly9XYmKizT4ODg6aMmWKfv/9d82dO1fr16/XoEGDrPrcuHFDU6ZM0cKFC7VmzRpFRkaqZcuWWr16tVavXq2vv/5aX3zxhRYvXmx13UcffaRKlSppz549GjJkiPr06aOffvrJZh23bt1ScHCwcuXKpU2bNmnLli1yd3dXw4YNlZSUpOTkZLVo0UJ16tTRb7/9pm3btqlbt24ymUzZ82UBAAAAAO6LKS0tLc3eRdjTkiVL1LVrV928eVNVqlRRnTp11KZNG1WsWNFm/8WLF6t79+66cOGCpH9mvDt37qwjR46oRIkSkqTu3bvr66+/1l9//SV3d3dJUsOGDVW8eHF9/vnnkv6Z8Q4ICNAPP/xgGbtNmza6evWqVq9eLemfzdWWLVumFi1aaN68efrggw8UExNjCdNJSUny8vLS8uXLFRQUpDx58igyMlJ16tQx5ssCAAAAAGTaEz3jLf3zjveZM2e0YsUKNWzYUJGRkapSpYoiIiIkST///LPq16+vwoULK1euXOrQoYMuXryoGzduWMbImTOnJXRLUoECBVS8eHFL6L7ddv78eat716hRI93nmJgYm3Xu3btXR44cUa5cuSwz9d7e3kpISNDRo0fl7e2tkJAQBQcHq1mzZpo8ebLOnj2b1a8HAAAAAJBFT3zwliQXFxe9+OKLGjZsmLZu3aqQkBANHz5cJ06cUNOmTVWxYkUtWbJEu3btsryHnZSUZLk+R44cVuOZTCabbampqfdd4/Xr11W1alVFR0dbHYcOHdLrr78uSZozZ462bdummjVratGiRSpVqpS2b99+3/cEAAAAAGQdwduGsmXLKj4+Xrt27VJqaqomTpyo6tWrq1SpUjpz5ky23ee/oXj79u0KCAiw2bdKlSo6fPiw8ufPL39/f6vD09PT0i8wMFBhYWHaunWrypcvr2+++Sbb6gUAAAAAZN4THbwvXryo559/XvPmzdNvv/2m48eP67vvvtOHH36o5s2by9/fX7du3dLUqVN17Ngxff3115Z3tLPDli1b9OGHH+rQoUOaNm2avvvuO/Xp08dm33bt2ilv3rxq3ry5Nm3apOPHjysyMlK9e/fWqVOndPz4cYWFhWnbtm06efKk1q5dq8OHD98xyAMAAAAAHown+ne83d3d9cwzz2jSpEk6evSobt26JV9fX3Xt2lXvvvuuXF1d9fHHH2v8+PEKCwvTc889p/DwcHXs2DFb7v/OO+8oKipKI0eOlIeHhz7++GMFBwfb7JszZ05t3LhRgwcP1ssvv6xr166pcOHCql+/vjw8PHTz5k398ccfmjt3ri5evKiCBQvq7bff1ptvvpkttQIAAAAA7s8Tv6u5vRQvXlx9+/ZV37597V0KAAAAAMBAT/RScwAAAAAAjEbwBgAAAADAQCw1BwAAAADAQMx4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAZysncBwIN268Ixe5cAPNbcCj9n7xKAx15+Ny97lwA8EU5d2m/vEvCYYMYbAAAAAAADEbwBAAAAADAQwRsAAAAAHgImk0nLly+3aps1a5YaNGhg+RwSEqIWLVo82ML+vwMHDqhIkSKKj4+3y/0fZQRvAAAAAMiClJQU1axZUy+//LJVe1xcnHx9ffXee+/d17gJCQkaNmyYhg8fnqnrNm7cqGbNmqlQoUI2w7wtly5dUq9evVS6dGm5urqqaNGi6t27t+Li4ix9ypYtq+rVq+vjjz+2urZu3boymUwaN25cunGbNGkik8mkESNGZOoZHjcEbwAAAADIAkdHR0VERGjNmjWaP3++pb1Xr17y9vbOdHC+bfHixfLw8FCtWrUydV18fLwqVaqkadOmZfiaM2fO6MyZM5owYYL2799veZ4uXbpY9evcubOmT5+u5ORkq3ZfX19FRERYtZ0+fVrr1q1TwYIFM1X/44jgDQAAAABZVKpUKY0bN069evXS2bNn9f3332vhwoX66quvlCNHDvn7+2vChAlW10RHR8tkMunIkSM2x1y4cKGaNWtm89zIkSOVL18+eXh4qHv37kpKSrKca9SokT744AO1bNkyw/WXL19eS5YsUbNmzVSiRAk9//zzGjNmjP73v/9ZhewXX3xRly5d0oYNG6yub9q0qS5cuKAtW7ZY2ubOnasGDRoof/78Vn2//vprBQUFKVeuXPLx8dHrr7+u8+fPW85HRkbKZDJp3bp1CgoKUs6cOVWzZk0dPHgww8/zsCF4AwAAAIANiYmJunr1qtWRmJh4x/69evVSpUqV1KFDB3Xr1k3vv/++KlWqJJPJpNDQUM2ZM8eq/5w5c/Tcc8/J39/f5nibN29WUFBQuvZ169YpJiZGkZGRWrBggZYuXaqRI0dm7WFtiIuLk4eHh5yc/u9XqJ2dnVW5cmVt2rTJqq+zs7PatWtn9YwREREKDQ1NN+6tW7c0evRo7d27V8uXL9eJEycUEhKSrt97772niRMnKioqSk5OTjbHelQQvAEAAADAhvDwcHl6elod4eHhd+xvMpk0ffp0rVu3TgUKFNCQIUMs50JCQnTw4EH9+uuvkv4Jn998880dw+SVK1cUFxenQoUKpTvn7Oys2bNnq1y5cmrSpIlGjRqlKVOmKDU1NYtP/H8uXLig0aNHq1u3bunOFSpUSCdPnkzXHhoaqm+//Vbx8fHauHGj4uLi1LRpU5v9GjVqJD8/P1WvXl1TpkzRDz/8oOvXr1v1GzNmjOrUqaOyZctqyJAh2rp1qxISErLtGR8kgjcAAAAA2BAWFqa4uDirIyws7K7XzJ49Wzlz5tTx48d16tQpS3uhQoXUpEkTzZ49W5L0v//9T4mJiWrVqpXNcW7evClJcnFxSXeuUqVKypkzp+VzjRo1dP36df35558Zeq6xY8fK3d3dcsTGxlqdv3r1qpo0aaKyZcva3BTN1dVVN27csFlXyZIltXjxYs2ePVsdOnSwmi2/bdeuXWrWrJmKFi2qXLlyqU6dOpKUro6KFSta/nz7PfF/L0l/lBC8AQAAAMAGs9ksDw8Pq8NsNt+x/9atWzVp0iStXLlS1apVU5cuXZSWlmY5/8Ybb2jhwoW6efOm5syZo9dee80qQP9bnjx5ZDKZdPny5Wx/ru7duys6Otpy/HtW/dq1a2rYsKFy5cqlZcuWKUeOHOmuv3TpkvLly2dz7NDQUE2bNk2LFy+2OZsfHx+v4OBgeXh4aP78+dq5c6eWLVsmSVbvqUuyurfJZJKkbJ3Vf5AI3gAAAACQRTdu3FBISIjeeust1atXT7NmzdKvv/6qzz//3NKncePGcnNz0/Tp07VmzZq7vrPs7OyssmXL6sCBA+nO7d271zIjLknbt2+Xu7u7fH19M1Srt7e3/P39LcftWemrV6+qQYMGcnZ21ooVK2zOtkvS/v37FRgYaPPc66+/rn379ql8+fIqW7ZsuvN//PGHLl68qHHjxql27doqU6bMIzuLnRkEbwAAAADIorCwMKWlpVl+y7p48eKaMGGCBg0apBMnTkj652fHQkJCFBYWppIlS6pGjRp3HTM4OFibN29O156UlKQuXbrowIEDWr16tYYPH66ePXvKweGfeHf9+nXLbLYkHT9+XNHR0emWcv/b7dAdHx+vWbNm6erVqzp37pzOnTunlJQUS78TJ07o9OnTeuGFF2yOkzt3bp09e1br1q2zeb5o0aJydnbW1KlTdezYMa1YsUKjR4++6/fwOCB4AwAAAEAWbNiwQdOmTdOcOXOslo6/+eabqlmzptWS8y5duigpKUmdO3e+57hdunTR6tWrFRcXZ9Vev359lSxZUs8995xee+01vfTSS1bvYkdFRSkwMNAyK92/f38FBgbq/fffv+O9du/erR07dmjfvn3y9/dXwYIFLce/3x1fsGCBGjRooGLFit1xLC8vL7m5udk8ly9fPkVEROi7775T2bJlNW7cuHQ/s/Y4MqX9+6UD4Alw68Ixe5cAPNbcCj9n7xKAx15+Ny97lwA8EU5d2p/tY27atEn169fXn3/+qQIFCtyzf6tWrVSlSpV7bur2ICQlJalkyZL65ptvVKtWLXuX80hhxhsAAAAADJaYmKhTp05pxIgRatWqVYZCtyR99NFHcnd3N7i6jImNjdW7775L6L4PzHjjicOMN2AsZrwB4zHjDTwY2TnjHRERoS5duqhy5cpasWKFChcunG1j4+FH8MYTh+ANGIvgDRiP4A08GEYsNceTiaXmAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBnOxdAJ4cYWFhSkhIyJaxJk2alC3jAAAAAIDRCN54YBISEgjMAAAAAJ44LDUHAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAMiiunXrqm/fvvYuAw8pgjcAAAAAGGzp0qVq0KCB8uTJI5PJpOjo6Htec+LECXXp0kVPPfWUXF1dVaJECQ0fPlxJSUnGF4xs5WTvAgAAAADgcRcfH69nn31WrVu3VteuXTN0zR9//KHU1FR98cUX8vf31/79+9W1a1fFx8drwoQJBleM7ETwBgAAAIBskJycrJ49e+rrr79Wjhw59NZbb2nUqFEymUzq0KGDpH9msTOqYcOGatiwoeWzn5+fDh48qOnTpxO8HzEsNQcAAAAAGxITE3X16lWrIzEx8Y79586dKycnJ/3666+aPHmyPv74Y82cOTNba4qLi5O3t3e2jgnjEbwBAAAAwIbw8HB5enpaHeHh4Xfs7+vrq0mTJql06dJq166devXqpUmTJmVbPUeOHNHUqVP15ptvZtuYeDAI3gAAAABgQ1hYmOLi4qyOsLCwO/avXr26TCaT5XONGjV0+PBhpaSk3PNe3bt3l7u7u+X4r9OnT6thw4Zq1apVht8Rx8ODd7wBAAAAwAaz2Syz2fxA7jVq1CgNGDDA5rkzZ86oXr16qlmzpr788ssHUg+yF8EbAAAAALLBjh07rD5v375dJUuWlKOj4z2vzZ8/v/Lnz5+u/fTp06pXr56qVq2qOXPmyMGBRcuPIoI3AAAAAGSD2NhY9e/fX2+++aZ2796tqVOnauLEiZKkS5cuKTY2VmfOnJEkHTx4UJLk4+MjHx8fm+OdPn1adevWVbFixTRhwgT9/ffflnN3ugYPJ4I3AAAAAGSDjh076ubNm6pWrZocHR3Vp08fdevWTZK0YsUKde7c2dK3TZs2kqThw4drxIgRNsf76aefdOTIER05ckRFihSxOpeWlmbMQ8AQpjT+xvCA9OvXL1t3dbxfty4cs3cJwGPNrfBz9i4BeOzld/OydwnAE+HUpf32LgGPCV4QAAAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAIDH0IgRI1SgQAGZTCYtX77c3uVk2KxZs9SgQQN7lyFJ+vzzz9WsWbMsj0PwBgAAAIDHTExMjEaOHKkvvvhCZ8+eVaNGjexdUoYkJCRo2LBhGj58eKau+/LLL1W3bl15eHjIZDLpypUr97zm4sWLatiwoQoVKiSz2SxfX1/17NlTV69etfQJDQ3V7t27tWnTpsw+ihWCNwAAAAA8RJKSkrI8xtGjRyVJzZs3l4+Pj8xmc5bHfBAWL14sDw8P1apVK1PX3bhxQw0bNtS7776b4WscHBzUvHlzrVixQocOHVJERIR+/vlnde/e3dLH2dlZr7/+uqZMmZKpetLdK0tXAwAAAACUmpqqDz/8UP7+/jKbzSpatKjGjBkjSdq3b5+ef/55ubq6Kk+ePOrWrZuuX79uuTYkJEQtWrTQmDFjVKhQIZUuXVqS9Oeff6p169by8vKSt7e3mjdvrhMnTtyzlhEjRliWRzs4OMhkMkmS6tatq759+1r1bdGihUJCQiyfExMTNXjwYPn6+spsNsvf31+zZs2ynN+wYYOqVasms9msggULasiQIUpOTracX7x4sSpUqGB51hdeeEHx8fGW8zNnzlRAQIBcXFxUpkwZffbZZ1b1LFy4MN3S7p07d+rFF19U3rx55enpqTp16mj37t1Wffr27ashQ4aoevXq9/x+bsudO7feeustBQUFqVixYqpfv7569OiRbna7WbNmWrFihW7evJnhsf+L4A0AAAAANiQmJurq1atWR2Jios2+YWFhGjdunIYNG6YDBw7om2++UYECBRQfH6/g4GDlzp1bO3fu1Hfffaeff/5ZPXv2tLp+3bp1OnjwoH766SetXLlSt27dUnBwsHLlyqVNmzZpy5Ytcnd3V8OGDe85Iz5gwADNmTNHknT27FmdPXs2w8/csWNHLViwQFOmTFFMTIy++OILubu7S5JOnz6txo0b6+mnn9bevXs1ffp0zZo1Sx988IHlXm3btlVoaKhiYmIUGRmpl19+WWlpaZKk+fPn6/3339eYMWMUExOjsWPHatiwYZo7d67l/ps3b1ZQUJBVTdeuXVOnTp20efNmbd++XSVLllTjxo117dq1DD9XRpw5c0ZLly5VnTp1rNqDgoKUnJysHTt23PfYTlktDgAAAAAeR+Hh4Ro5cqRV2/DhwzVixAirtmvXrmny5Mn69NNP1alTJ0lSiRIl9Oyzz2rGjBlKSEjQV199JTc3N0nSp59+qmbNmmn8+PEqUKCAJMnNzU0zZ86Us7OzJGnevHlKTU3VzJkzLTPWc+bMkZeXlyIjI++6+Zi7u7u8vLwkST4+Phl+3kOHDunbb7/VTz/9pBdeeEGS5OfnZzn/2WefydfXV59++qlMJpPKlCmjM2fOaPDgwXr//fd19uxZJScn6+WXX1axYsUkSRUqVLD67iZOnKiXX35ZkvTUU0/pwIED+uKLL9SpUydduXJFcXFxKlSokFVdzz//vNXnL7/8Ul5eXtqwYYOaNm2a4ee7k7Zt2+r777/XzZs31axZM82cOdPqfM6cOeXp6amTJ0/e9z2Y8QYAAAAAG8LCwhQXF2d1hIWFpesXExOjxMRE1a9f3+a5SpUqWUK3JNWqVUupqak6ePCgpa1ChQqW0C1Je/fu1ZEjR5QrVy65u7vL3d1d3t7eSkhIsLy/nd2io6Pl6OiYbsb3389So0YNyz8E3H6W69ev69SpU6pUqZLq16+vChUqqFWrVpoxY4YuX74sSYqPj9fRo0fVpUsXy/O4u7vrgw8+sDzP7aXcLi4uVvf966+/1LVrV5UsWVKenp7y8PDQ9evXFRsbm+Fna9SokeWe5cqVszo3adIk7d69W99//72OHj2q/v37p7ve1dVVN27cyPD9/osZbwAAAACwwWw2Z2hTMldX1yzf69/BXJKuX7+uqlWrav78+en65suX777u4eDgYFn2fdutW7csf87qczg6Ouqnn37S1q1btXbtWk2dOlXvvfeeduzYoZw5c0qSZsyYoWeeeSbddZKUJ08emUwmS1i/rVOnTrp48aImT56sYsWKyWw2q0aNGpnahG7mzJmWYJ8jRw6rcz4+PvLx8VGZMmXk7e2t2rVra9iwYSpYsKClz6VLl+77e5eY8QYAAACALClZsqRcXV21bt26dOcCAgK0d+9eqw3GtmzZIgcHB8smarZUqVJFhw8fVv78+eXv7291eHp63led+fLls3rfOyUlRfv377d8rlChglJTU7Vhwwab1wcEBGjbtm1W4X3Lli3KlSuXihQpIkkymUyqVauWRo4cqT179sjZ2VnLli1TgQIFVKhQIR07dizd8zz11FOS/tlBvGzZsjpw4IDVfbds2aLevXurcePGKleunMxmsy5cuJCpZy9cuLDlfreXwduSmpoqSVbv8h89elQJCQkKDAzM1D3/jeANAAAAAFng4uKiwYMHa9CgQfrqq6909OhRbd++XbNmzVK7du3k4uKiTp06af/+/frll1/Uq1cvdejQwfJ+ty3t2rVT3rx51bx5c23atEnHjx9XZGSkevfurVOnTt1Xnc8//7xWrVqlVatW6Y8//tBbb71l9XvXxYsXV6dOnRQaGqrly5db7vntt99Kknr06KE///xTvXr10h9//KHvv/9ew4cPV//+/eXg4KAdO3Zo7NixioqKUmxsrJYuXaq///5bAQEBkqSRI0cqPDxcU6ZM0aFDh7Rv3z7NmTNHH3/8saWG4OBgbd682arukiVL6uuvv1ZMTIx27Nihdu3apZudP3funKKjo3XkyBFJ/+wkHx0drUuXLt3x+1i9erXmzJmj/fv368SJE1q1apW6d++uWrVqqXjx4pZ+mzZtkp+fn0qUKHFf37tE8AYAAACALBs2bJjeeecdvf/++woICNBrr72m8+fPK2fOnPrxxx916dIlPf3003r11VdVv359ffrpp3cdL2fOnNq4caOKFi2ql19+WQEBAerSpYsSEhLk4eFxXzWGhoaqU6dO6tixo+rUqSM/Pz/Vq1fPqs/06dP16quvqkePHipTpoy6du1qma0vXLiwVq9erV9//VWVKlVS9+7d1aVLFw0dOlSS5OHhoY0bN6px48YqVaqUhg4dqokTJ6pRo0aSpDfeeEMzZ87UnDlzVKFCBdWpU0cRERGWGW9J6tKli1avXq24uDhL26xZs3T58mVVqVJFHTp0UO/evZU/f36ruj///HMFBgaqa9eukqTnnntOgYGBWrFixR2/D1dXV82YMUPPPvusAgIC1K9fP7300ktauXKlVb8FCxZYxr1fprT/LvK/T2FhYUpISMiOofCYOn/+vM13VB60WxeO2bsE4LHmVvg5e5cAPPbyu3nZuwTgiXDq0v57d0K2a9WqlapUqWJzI7sH7ffff9fzzz+vQ4cO3fcSfykbN1dLSEjQpEmTsms4PIb69etn7xIAAAAAPOQ++ugj/e9//7N3GZL++W3yr776KkuhW2JXcwAAAAB45Li7u9/x3A8//KDatWs/wGqyV/HixdWrVy97lyFJlt8zzyqCNwAAAAA8YqKjo+94rnDhwg+uEGQIwRsAAAAAHjH+/v72LgGZwK7mAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABjIyd4FAA9a8tal9i4BeKzldnG3dwnAY+9ywnV7lwAAyARmvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAA4AEYM2aMatasqZw5c8rLyytD10RGRqp58+YqWLCg3NzcVLlyZc2fP9/YQpHtCN4AAAAA8AAkJSWpVatWeuuttzJ8zdatW1WxYkUtWbJEv/32mzp37qyOHTtq5cqVBlaK7MbPiQEAAABANlizZo0++OAD7d+/X46OjqpRo4YmT56sEiVKSJJGjhwpSYqIiMjwmO+++67V5z59+mjt2rVaunSpmjZtmm21w1jMeAMAAABANoiPj1f//v0VFRWldevWycHBQS1btlRqamq23icuLk7e3t7ZOiaMxYw3AAAAANiQmJioxMREqzaz2Syz2Wyz/yuvvGL1efbs2cqXL58OHDig8uXLZ0tN3377rXbu3KkvvvgiW8bDg8GMNwAAAADYEB4eLk9PT6sjPDz8jv0PHz6stm3bys/PTx4eHipevLgkKTY2NkP3K1eunNzd3eXu7q5GjRqlO//LL7+oc+fOmjFjhsqVK3dfzwT7YMYbAAAAAGwICwtT//79rdruNNstSc2aNVOxYsU0Y8YMFSpUSKmpqSpfvrySkpIydL/Vq1fr1q1bkiRXV1ercxs2bFCzZs00adIkdezYMZNPAnsjeAMAAACADXdbVv5fFy9e1MGDBzVjxgzVrl1bkrR58+ZM3a9YsWI22yMjI9W0aVONHz9e3bp1y9SYeDgQvAEAAAAgi3Lnzq08efLoyy+/VMGCBRUbG6shQ4ZY9YmNjdWlS5cUGxurlJQURUdHS5L8/f3l7u5uc9xffvlFTZs2VZ8+ffTKK6/o3LlzkiRnZ2c2WHuE8I43AAAAAGSRg4ODFi5cqF27dql8+fLq16+fPvroI6s+77//vgIDAzV8+HBdv35dgYGBCgwMVFRU1B3HnTt3rm7cuKHw8HAVLFjQcrz88stGPxKyETPeAAAAAJANXnjhBR04cMCqLS0tzfLniIiITP2G9/1eg4cPM94AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIh3vPHQCQsLU0JCwl37TJo06QFVAwAAAABZQ/DGQychIYFgDQAAAOCxwVJzAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAzkZO8CgAfNlK+wvUsAHmtJKcn2LgF47Dk78P/CAcCjhBlvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAACAR1SHDh00duxYy+fixYvrk08+sUstn3/+uZo1a3bf10dERMjLyyv7CnqIELzxwLi4uKhfv373PM6fP2/vUgEAAIBMCQkJkclkUvfu3dOde/vtt2UymRQSEpKt99y7d69Wr16t3r17Z+q6L7/8UnXr1pWHh4dMJpOuXLmSqesvXryoIkWKpLs2NDRUu3fv1qZNm9Jd88svv6hp06bKly+fXFxcVKJECb322mvauHFjpu79b3Xr1pXJZNLChQut2j/55BMVL148Xf+bN2/K29tbefPmVWJi4n3f936wJSYemPDw8Az169evn8GVAAAAANnP19dXCxcu1KRJk+Tq6ipJSkhI0DfffKOiRYtm+/2mTp2qVq1ayd3dPVPX3bhxQw0bNlTDhg0VFhaW6ft26dJFFStW1OnTp63anZ2d9frrr2vKlCmqXbu2pf2zzz5Tz5491aFDBy1atEglSpRQXFycfvnlF/Xr10+7du3KdA23ubi4aOjQoXrllVeUI0eOu/ZdsmSJypUrp7S0NC1fvlyvvfbafd83s5jxBgAAAIBsUKVKFfn6+mrp0qWWtqVLl6po0aIKDAy0tK1Zs0bPPvusvLy8lCdPHjVt2lRHjx61nD9x4oRMJpOWLl2qevXqKWfOnKpUqZK2bdtm6ZOSkqLFixfbXNp97do1tW3bVm5ubipcuLCmTZtmdb5v374aMmSIqlevnulnnD59uq5cuaIBAwbYPN+sWTOtWLFCN2/elCTFxsaqb9++6tu3r+bOnavnn39exYoVU8WKFdWnTx9FRUXd9X7/+9//9PTTT8vFxUV58+ZVy5Ytrc63bdtWV65c0YwZM+5Z+6xZs9S+fXu1b99es2bNSnf+999/V9OmTeXh4aFcuXKpdu3aVn8vWUHwBgAAAAAbEhMTdfXqVavjXkuUQ0NDNWfOHMvn2bNnq3PnzlZ94uPj1b9/f0VFRWndunVycHBQy5YtlZqaatXvvffe04ABAxQdHa1SpUqpbdu2Sk5OliT99ttviouLU1BQULoaPvroI1WqVEl79uzRkCFD1KdPH/3000/3+zVYHDhwQKNGjdJXX30lBwfbUTIoKEjJycnasWOHpH9mmW/duqVBgwbZ7G8yme54v1WrVqlly5Zq3Lix9uzZo3Xr1qlatWpWfTw8PPTee+9p1KhRio+Pv+NYR48e1bZt29S6dWu1bt1amzZt0smTJy3nT58+reeee05ms1nr16/Xrl27FBoaavm+s4rgDQAAAAA2hIeHy9PT0+q41+uT7du31+bNm3Xy5EmdPHlSW7ZsUfv27a36vPLKK3r55Zfl7++vypUra/bs2dq3b58OHDhg1W/AgAFq0qSJSpUqpZEjR+rkyZM6cuSIJOnkyZNydHRU/vz509VQq1YtDRkyRKVKlVKvXr306quvatKkSVn6LhITE9W2bVt99NFHd102nzNnTnl6elpC7aFDh+Th4SEfHx9LnyVLlsjd3d1y7Nu3z+ZYY8aMUZs2bTRy5EgFBASoUqVKNpfG9+jRQy4uLvr444/vWNfs2bPVqFEj5c6dW97e3goODrb6B5Jp06bJ09NTCxcuVFBQkEqVKqXOnTurdOnS9/xuMoLgDQAAAAA2hIWFKS4uzuq41zvR+fLlU5MmTRQREaE5c+aoSZMmyps3r1Wfw4cPq23btvLz85OHh4dlI7DY2FirfhUrVrT8uWDBgpJk2Yj45s2bMpvNNmeMa9Soke5zTExMxh5aUqNGjSyhuFy5cpL++S4CAgLS/SOCLa6urrpx44bl839rDA4OVnR0tFatWqX4+HilpKTYHCc6Olr169e/5/3MZrNGjRqlCRMm6MKFC+nOp6SkaO7cuVa1t2/fXhEREZZVBtHR0apdu/Y93xO/X2yuBgAAAAA2mM1mmc3mTF8XGhqqnj17SlK696ulf96DLlasmGbMmKFChQopNTVV5cuXV1JSklW/f4fA2+H1dlDMmzevbty4oaSkJDk7O2e6xruZOXOm5R3t2zWsX79e+/bt0+LFiyVJaWlpljree+89jRw50nL9pUuXlC9fPklSyZIlFRcXp3Pnzllmvd3d3eXv7y8np7vH0dsb1GVE+/btNWHCBH3wwQfpdjT/8ccfdfr06XSbqaWkpGjdunV68cUXM3Wv+8GMNwAAAABko4YNGyopKUm3bt1ScHCw1bmLFy/q4MGDGjp0qOrXr6+AgABdvnw50/eoXLmyJKVbni5J27dvT/c5ICAgw2MXLlxY/v7+8vf3V7FixST9szx87969io6OVnR0tGbOnClJ2rRpk95++23LtUePHlVCQoJlM7lXX31VOXLk0Pjx4zP1fNI/M/7r1q3LUF8HBweFh4dr+vTpOnHihNW5WbNmqU2bNpbabx9t2rSxbLJWsWJFbdq0Sbdu3cp0nRnBjDcAAAAAZCNHR0fL0m5HR0erc7lz51aePHn05ZdfqmDBgoqNjdWQIUMyfY98+fKpSpUq2rx5syWE37ZlyxZ9+OGHatGihX766Sd99913WrVqleX8uXPndO7cOcv74vv27VOuXLlUtGhReXt727xfiRIlrD7fXtIdEBAgLy8vS/umTZvk5+dn6V+0aFFNnDhRffr00aVLlxQSEqKnnnpKly5d0rx582x+R7cNHz5c9evXV4kSJdSmTRslJydr9erVGjx4sM3+TZo00TPPPKMvvvhCBQoUkCT9/fff+t///qcVK1aofPnyVv07duyoli1b6tKlS+rZs6emTp2qNm3aKCwsTJ6entq+fbuqVauWLe95M+MNAAAAANnMw8NDHh4e6dodHBy0cOFC7dq1S+XLl1e/fv300Ucf3dc93njjDc2fPz9d+zvvvKOoqCgFBgbqgw8+0Mcff2w18/75558rMDBQXbt2lSQ999xzCgwM1IoVK+6rjn9bsGCBZdzbevXqpbVr1+rvv//Wq6++qpIlS6px48Y6fvy41qxZowoVKtgcq27duvruu++0YsUKVa5cWc8//7x+/fXXu95//PjxSkhIsHz+6quv5ObmZvNd8fr168vV1VXz5s1Tnjx5tH79el2/fl116tRR1apVNWPGjGx759uUdntxfhb169cvyzvlAZLx/y0lbFtg2NgApIIvDrN3CQAAZIvL14/Yu4S7unnzpkqXLq1Fixal21DNHn7//Xc9//zzOnTokDw9Pe1dzkOFGW8AAAAAeAS5urrqq6++srmTtz2cPXtWX331FaHbBt7xBgAAAIBHVN26de1dgsULL7xg7xIeWsx4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAgJ3sXAPyXi4uL+vXrd9c+kyZNekDVAAAAAEDWELzx0AkPD7d3CQAAAACQbVhqDgAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAADwiOrQoYPGjh1r7zIkSW3atNHEiRPtXcZDieANAAAAAFkUEhIik8mk7t27pzv39ttvy2QyKSQkJFvvuXfvXq1evVq9e/fO1HW9e/dW1apVZTabVbly5Qxds3nzZtWqVUt58uSRq6urypQpo0mTJln1GTp0qMaMGaO4uDhLW2RkpEwmk3Lnzq2EhASr/jt37pTJZJLJZMpU/Y8igjcAAAAAZANfX18tXLhQN2/etLQlJCTom2++UdGiRbP9flOnTlWrVq3k7u6e6WtDQ0P12muvZbi/m5ubevbsqY0bNyomJkZDhw7V0KFD9eWXX1r6lC9fXiVKlNC8efPSXZ8rVy4tW7bMqm3WrFmGfC8PI4I3AAAAAGSDKlWqyNfXV0uXLrW0LV26VEWLFlVgYKClbc2aNXr22Wfl5eWlPHnyqGnTpjp69Kjl/IkTJ2QymbR06VLVq1dPOXPmVKVKlbRt2zZLn5SUFC1evFjNmjWzquHrr79WUFCQcuXKJR8fH73++us6f/68VZ8pU6bo7bfflp+fX4afLTAwUG3btlW5cuVUvHhxtW/fXsHBwdq0aZNVv2bNmmnhwoXpru/UqZNmz55t+Xzz5k0tXLhQnTp1sup38eJFtW3bVoULF1bOnDlVoUIFLViwwKpP3bp11bt3bw0aNEje3t7y8fHRiBEjMvws9kDwBgAAAAAbEhMTdfXqVasjMTHxrteEhoZqzpw5ls+zZ89W586drfrEx8erf//+ioqK0rp16+Tg4KCWLVsqNTXVqt97772nAQMGKDo6WqVKlVLbtm2VnJwsSfrtt98UFxenoKAgq2tu3bql0aNHa+/evVq+fLlOnDiR7UvcJWnPnj3aunWr6tSpY9VerVo1/frrr+m+pw4dOmjTpk2KjY2VJC1ZskTFixdXlSpVrPolJCSoatWqWrVqlfbv369u3bqpQ4cO+vXXX636zZ07V25ubtqxY4c+/PBDjRo1Sj/99FO2P2d2IXgDAAAAgA3h4eHy9PS0OsLDw+96Tfv27bV582adPHlSJ0+e1JYtW9S+fXurPq+88opefvll+fv7q3Llypo9e7b27dunAwcOWPUbMGCAmjRpolKlSmnkyJE6efKkjhw5Ikk6efKkHB0dlT9/fqtrQkND1ahRI/n5+al69eqaMmWKfvjhB12/fj0bvhGpSJEiMpvNCgoK0ttvv6033njD6nyhQoWUlJSkc+fOWbXnz59fjRo1UkREhKR//kEiNDQ03fiFCxfWgAEDVLlyZfn5+alXr15q2LChvv32W6t+FStW1PDhw1WyZEl17NhRQUFBWrduXbY8oxEI3gAAAABgQ1hYmOLi4qyOsLCwu16TL18+NWnSRBEREZozZ46aNGmivHnzWvU5fPiw2rZtKz8/P3l4eKh48eKSZJkNvq1ixYqWPxcsWFCSLMvGb968KbPZnG5jsl27dqlZs2YqWrSocuXKZZmR/u/Yd+Pu7m45/rtZ3KZNmxQVFaXPP/9cn3zySbpl4K6urpKkGzdupBs3NDRUEREROnbsmLZt26Z27dql65OSkqLRo0erQoUK8vb2lru7u3788ce7fjfSP9/Pf5fUP0yc7F0AAAAAADyMzGazzGZzpq8LDQ1Vz549JUnTpk1Ld75Zs2YqVqyYZsyYoUKFCik1NVXly5dXUlKSVb8cOXJY/nw7YN9ejp43b17duHFDSUlJcnZ2lvTPEvbg4GAFBwdr/vz5ypcvn2JjYxUcHJxu7LuJjo62/NnDw8Pq3FNPPSVJqlChgv766y+NGDFCbdu2tZy/dOmSpH/+AeK/GjVqpG7duqlLly5q1qyZ8uTJk67PRx99pMmTJ+uTTz5RhQoV5Obmpr59+971u5H++X7+u1T/YULwBgAAAIBs1LBhQyUlJclkMik4ONjq3MWLF3Xw4EHNmDFDtWvXlvTPT3Vl1u2fATtw4IDlz3/88YcuXryocePGydfXV5IUFRWV6bH9/f0z1C81NTXdu9z79+9XkSJF0s3yS5KTk5M6duyoDz/8UD/88IPNMbds2aLmzZtbluenpqbq0KFDKlu2bCaf4uFC8AYAAACAbOTo6KiYmBjLn/8td+7cypMnj7788ksVLFhQsbGxGjJkSKbvkS9fPlWpUkWbN2+2BO+iRYvK2dlZU6dOVffu3bV//36NHj063bVHjhzR9evXde7cOd28edMyw122bFnL7Pl/TZs2TUWLFlWZMmUkSRs3btSECRPS/Yb4pk2b1KBBgzvWPXr0aA0cONDmbLcklSxZUosXL9bWrVuVO3duffzxx/rrr78I3gAAAAAAa/9don2bg4ODFi5cqN69e6t8+fIqXbq0pkyZorp162b6Hm+88Ya++uory7L2fPnyKSIiQu+++66mTJmiKlWqaMKECXrppZfSXbdhwwbL59s/dXb8+HHL++b/lZqaqrCwMB0/flxOTk4qUaKExo8frzfffNPSJyEhQcuXL9eaNWvuWLOzs7PN2fDbhg4dqmPHjik4OFg5c+ZUt27d1KJFC8XFxd3z+3iYmdLS0tKyY6B+/fpp0qRJ2TEUYKiEbQvu3QnAfSv44jB7lwAAQLa4fP2IvUu4q5s3b6p06dJatGiRatSoYe9yNH36dC1btkxr1661dykPHXY1BwAAAIBHkKurq7766itduHDB3qVI+mfDs6lTp9q7jIcSS80BAAAA4BF1P0vUjfLf3/TG/2HGGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAIIvq1q2rvn372rsMPKQI3gAAAABgsKVLl6pBgwbKkyePTCaToqOjM3V9YmKiKleufF/Xwv4I3gAAAABgsPj4eD377LMaP378fV0/aNAgFSpUKJurwoNC8AYAAACAbJCcnKyePXvK09NTefPm1bBhw5SWliZJ6tChg95//3298MILmR73hx9+0Nq1azVhwoTsLhkPCMEbAAAAALLB3Llz5eTkpF9//VWTJ0/Wxx9/rJkzZ2ZpzL/++ktdu3bV119/rZw5c2ZTpXjQnOxdAAAAAAA8jBITE5WYmGjVZjabZTabbfb39fXVpEmTZDKZVLp0ae3bt0+TJk1S165d7+v+aWlpCgkJUffu3RUUFKQTJ07c1ziwP2a8AQAAAMCG8PBweXp6Wh3h4eF37F+9enWZTCbL5xo1aujw4cNKSUm55726d+8ud3d3yyFJU6dO1bVr1xQWFpb1h4FdMeMNAAAAADaEhYWpf//+Vm13mu3OqlGjRmnAgAFWbevXr9e2bdvS3TMoKEjt2rXT3LlzDakF2Y/gDQAAAAA23G1ZuS07duyw+rx9+3aVLFlSjo6O97w2f/78yp8/v1XblClT9MEHH1g+nzlzRsHBwVq0aJGeeeaZDNcF+yN4AwAAAEA2iI2NVf/+/fXmm29q9+7dmjp1qiZOnChJunTpkmJjY3XmzBlJ0sGDByVJPj4+8vHxsTle0aJFrT7fXoJeokQJFSlSxKjHgAEI3gAAAACQDTp27KibN2+qWrVqcnR0VJ8+fdStWzdJ0ooVK9S5c2dL3zZt2kiShg8frhEjRtijXDxAprTbPyyXRf369dOkSZOyYyjAUAnbFti7BOCxVvDFYfYuAQCAbHH5+hF7l4DHBLuaAwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAADwAY8aMUc2aNZUzZ055eXll6JqDBw+qXr16KlCggFxcXOTn56ehQ4fq1q1bxhaLbOVk7wIAAAAA4EmQlJSkVq1aqUaNGpo1a1aGrsmRI4c6duyoKlWqyMvLS3v37lXXrl2VmpqqsWPHGlwxsgvBGwAAAACywZo1a/TBBx9o//79cnR0VI0aNTR58mSVKFFCkjRy5EhJUkRERIbH9PPzk5+fn+VzsWLFFBkZqU2bNmVr7TAWS80BAAAAIBvEx8erf//+ioqK0rp16+Tg4KCWLVsqNTU12+5x5MgRrVmzRnXq1Mm2MWE8ZrwBAAAAwIbExEQlJiZatZnNZpnNZpv9X3nlFavPs2fPVr58+XTgwAGVL18+S7XUrFlTu3fvVmJiorp166ZRo0ZlaTw8WMx4AwAAAIAN4eHh8vT0tDrCw8Pv2P/w4cNq27at/Pz85OHhoeLFi0uSYmNjM3S/cuXKyd3dXe7u7mrUqJHVuUWLFmn37t365ptvtGrVKk2YMOG+nwsPHjPeAAAAAGBDWFiY+vfvb9V2p9luSWrWrJmKFSumGTNmqFChQkpNTVX58uWVlJSUofutXr3aslu5q6ur1TlfX19JUtmyZZWSkqJu3brpnXfekaOjY2YeCXZC8AYAAAAAG+62rPy/Ll68qIMHD2rGjBmqXbu2JGnz5s2Zul+xYsUy1C81NVW3bt1SamoqwfsRQfAGAAAAgCzKnTu38uTJoy+//FIFCxZUbGyshgwZYtUnNjZWly5dUmxsrFJSUhQdHS1J8vf3l7u7u81x58+frxw5cqhChQoym82KiopSWFiYXnvtNeXIkcPox0I2IXgDAAAAQBY5ODho4cKF6t27t8qXL6/SpUtrypQpqlu3rqXP+++/r7lz51o+BwYGSpJ++eUXq37/5uTkpPHjx+vQoUNKS0tTsWLF1LNnT/Xr18/Ix0E2M6WlpaVlx0D9+vXTpEmTsmMowFAJ2xbYuwTgsVbwxWH2LgEAgGxx+foRe5eAxwS7mgMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAADAA2IymbR8+XJ7l4EHjOANAAAAANng3Llz6tOnj/z9/eXi4qICBQqoVq1amj59um7cuJHp8SIiImQymdSwYUOr9itXrshkMikyMjLdNW+++aYcHR313Xff3e9jwAAEbwAAAADIomPHjikwMFBr167V2LFjtWfPHm3btk2DBg3SypUr9fPPP9/XuE5OTvr555/1yy+/3LPvjRs3tHDhQg0aNEizZ8++r/vBGARvAAAAAMiiHj16yMnJSVFRUWrdurUCAgLk5+en5s2ba9WqVWrWrJnN606dOqW2bdvK29tbbm5uCgoK0o4dOyzn3dzcFBoaqiFDhtyzhu+++05ly5bVkCFDtHHjRv35559W5xMTEzV48GD5+vrKbDbL399fs2bNytqDI0Oc7F0AAAAAADyMEhMTlZiYaNVmNptlNput2i5evGiZ6XZzc7M5lslkStd2/fp11alTR4ULF9aKFSvk4+Oj3bt3KzU11arfiBEj5O/vr8WLF+vVV1+9Y72zZs1S+/bt5enpqUaNGikiIkLDhg2znO/YsaO2bdumKVOmqFKlSjp+/LguXLhwz+8BWceMNwAAAADYEB4eLk9PT6sjPDw8Xb8jR44oLS1NpUuXtmrPmzev3N3d5e7ursGDB6e77ptvvtHff/+t5cuX69lnn5W/v79at26tGjVqWPUrVKiQ+vTpo/fee0/Jyck2az18+LC2b9+u1157TZLUvn17zZkzR2lpaZKkQ4cO6dtvv9Xs2bPVsmVL+fn5qX79+pb+MBbBGwAAAABsCAsLU1xcnNURFhaW4et//fVXRUdHq1y5culmziUpOjpagYGB8vb2vudYgwcP1t9//33Hd7dnz56t4OBg5c2bV5LUuHFjxcXFaf369ZZ7OTo6qk6dOhmuH9mHpeYAAAAAYIOtZeW2+Pv7y2Qy6eDBg1btfn5+kiRXV1eb192p3RYvLy+FhYVp5MiRatq0qdW5lJQUzZ07V+fOnZOTk5NV++zZs1W/fv1M3QvZjxlvAAAAAMiCPHny6MUXX9Snn36q+Pj4DF9XsWJFRUdH69KlSxnq36tXLzk4OGjy5MlW7atXr9a1a9e0Z88eRUdHW44FCxZo6dKlunLliipUqKDU1FRt2LAhU8+G7EHwBgAAAIAs+uyzz5ScnKygoCAtWrRIMTExOnjwoObNm6c//vhDjo6O6a5p27atfHx81KJFC23ZskXHjh3TkiVLtG3bNpv3cHFx0ciRIzVlyhSr9lmzZqlJkyaqVKmSypcvbzlat24tLy8vzZ8/X8WLF1enTp0UGhqq5cuX6/jx44qMjNS3335ryPcBawRvAAAAAMiiEiVKaM+ePXrhhRcUFhamSpUqKSgoSFOnTtWAAQM0evTodNc4Oztr7dq1yp8/vxo3bqwKFSpo3LhxNkP6bZ06dbIsYZekv/76S6tWrdIrr7ySrq+Dg4Natmxp+cmw6dOn69VXX1WPHj1UpkwZde3aNVMz9Lh/prTb29xlUb9+/TRp0qTsGAowVMK2BfYuAXisFXxx2L07AQDwCLh8/Yi9S8BjghlvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAHjARowYocqVK9u7jIfCiRMnZDKZFB0dbe9SDEPwBgAAAIAsMJlMdz1GjBiR7poBAwZo3bp1D75Y2IWTvQsAAAAAgEfZ2bNnLX9etGiR3n//fR08eNDS5u7ubvlzWlqaUlJS5O7ubtUOY926dUs5cuSw2/2Z8QYAAACALPDx8bEcnp6eMplMls9//PGHcuXKpR9++EFVq1aV2WzW5s2b0y01DwkJUYsWLTR27FgVKFBAXl5eGjVqlJKTkzVw4EB5e3urSJEimjNnjtW9Bw8erFKlSilnzpzy8/PTsGHDdOvWLcv52/f5+uuvVbx4cXl6eqpNmza6du2apU/dunXVu3dvDRo0SN7e3vLx8Uk3Sx8bG6vmzZvL3d1dHh4eat26tf766y/L+b1796pevXrKlSuXPDw8VLVqVUVFRSk+Pl4eHh5avHix1XjLly+Xm5ubVR3/9vvvv6tp06by8PBQrly5VLt2bR09elSSlJqaqlGjRqlIkSIym82qXLmy1qxZY7n29tL1RYsWqU6dOnJxcdH8+fMlSTNnzlRAQIBcXFxUpkwZffbZZxn4G846gjcAAAAA2JCYmKirV69aHYmJifc11pAhQzRu3DjFxMSoYsWKNvusX79eZ86c0caNG/Xxxx9r+PDhatq0qXLnzq0dO3aoe/fuevPNN3Xq1CnLNbly5VJERIQOHDigyZMna8aMGZo0aZLVuEePHtXy5cu1cuVKrVy5Uhs2bNC4ceOs+sydO1dubm7asWOHPvzwQ40aNUo//fSTpH+CbvPmzXXp0iVt2LBBP/30k44dO6bXXnvNcn27du1UpEgR7dy5U7t27dKQIUOUI0cOubm5qU2bNun+wWDOnDl69dVXlStXrnTfw+nTp/Xcc8/JbDZr/fr12rVrl0JDQ5WcnCxJmjx5siZOnKgJEybot99+U3BwsF566SUdPnw43Xfep08fxcTEKDg4WPPnz9f777+vMWPGKCYmRmPHjtWwYcM0d+7ce/31ZRlLzQEAAADAhvDwcI0cOdKqbfjw4Tbf2b6XUaNG6cUXX7xrH29vb02ZMkUODg4qXbq0PvzwQ924cUPvvvuuJCksLEzjxo3T5s2b1aZNG0nS0KFDLdcXL15cAwYM0MKFCzVo0CBLe2pqqiIiIiwht0OHDlq3bp3GjBlj6VOxYkUNHz5cklSyZEl9+umnWrdunV588UWtW7dO+/bt0/Hjx+Xr6ytJ+uqrr1SuXDnt3LlTTz/9tGJjYzVw4ECVKVPGMsZtb7zxhmrWrKmzZ8+qYMGCOn/+vFavXq2ff/7Z5vcwbdo0eXp6auHChZbl4aVKlbKcnzBhggYPHmz5DsaPH69ffvlFn3zyiaZNm2bp17dvX7388suWz8OHD9fEiRMtbU899ZQOHDigL774Qp06dbrr301WMeMNAAAAADaEhYUpLi7O6ggLC7uvsYKCgu7Zp1y5cnJw+L+IVqBAAVWoUMHy2dHRUXny5NH58+ctbYsWLVKtWrXk4+Mjd3d3DR06VLGxsVbjFi9e3Gpm+Xb4/bf/zsL/u09MTIx8fX0toVuSypYtKy8vL8XExEiS+vfvrzfeeEMvvPCCxo0bZ1kWLknVqlVTuXLlLDPL8+bNU7FixfTcc8/Z/B6io6NVu3Ztm+9kX716VWfOnFGtWrWs2mvVqmWp5bZ/f+fx8fE6evSounTpYnm/3t3dXR988IFVrUYheAMAAACADWazWR4eHlaH2Wy+r7Hc3Nzu2ee/QdNkMtlsS01NlSRt27ZN7dq1U+PGjbVy5Urt2bNH7733npKSku457u0xMtPnbkaMGKHff/9dTZo00fr161W2bFktW7bMcv6NN95QRESEpH+WmXfu3Fkmk8nmWK6urhm+7938+zu/fv26JGnGjBmKjo62HPv379f27duz5X53Q/AGAAAAgEfQ1q1bVaxYMb333nsKCgpSyZIldfLkyWy/T0BAgP7880/9+eeflrYDBw7oypUrKlu2rKWtVKlS6tevn9auXauXX37Z6r3u9u3b6+TJk5oyZYoOHDhw16XdFStW1KZNm6w2ibvNw8NDhQoV0pYtW6zat2zZYlXLfxUoUECFChXSsWPH5O/vb3U89dRTGfoesoJ3vAEAAADgEVSyZEnFxsZq4cKFevrpp7Vq1SqrWebs8sILL6hChQpq166dPvnkEyUnJ6tHjx6qU6eOgoKCdPPmTQ0cOFCvvvqqnnrqKZ06dUo7d+7UK6+8Yhkjd+7cevnllzVw4EA1aNBARYoUueP9evbsqalTp6pNmzYKCwuTp6entm/frmrVqql06dIaOHCghg8frhIlSqhy5cqaM2eOoqOjLTuX38nIkSPVu3dveXp6qmHDhkpMTFRUVJQuX76s/v37Z9v3ZQsz3gAAAADwCHrppZfUr18/9ezZU5UrV9bWrVs1bNiwbL+PyWTS999/r9y5c+u5557TCy+8ID8/Py1atEjSP++eX7x4UR07dlSpUqXUunVrNWrUKN3GdF26dFFSUpJCQ0Pver88efJo/fr1un79uurUqaOqVatqxowZluXwvXv3Vv/+/fXOO++oQoUKWrNmjVasWGG1oZstb7zxhmbOnKk5c+aoQoUKqlOnjiIiIh7IjLcpLS0tLTsG6tevX7pt64GHUcK2BfYuAXisFXwx+/8PPgAA9nD5+hF7l/BY+frrr9WvXz+dOXNGzs7O9i7ngWKpOQAAAADAMDdu3NDZs2c1btw4vfnmm09c6JZYag4AAAAAMNCHH36oMmXKyMfH575/ju1Rx1JzPHGcnAvbuwTgsXZ9+3R7lwA89lKjfrF3CcATIWc38g2yBzPeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAABiI4A0AAAAAgIEI3gAAAAAAGIjgDQAAAACAgQjeAAAAAAAYiOANAAAAAICBCN4AAAAAkEV169ZV37597V0GHlIEbwAAAAAw2NKlS9WgQQPlyZNHJpNJ0dHRGbrupZdeUtGiReXi4qKCBQuqQ4cOOnPmjLHFItsRvAEAAADAYPHx8Xr22Wc1fvz4TF1Xr149ffvttzp48KCWLFmio0eP6tVXXzWoShjFyd4FAAAAAMDjIDk5WT179tTXX3+tHDly6K233tKoUaNkMpnUoUMHSdKJEycyNWa/fv0sfy5WrJiGDBmiFi1a6NatW8qRI0d2lg8DPfHBOywsTAkJCfYuA5k0adIke5cAAAAAWJk7d666dOmiX3/9VVFRUerWrZuKFi2qrl27Zsv4ly5d0vz581WzZk1C9yPmiQ/eCQkJhDgAAAAA6SQmJioxMdGqzWw2y2w22+zv6+urSZMmyWQyqXTp0tq3b58mTZqU5eA9ePBgffrpp7px44aqV6+ulStXZmk8PHi84w0AAAAANoSHh8vT09PqCA8Pv2P/6tWry2QyWT7XqFFDhw8fVkpKyj3v1b17d7m7u1uOfxs4cKD27NmjtWvXytHRUR07dlRaWtr9PxgeuCd+xhsAAAAAbAkLC1P//v2t2u40251Vo0aN0oABA2yey5s3r/LmzatSpUopICBAvr6+2r59u2rUqGFILch+BG8AAAAAsOFuy8pt2bFjh9Xn7du3q2TJknJ0dLzntfnz51f+/Pnv2S81NVWS0i2Bx8ON4A0AAAAA2SA2Nlb9+/fXm2++qd27d2vq1KmaOHGipH82RouNjbX8BvfBgwclST4+PvLx8bE53o4dO7Rz5049++yzyp07t44ePaphw4apRIkSzHY/YnjHGwAAAACyQceOHXXz5k1Vq1ZNb7/9tvr06aNu3bpJklasWKHAwEA1adJEktSmTRsFBgbq888/v+N4OXPm1NKlS1W/fn2VLl1aXbp0UcWKFbVhwwbDlrzDGMx4AwAAAEAWRUZGWv48ffr0dOdDQkIUEhKSqTErVKig9evXZ7EyPAyY8QYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAAAAwEAEbwAAAAAADETwBgAAAADAQARvAAAAAAAMRPAGAAAAAMBABG8AAAAAAAxE8AYAAACAB8RkMmn58uX2LgMPGMEbAAAAALLBuXPn1KdPH/n7+8vFxUUFChRQrVq1NH36dN24ccPe5cGOnOxdAAAAAAA86o4dO6ZatWrJy8tLY8eOVYUKFWQ2m7Vv3z59+eWXKly4sF566SV7lwk7IXgDAAAAQBb16NFDTk5OioqKkpubm6Xdz89PzZs3V1pams3rTp06pYEDB+rHH39UYmKiAgICNG3aND3zzDMPqnQ8AARvAAAAAMiCixcvau3atRo7dqxV6P43k8mUru369euqU6eOChcurBUrVsjHx0e7d+9Wamqq0SXjAcu24O3i4qJ+/fpl13APzPnz5+1dAgAAAICHUGJiohITE63azGazzGazVduRI0eUlpam0qVLW7XnzZtXCQkJkqS3335b48ePtzr/zTff6O+//9bOnTvl7e0tSfL398/ux8BDINuCd3h4eHYN9UA9iv9YAAAAAMB44eHhGjlypFXb8OHDNWLEiAxd/+uvvyo1NVXt2rVLF+AlKTo6WoGBgZbQjccXS80BAAAAwIawsDD179/fqu2/s93SP7PUJpNJBw8etGr38/OTJLm6utoc/07tePzwc2IAAAAAYIPZbJaHh4fVYSt458mTRy+++KI+/fRTxcfHZ3j8ihUrKjo6WpcuXcrOsvEQIngDAAAAQBZ99tlnSk5OVlBQkBYtWqSYmBgdPHhQ8+bN0x9//CFHR8d017Rt21Y+Pj5q0aKFtmzZomPHjmnJkiXatm2bHZ4ARmKpOQAAAABkUYkSJbRnzx6NHTtWYWFhOnXqlMxms8qWLasBAwaoR48e6a5xdnbW2rVr9c4776hx48ZKTk5W2bJlNW3aNDs8AYxE8AYAAACAbFCwYEFNnTpVU6dOvWOf//6ed7FixbR48WKjS4OdsdQcAAAAAAADEbwBAAAAADAQS83xxGnkE2jvEoDH2sSWi+xdAvDYq56QbO8SgCdC/W72rgCPC2a8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAAAAMRvAEAAAAAMBDBGwAAAAAAAxG8AQAAAAAwEMEbAAAAAAADEbwBAAAAADAQwRsAAAAA7KBu3brq27dvpq/r0KGDxo4dm/0F3Yfq1atryZIl2TLWiBEjVLly5WwZ62FD8AYAAACALAoJCVGLFi2s2hYvXiwXFxdNnDgx2+6zd+9erV69Wr17987UNW3btpWvr69cXV0VEBCgyZMn3/O6iIgImUwmq8PFxcWqz9ChQzVkyBClpqZaXefl5WXVLyYmRr6+vmrVqpWSkpIyXPvjguANAAAAANls5syZateunaZPn6533nkn28adOnWqWrVqJXd39wxfs2vXLuXPn1/z5s3T77//rvfee09hYWH69NNP73mth4eHzp49azlOnjxpdb5Ro0a6du2afvjhhzuOsXPnTtWuXVsNGzbUokWL5OzsnOHaHxcEbwAAAADIRh9++KF69eqlhQsXqnPnzpKk+Ph4dezYUe7u7ipYsKDNWfDLly+rY8eOyp07t3LmzKlGjRrp8OHDlvMpKSlavHixmjVrZnXd119/raCgIOXKlUs+Pj56/fXXdf78ecv50NBQTZ48WXXq1JGfn5/at2+vzp07a+nSpfd8FpPJJB8fH8tRoEABq/OOjo5q3LixFi5caPP69evX6/nnn1eXLl00Y8YMOTj8XwQdN26cChQooFy5cqlLly5KSEiwujY1NVWjRo1SkSJFZDabVblyZa1Zs+aeNT+MCN4AAAAAYENiYqKuXr1qdSQmJt71msGDB2v06NFauXKlWrZsaWkfOHCgNmzYoO+//15r165VZGSkdu/ebXVtSEiIoqKitGLFCm3btk1paWlq3Lixbt26JUn67bffFBcXp6CgIKvrbt26pdGjR2vv3r1avny5Tpw4oZCQkLvWGRcXJ29v73t+B9evX1exYsXk6+ur5s2b6/fff0/Xp1q1atq0aVO69mXLlqlJkyYaOnSoxo8fb3Xu22+/1YgRIzR27FhFRUWpYMGC+uyzz6z6TJ48WRMnTtSECRP022+/KTg4WC+99JLVP0Y8KkxpaWlp9i7CnsLCwtL9ywoefpMmTbrva5sVbZqNlQD4r+omT3uXADz2qick27sE4Imw6a0AjRw50qpt+PDhGjFiRLq+ISEhWrBggZKSkrRu3To9//zzlnPXr19Xnjx5NG/ePLVq1UqSdOnSJRUpUkTdunXTJ598osOHD6tUqVLasmWLatasKUm6ePGifH19NXfuXLVq1UrLly/Xq6++qlu3bslkMt2x7qioKD399NO6du2azSXpW7duVZ06dbRq1So1aNDgjuNs27ZNhw8fVsWKFRUXF6cJEyZo48aN+v3331WkSBFLvxUrVqhly5a6deuWHBwcFBERoTfeeEOS9O6772rUqFHpxq5Zs6YCAwM1bdo0S1v16tWVkJCg6OhoSVLhwoX19ttv691337X0qVatmp5++mmr6x4FTvYuwN7Cw8PtXQIAAACAh1BYWJj69+9v1WY2m+/Yv2LFirpw4YKGDx+uatWqWULv0aNHlZSUpGeeecbS19vbW6VLl7Z8jomJkZOTk1WfPHnyqHTp0oqJiZEk3bx5U2azOV3o3rVrl0aMGKG9e/fq8uXLlo3OYmNjVbZsWau++/fvV/PmzTV8+HBL6P5vv3fffVfvvvuuatSooRo1aljaa9asqYCAAH3xxRcaPXq0pd3V1VWpqalKTEyUq6urpe3ZZ5/VjBkz1LZtWwUEBFjVERMTo+7du1u11ahRQ7/88osk6erVqzpz5oxq1apl1adWrVrau3evHjUsNQcAAAAAG8xmszw8PKyOuwXvwoULKzIyUqdPn1bDhg117dq1bK0nb968unHjhtWu4PHx8QoODpaHh4fmz5+vnTt3atmyZZKUbvfwAwcOqH79+urWrZuGDh1qaS9UqJCio6Mtx38D8W05cuRQYGCgjhw5YtV+6dIlubm5WUK39M+738uXL1eVKlVUr149yz8ePKkI3gAAAACQTYoVK6YNGzbo3LlzlvBdokQJ5ciRQzt27LD0u3z5sg4dOmT5HBAQoOTkZKs+Fy9e1MGDBy2z0bd/4/rAgQOWPn/88YcuXryocePGqXbt2ipTpozVxmq3/f7776pXr546deqkMWPGWJ1zcnKSv7+/5bjTu98pKSnat2+fChYsaNW+f/9+BQYGputvNpu1dOlSPf3006pXr55V3QEBAVbPKknbt2+3/NnDw0OFChXSli1brPps2bIl3Sz+o4DgDQAAAADZyNfXV5GRkTp//ryCg4OVmpqqLl26aODAgVq/fr3279+vkJAQqx2+S5YsqebNm6tr167avHmz9u7dq/bt26tw4cJq3ry5JClfvnyqUqWKNm/ebLmuaNGicnZ21tSpU3Xs2DGtWLHCahm49E8wrlevnho0aKD+/fvr3LlzOnfunP7++++7PseoUaO0du1aHTt2TLt371b79u118uRJy/vbt23atOmO74qbzWYtWbJEzzzzjOrVq2fZnK1Pnz6aPXu25syZo0OHDmn48OHpNm4bOHCgxo8fr0WLFungwYMaMmSIoqOj1adPn3v8DTx8CN4AAAAAkM2KFCmiyMhIXbhwQcHBwRo/frxq166tZs2a6YUXXtCzzz6rqlWrWl0zZ84cVa1aVU2bNlWNGjWUlpam1atXK0eOHJY+b7zxhubPn2/5nC9fPkVEROi7775T2bJlNW7cOE2YMMFq3MWLF+vvv//WvHnzVLBgQcvx9NNP3/UZLl++rK5duyogIECNGzfW1atXtXXrVqsZ59OnT2vr1q2Wn02zxdnZWYsXL1bNmjVVr1497d+/X6+99pqGDRumQYMGqWrVqjp58qTeeustq+t69+6t/v3765133lGFChW0Zs0arVixQiVLlrxr3Q+jJ35Xczx52NUcMBa7mgPGY1dz4MGo/9cie5eQzs2bN1W6dGktWrTIauMzexk8eLAuX76sL7/80t6lPNSe+F3NAQAAAOBR4erqqq+++koXLlywdymSpPz586fb+R3pMeONJw4z3oCxmPEGjMeMN/BgPIwz3ng08Y43AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAAABiJ4AwAAAABgIII3AAAAAAAGIngDAAAAAGAggjcAAAAAAAYieAMAAAAAYCCCNwAAAAA8ICaTScuXL7d3GVly8eJF5c+fXydOnLB3Kbpw4YLy58+vU6dOZct4devWVd++fbNlrH8jeAMAAABANjh37pz69Okjf39/ubi4qECBAqpVq5amT5+uGzduZHq8iIgImUwmNWzY0Kr9ypUrMplMioyMTHfNm2++KUdHR3333Xf3+xgKDg6Wo6Ojdu7cafP8mDFj1Lx5cxUvXjzDY/7+++965ZVXVLx4cZlMJn3yySeZrqt79+7prs2bN686duyo4cOHW/UNCQlRixYtrNoWL14sFxcXTZw4MdP3ziqCNwAAAABk0bFjxxQYGKi1a9dq7Nix2rNnj7Zt26ZBgwZp5cqV+vnnn+9rXCcnJ/3888/65Zdf7tn3xo0bWrhwoQYNGqTZs2ff1/1iY2O1detW9ezZ0+YYN27c0KxZs9SlS5dMjXvjxg35+flp3Lhx8vHxyXRdy5Yt0/bt21WoUKF05zp37qz58+fr0qVLd7x+5syZateunaZPn6533nkn0/fPKoI3AAAAAGRRjx495OTkpKioKLVu3VoBAQHy8/NT8+bNtWrVKjVr1szmdadOnVLbtm3l7e0tNzc3BQUFaceOHZbzbm5uCg0N1ZAhQ+5Zw3fffaeyZctqyJAh2rhxo/7880+r84mJiRo8eLB8fX1lNpvl7++vWbNmWfWZM2eOmjZtqrfeeksLFizQzZs3rc6vXr1aZrNZ1atXt7SlpKSoS5cueuqpp+Tq6qrSpUtr8uTJVtc9/fTT+uijj9SmTRuZzeZ7Psu/nT59Wr169dL8+fOVI0eOdOfLlSunQoUKadmyZTav//DDD9WrVy8tXLhQnTt3trTHx8erY8eOcnd3V8GCBW3OhF++fFkdO3ZU7ty5lTNnTjVq1EiHDx/OVP0SwRsAAAAAbEpMTNTVq1etjsTExHT9Ll68qLVr1+rtt9+Wm5ubzbFMJlO6tuvXr6tOnTo6ffq0VqxYob1792rQoEFKTU216jdixAjt27dPixcvvmu9s2bNUvv27eXp6alGjRopIiLC6nzHjh21YMECTZkyRTExMfriiy/k7u5uOZ+WlqY5c+aoffv2KlOmjPz9/dPdc9OmTapatapVW2pqqooUKaLvvvtOBw4c0Pvvv693331X33777V3rzYjU1FR16NBBAwcOVLly5e7Yr1q1atq0aVO69sGDB2v06NFauXKlWrZsaXVu4MCB2rBhg77//nutXbtWkZGR2r17t1WfkJAQRUVFacWKFdq2bZvS0tLUuHFj3bp1K1PPQfAGAAAAABvCw8Pl6elpdYSHh6frd+TIEaWlpal06dJW7Xnz5pW7u7vc3d01ePDgdNd98803+vvvv7V8+XI9++yz8vf3V+vWrVWjRg2rfoUKFVKfPn303nvvKTk52Wathw8f1vbt2/Xaa69Jktq3b685c+YoLS1NknTo0CF9++23mj17tlq2bCk/Pz/Vr1/f0l+Sfv75Z924cUPBwcGWMf47I37y5Ml0y71z5MihkSNHKigoSE899ZTatWunzp07Z0vwHj9+vJycnNS7d++79itUqJBOnjxp1fbDDz/oww8/1Pfff6/69etbnbt+/bpmzZqlCRMmqH79+qpQoYLmzp1r9f0ePnxYK1as0MyZM1W7dm1VqlRJ8+fP1+nTpzO9QR7BGwAAAABsCAsLU1xcnNURFhaW4et//fVXRUdHq1y5cjZnyqOjoxUYGChvb+97jjV48GD9/fffd3x3e/bs2QoODlbevHklSY0bN1ZcXJzWr19vuZejo6Pq1Klzx3vMnj1br732mpycnCRJbdu21ZYtW3T06FFLn5s3b8rFxSXdtdOmTVPVqlWVL18+ubu768svv1RsbOw9n+u2+fPnW/6Rwt3dXZs2bdKuXbs0efJkyyZzd+Pq6ppuA7uKFSuqePHiGj58uK5fv2517ujRo0pKStIzzzxjafP29rb6x5OYmBg5OTlZ9cmTJ49Kly6tmJiYDD+bRPAGAAAAAJvMZrM8PDysDlvvJ/v7+8tkMungwYNW7X5+fvL395erq6vN8e/UbouXl5fCwsI0cuTIdAEzJSVFc+fO1apVq+Tk5CQnJyflzJlTly5dsgT1e93r0qVLWrZsmT777DPLGIULF1ZycrJV2M+bN68uX75sde3ChQs1YMAAdenSRWvXrlV0dLQ6d+6spKSkDD/fSy+9pOjoaMsRFBSkTZs26fz58ypatKilppMnT+qdd95Jt6P6pUuXlC9fPqu2woULKzIyUqdPn1bDhg117dq1DNeT3QjeAAAAAJAFefLk0YsvvqhPP/1U8fHxGb6uYsWKio6Ovutu3P/Wq1cvOTg4pNu4bPXq1bp27Zr27NljFV4XLFigpUuX6sqVK6pQoYJSU1O1YcMGm2PPnz9fRYoU0d69e63GmDhxoiIiIpSSkiJJCgwM1IEDB6yu3bJli2rWrKkePXooMDBQ/v7+VrPkGZErVy75+/tbDldXV3Xo0EG//fabVT2FChXSwIED9eOPP1pdv3//fgUGBqYbt1ixYtqwYYPOnTtnFb5LlCihHDlyWG1kd/nyZR06dMjyOSAgQMnJyVZ9Ll68qIMHD6ps2bKZej6CNwAAAABk0Weffabk5GQFBQVp0aJFiomJ0cGDBzVv3jz98ccfcnR0THdN27Zt5ePjoxYtWmjLli06duyYlixZom3bttm8h4uLi0aOHKkpU6ZYtc+aNUtNmjRRpUqVVL58ecvRunVreXl5af78+SpevLg6deqk0NBQLV++XMePH1dkZKTlPexZs2bp1Vdftbq+fPny6tKliy5cuKA1a9ZI+uc3vn///XerWe+SJUsqKipKP/74ow4dOqRhw4al+w3wpKQkS3hOSkrS6dOnFR0drSNHjtzxO82TJ0+6enLkyCEfHx+rJeE3btzQrl271KBBA5vj+Pr6KjIyUufPn1dwcLCuXr0qd3d3denSRQMHDtT69eu1f/9+hYSEyMHh/yJyyZIl1bx5c3Xt2lWbN2/W3r171b59exUuXFjNmze/Y922ELwBAAAAIItKlCihPXv26IUXXlBYWJgqVaqkoKAgTZ06VQMGDNDo0aPTXePs7Ky1a9cqf/78avz/2rv3qKrrfP/jr43IhiAwKCTAG2wjU1RGymteRlLziJdKymbGS+rk8ZItjqZ7WjNINlKT04yX8oyOUueYR1Y2qatsQkF09JjmZbv0qCRDHkdUIkwKyk3E/v0xv/acPWwVhQ+b8PlY67tW+/39fD68P7X449Xny3ePHKnExES99NJLXkP69yZNmqS4uDj359LSUr3//vt69NFH64z18/PTuHHj3C9IW7VqlR577DHNnDlT9957r6ZPn66qqiodOnRIR48e9bpGWFiYhg4d6l4jMTFRP/rRjzxenPb000/rkUce0eOPP67evXurvLxcM2fO9Fjn/PnzSkpKUlJSki5cuKClS5cqKSlJ06ZNu86/2evbsmWL2rdvrwcffPCqY2JjY1VQUKDPP//cHb5feeUVPfjgg0pNTVVKSooGDBhQ543t2dnZ6tWrl0aNGqW+ffvK5XJp27ZtXr/W7Fosru9fcwfcIlLbj/J1C0CL1scS5usWgBavzxXvbzUG0LiGlub4uoVm6f3339f8+fN1/PhxjxNiX+nTp4+eeeYZPfnkk75u5ar8fd0AAAAAAOCH41/+5V90+vRplZSUqF27dj7t5fPPP9cjjzyiCRMm+LSP6+HEG7ccTrwBszjxBszjxBtoGpx4o7H4/rkAAAAAAABaMII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAmojFYtHmzZuNrL127VoNGzbM/Xny5MkaO3askZ91PSdOnFBsbKyqqqoaZb2OHTvq97//faOs5QsEbwAAAABoBBcvXtTcuXNls9kUGBiotm3bqn///lq1apW+/vrrG17vjTfekMVi0YgRIzzqly9flsViUUFBgbt25coV/fKXv1RGRsYN/Yzdu3crNTVV0dHR9f6fAmfOnJHFYvF6vf3225Kk++67T3369NGrr77qMXfw4MF69tlnPWrLli2T1WrVxo0bb6j3HxKCNwAAAAA0UHFxsZKSkpSbm6slS5boyJEj2rdvn5577jm999572rFjx02t6+/vrx07dmjnzp3XHLdp0yaFhoaqf//+N7R+VVWVevTooddee63ec9q1a6cLFy54XJmZmQoJCdHDDz/sHjdlyhStWrVKNTU1V10rIyNDv/jFL7RlyxY98cQTN9T7DwnBGwAAAAAaaObMmfL399fBgweVlpamLl26KC4uTmPGjNH777+v1NRUr/POnTunCRMmKDw8XMHBwUpOTtb+/fvd94ODg/XUU09p4cKF1/z5GzduvOrPyMzM1F133aXQ0FDNmDFD1dXV7nsPP/ywXnzxRY0bN67ee23VqpWioqI8rnfffVdpaWkKCQlxj3vooYd06dIl7dq1q84aLpdLc+bM0fLly7V9+3aPU/3PPvtMqampCgoKUqdOnfTWW2/VmX/27FmNGTNGISEhCg0NVVpamkpLS+u9h6bm7+sGAAAAAKA5cjqdcjqdHjWr1Sqr1epRKy8vd590BwcHe13LYrHUqVVWVmrQoEGKiYnR1q1bFRUVpcOHD6u2ttZj3KJFi2Sz2bRp0yY99thjXtffs2ePfvazn9Wp5+XlKTAwUAUFBTpz5oymTJmiiIgI/frXv77m3m/EoUOH5HA46pyaBwQEqGfPnvrLX/6ioUOHuus1NTX66U9/qvz8fO3atUvdu3f3mDd58mSdP39eO3fuVOvWrfXMM8/os88+c9+vra11h+5du3appqZGs2bN0uOPP+7x+H1zQvAGAAAAAC+ysrKUmZnpUcvIyNCiRYs8akVFRXK5XEpISPCo33nnnbpy5YokadasWXr55Zc97m/YsEFlZWX6+OOPFR4eLkmy2Wx1+oiOjtbcuXP1/PPPe31Z2uXLl1VRUaHo6Og69wICArRu3Trddttt6tq1q1544QXNnz9fixcvlp9f4zwAvXbtWnXp0kX9+vXz2vv//u//etTWrFkjSTp69Kjuvfdej3uffPKJPvjgAx04cED333+/x/rfy8vL07Fjx/Tpp5+qXbt2kqT/+I//UNeuXfXxxx+75zUnPGoOAAAAAF7Y7XZVVFR4XHa7vd7zDxw4IIfDoa5du9Y5OZckh8OhpKQkd+i+lgULFqisrEzr1q2rc++bb76RJAUGBta516NHD912223uz3379lVlZaX+9re/1WsPS5YsUUhIiPs6e/ZsnZ+9YcMGTZ061ev8oKCgOi+WGzBggEJCQvTLX/6yzt9/nzx5Uv7+/urVq5e7du+996pNmzYeY9q1a+cO3dLfX+bWpk0bnTx5sl77amoEbwAAAADwwmq1KjQ01OP658fMpb+fUlssFhUWFnrU4+LiZLPZFBQU5HX9q9W9adOmjex2uzIzM+sE2YiICFksFn3xxRf1Xq++ZsyYIYfD4b7++VR906ZN+vrrrzVx4kSv8y9duqS77rrLo5aYmKi8vDzt3LlTjz/++DVfvtZSELwBAAAAoAEiIiL00EMPaeXKlTf0vdXdu3eXw+HQpUuX6jV+zpw58vPz07JlyzzqAQEBuu+++3TixIk6c44ePeo+EZekjz76SCEhIR6nxdcSHh4um83mvvz9Pf9aee3atRo9enSdcP2948ePKykpqU69Z8+eysvL0+7du5WWlqZvv/1W0t9Pt2tqanTo0CH32MLCQl2+fNn9uUuXLvrb3/7mcWp/4sQJXb58Wffdd1+99tXUCN4AAAAA0ECvv/66ampqlJycrJycHJ08eVKFhYVav369Tp06pVatWtWZM2HCBEVFRWns2LHau3eviouL9c4772jfvn1ef0ZgYKAyMzO1fPnyOveGDx+uPXv21KlXV1dr6tSpOnHihLZt26aMjAzNnj3b/ffdlZWV7tNsSfr000/lcDjqPFLuTVFRkXbv3q1p06Z5vX/mzBmVlJQoJSXF6/0ePXooPz9fe/bscYfvhIQEjRgxQk8//bT279+vQ4cOadq0aR5PB6SkpCgxMVE/+clPdPjwYR04cEATJ07UoEGDlJycfN2+fYHgDQAAAAANFB8fryNHjiglJUV2u109evRQcnKyVqxYoXnz5mnx4sV15gQEBCg3N1eRkZEaOXKkEhMT9dJLL3kN6d+bNGmS4uLi6tSnTp2qbdu2qaKiwqM+dOhQde7cWQMHDtTjjz+u0aNHe7wc7uDBg0pKSnKfSqenpyspKUm/+tWvrrvndevWKTY2VsOGDfN6/7/+6780bNgwdejQ4aprJCYmKj8/X//93/+t8ePHq7q6WtnZ2YqOjtagQYP0yCOP6Oc//7kiIyPdcywWi7Zs2aI77rhDAwcOVEpKiuLi4pSTk3Pdnn3F4nK5XL5uAmhKqe1H+boFoEXrYwnzdQtAi9fnSsv/e0igORha2nyDnDfjx4/Xj370oxt6AZwp1dXV6ty5szZs2KD+/fv7uh2f48QbAAAAAFqAV155RSEhIb5uQ5J09uxZ/eIXvyB0/3+ceOOWw4k3YBYn3oB5nHgDTeOHduKN5osTbwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAADTQ4MGD9eyzz/q6DTRTBG8AAAAAMOxPf/qThg0bpoiICFksFjkcjuvOKSgokMVi8Xp9/PHH5ptGoyF4AwAAAIBhVVVVGjBggF5++eV6z+nXr58uXLjgcU2bNk2dOnVScnKywW7R2Px93QAAAAAAtAQ1NTWaPXu2/vM//1OtW7fWv/7rv+qFF16QxWLRz372M0nSmTNn6r1eQECAoqKi3J+//fZbbdmyRXPmzJHFYmns9mEQJ94AAAAA0AjefPNN+fv768CBA1q2bJleffVV/fGPf2y09bdu3ary8nJNmTKl0dZE0+DEGwAAAAC8cDqdcjqdHjWr1Sqr1ep1fLt27fS73/1OFotFCQkJOnbsmH73u99p+vTpjdLP2rVrNXz4cMXGxjbKemg6nHgDAAAAgBdZWVkKCwvzuLKysq46vk+fPh6PgPft21enT5/Wd999d92fNWPGDIWEhLivf3bu3Dl9+OGHmjp16s1tBj7FiTcAAAAAeGG325Wenu5Ru9ppd0O98MILmjdv3lXvZ2dnKyIiQqNHjzby82EWwRsAAAAAvLjWY+Xe7N+/3+PzRx99pM6dO6tVq1bXnRsZGanIyEiv91wul7KzszVx4kS1bt263v2g+SB4AwAAAEAjOHv2rNLT0/X000/r8OHDWrFihX77299Kki5duqSzZ8/q/PnzkqTCwkJJUlRUlMeby73Jz8/Xp59+qmnTppndAIwheAMAAABAI5g4caK++eYbPfDAA2rVqpXmzp2rn//855L+/kby//s28ieeeEKSlJGRoUWLFl1z3bVr16pfv3669957jfUOsywul8vl6yaAppTafpSvWwBatD6WMF+3ALR4fa7U+LoF4JYwtDTH1y2gheCt5gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADDzpw5o6lTp6pTp04KCgpSfHy8MjIyVF1dfd25TqdTzz//vDp06CCr1aqOHTtq3bp1TdA1Gou/rxsAAAAAgJbu1KlTqq2t1R/+8AfZbDYdP35c06dPV1VVlZYuXXrNuWlpaSotLdXatWtls9l04cIF1dbWNlHnaAwWl8vl8nUTQFNKbT/K1y0ALVofS5ivWwBavD5XanzdAnBLGFqac0Pj//znP+vFF1/U8ePH1apVK/Xt21fLli1TfHy81/GvvPKKVq1apeLi4muu+cQTT6i4uFjh4eE31A+aDx41BwAAAAAvnE6nvvzyS4/L6XRedXxVVZXS09N18OBB5eXlyc/PT+PGjbvq6XRFRcV1w/TWrVuVnJys3/zmN4qJidE999yjefPm6ZtvvmnQ3tC0eNQcAAAAALzIyspSZmamRy0jI0OLFi3yOv7RRx/1+Lxu3TrdddddOnHihLp16+Zxr6ioSCtWrLjuY+bFxcXas2ePAgMD9e677+rzzz/XzJkzVV5eruzs7BvfFHyCE28AAAAA8MJut6uiosLjstvtVx1/+vRpTZgwQXFxcQoNDVXHjh0lSWfPnvUYV1JSohEjRmj8+PGaPn26ux4SEuK+ZsyYIUmqra2VxWLRW2+9pQceeEAjR47Uq6++qjfffJNT7x8QTrwBAAAAwAur1Sqr1Vrv8ampqerQoYPWrFmj6Oho1dbWqlu3bh5vLj9//ryGDBmifv36afXq1R7zHQ6H+59DQ0MlSXfffbdiYmIUFvaPd6h06dJFLpdL586dU+fOnW9yd2hKBG8AAAAAaKDy8nIVFhZqzZo1evDBByVJe/bs8RhTUlKiIUOGqFevXsrOzpafn+cDyDabrc66/fv319tvv63KykqFhIRIkj755BP5+fkpNjbW0G7Q2HjUHAAAAAAa6I477lBERIRWr16toqIi5efnKz093X2/pKREgwcPVvv27bV06VKVlZXp4sWLunjx4jXXffLJJxUREaEpU6boxIkT2r17t+bPn6+nnnpKQUFBpreFRsKJNwAAAAA0kJ+fnzZu3KhnnnlG3bp1U0JCgpYvX67BgwdLkrZv366ioiIVFRXVOam+1jc8h4SEaPv27ZozZ46Sk5MVERGhtLQ0vfjiiya3g0bG93jjlsP3eANm8T3egHl8jzfQNG70e7yBq+FRcwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAIBhZ86c0dSpU9WpUycFBQUpPj5eGRkZqq6urvcae/fulb+/v3r27GmuURjh7+sGAAAAAKClO3XqlGpra/WHP/xBNptNx48f1/Tp01VVVaWlS5ded/7ly5c1ceJEDR06VKWlpU3QMRoTJ94AAAAA0Aj+/Oc/a8CAAWrTpo0iIiI0atQo/fWvf5UkjRgxQtnZ2Ro2bJji4uI0evRozZs3T3/605/qtfaMGTP05JNPqm/fvia3AEMI3gAAAADghdPp1JdffulxOZ3Oq46vqqpSenq6Dh48qLy8PPn5+WncuHGqra31Or6iokLh4eHX7SM7O1vFxcXKyMi46b3AtwjeAAAAAOBFVlaWwsLCPK6srKyrjn/00Uf1yCOPyGazqWfPnlq3bp2OHTumEydO1BlbVFSkFStW6Omnn75mD6dPn9bChQu1fv16+fvzl8I/VARvAAAAAPDCbreroqLC47Lb7Vcdf/r0aU2YMEFxcXEKDQ1Vx44dJUlnz571GFdSUqIRI0Zo/Pjxmj59urseEhLivmbMmKHvvvtOTz75pDIzM3XPPfcY2SOaBv/LBAAAAAC8sFqtslqt9R6fmpqqDh06aM2aNYqOjlZtba26devm8eby8+fPa8iQIerXr59Wr17tMd/hcLj/OTQ0VF999ZUOHjyoI0eOaPbs2ZKk2tpauVwu+fv7Kzc3Vz/+8Y8btkk0CYI3AAAAADRQeXm5CgsLtWbNGj344IOSpD179niMKSkp0ZAhQ9SrVy9lZ2fLz8/zAWSbzebxuba2VseOHfOovf7668rPz9emTZvUqVMnAzuBCQRvAAAAAGigO+64QxEREVq9erXuvvtunT17VgsXLnTfLykp0eDBg9WhQwctXbpUZWVl7ntRUVFe1/Tz81O3bt08apGRkQoMDKxTR/NG8AYAAACABvLz89PGjRv1zDPPqFu3bkpISNDy5cs1ePBgSdL27dtVVFSkoqIixcbGesx1uVw+6BhNyeLivzJuMantR/m6BaBF62MJ83ULQIvX50qNr1sAbglDS3N83QJaCN5qDgAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAADABwYPHqxnn33W122gCRC8AQAAAKCBJk+erLFjx3rUNm3apMDAQP32t7/1TVNoNvx93QAAAAAAtDR//OMfNWvWLP37v/+7pkyZ4ut24GOceAMAAABAI/rNb36jOXPmaOPGje7QXVVVpYkTJyokJER3332311PwL774QhMnTtQdd9yh2267TQ8//LBOnz7d1O3DAII3AAAAADSSBQsWaPHixXrvvfc0btw4d33+/PnatWuXtmzZotzcXBUUFOjw4cMecydPnqyDBw9q69at2rdvn1wul0aOHKlvv/22qbeBRsaj5gAAAADghdPplNPp9KhZrVZZrVav4z/44ANt2bJFeXl5+vGPf+yuV1ZWau3atVq/fr2GDh0qSXrzzTcVGxvrHnP69Glt3bpVe/fuVb9+/SRJb731ltq1a6fNmzdr/Pjxjb09NCFOvAEAAADAi6ysLIWFhXlcWVlZVx3fvXt3dezYURkZGaqsrHTX//rXv6q6ulq9e/d218LDw5WQkOD+fPLkSfn7+3uMiYiIUEJCgk6ePNnIO0NTI3gDAAAAgBd2u10VFRUel91uv+r4mJgYFRQUqKSkRCNGjNBXX33VhN2iOSN4AwAAAIAXVqtVoaGhHtfVHjP/XocOHbRr1y5dvHjRHb7j4+PVunVr7d+/3z3uiy++0CeffOL+3KVLF9XU1HiMKS8vV2Fhoe67777G3xyaFMEbAAAAABpRu3btVFBQoM8++0zDhw9XbW2tpk6dqvnz5ys/P1/Hjx/X5MmT5ef3jzjWuXNnjRkzRtOnT9eePXt09OhR/fSnP1VMTIzGjBnjw92gMfByNQAAAABoZLGxsSooKNCQIUM0fPhwffjhh6qsrFRqaqpuv/12/du//ZsqKio85mRnZ2vu3LkaNWqUqqurNXDgQG3btk2tW7f20S7QWCwul8vl6yaAppTafpSvWwBatD6WMF+3ALR4fa7U+LoF4JYwtDTH1y2gheBRcwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAABoIhaLRZs3b/Z1G2hiBG8AAAAAaAQXL17U3LlzZbPZFBgYqLZt26p///5atWqVvv766xte74033pDFYtGIESM86pcvX5bFYlFBQUGdOU6nUz179pTFYpHD4XDXCwoKZLFY3Ffbtm316KOPqri4+Ib7wo0jeAMAAABAAxUXFyspKUm5ublasmSJjhw5on379um5557Te++9px07dtzUuv7+/tqxY4d27txZr/HPPfecoqOjr3q/sLBQ58+f19tvv63/+Z//UWpqqr777rub6g31R/AGAAAAgAaaOXOm/P39dfDgQaWlpalLly6Ki4vTmDFj9P777ys1NdXrvHPnzmnChAkKDw9XcHCwkpOTtX//fvf94OBgPfXUU1q4cOF1e/jggw+Um5urpUuXXnVMZGSk7r77bg0cOFC/+tWvdOLECRUVFd34hnFD/H3dAAAAAAD8kJWXl7tPuoODg72OsVgsdWqVlZUaNGiQYmJitHXrVkVFRenw4cOqra31GLdo0SLZbDZt2rRJjz32mNf1S0tLNX36dG3evFm33XZbvfoOCgqSJFVXV9drPG4ewRsAAAAAvHA6nXI6nR41q9Uqq9XqUSsqKpLL5VJCQoJH/c4779SVK1ckSbNmzdLLL7/scX/Dhg0qKyvTxx9/rPDwcEmSzWar00d0dLTmzp2r559/XmPHjq1z3+VyafLkyZoxY4aSk5N15syZ6+7twoULWrp0qWJiYur0jcbHo+YAAAAA4EVWVpbCwsI8rqysrHrPP3DggBwOh7p27VonwEuSw+FQUlKSO3Rfy4IFC1RWVqZ169bVubdixQp99dVXstvt110nNjZWwcHBio6OVlVVld555x0FBATUb0O4aZx4AwAAAIAXdrtd6enpHrV/Pu2W/n5KbbFYVFhY6FGPi4uT9I9Huv/Z1eretGnTRna7XZmZmRo1apTHvfz8fO3bt69Ob8nJyfrJT36iN9980137y1/+otDQUEVGRur222+v989Hw3DiDQAAAABeWK1WhYaGelzegndERIQeeughrVy5UlVVVfVev3v37nI4HLp06VK9xs+ZM0d+fn5atmyZR3358uU6evSoHA6HHA6Htm3bJknKycnRr3/9a4+xnTp1Unx8PKG7iRG8AQAAAKCBXn/9ddXU1Cg5OVk5OTk6efKkCgsLtX79ep06dUqtWrWqM2fChAmKiorS2LFjtXfvXhUXF+udd97Rvn37vP6MwMBAZWZmavny5R719u3bq1u3bu7rnnvukSTFx8crNja28TeLG0bwBgAAAIAGio+P15EjR5SSkiK73a4ePXooOTlZK1as0Lx587R48eI6cwICApSbm6vIyEiNHDlSiYmJeumll7yG9O9NmjTJ/Qg7fjgsLpfL5esmgKaU2n7U9QcBuGl9LGG+bgFo8fpcqfF1C8AtYWhpjq9bQAvBiTcAAAAAAAYRvAEAAAAAMIjgDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAABgEMEbAAAAAACDCN4AAAAAABhE8AYAAAAAwCCCNwAAAAAABhG8AQAAAAAwiOANAAAAAIBBBG8AAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAAAAAAAYRPAGAAAAAMAggjcAAAAANBGLxaLNmzf7ug00MYI3AAAAADSCixcvau7cubLZbAoMDFTbtm3Vv39/rVq1Sl9//fUNr/fGG2/IYrFoxIgRHvXLly/LYrGooKCgkTqHaf6+bgAAAAAAfuiKi4vVv39/tWnTRkuWLFFiYqKsVquOHTum1atXKyYmRqNHj77hdf39/bVjxw7t3LlTQ4YMMdA5mgLBGwAAAAAaaObMmfL399fBgwcVHBzsrsfFxWnMmDFyuVxe5507d07z58/Xhx9+KKfTqS5duui1115T7969JUnBwcFKS0vTwoULtX///ibZCxofwRsAAAAAvHA6nXI6nR41q9Uqq9XqUSsvL1dubq6WLFniEbr/L4vFUqdWWVmpQYMGKSYmRlu3blVUVJQOHz6s2tpaj3GLFi2SzWbTpk2b9NhjjzVwV/AF/sYbAAAAALzIyspSWFiYx5WVlVVnXFFRkVwulxISEjzqd955p0JCQhQSEqIFCxbUmbdhwwaVlZVp8+bNGjBggGw2m9LS0tS3b1+PcdHR0Zo7d66ef/551dTUNO4m0SQI3gAAAADghd1uV0VFhcdlt9vrPf/AgQNyOBzq2rVrnZNzSXI4HEpKSlJ4ePh111qwYIHKysq0bt26G9oDmgceNQcAAAAAL7w9Vu6NzWaTxWJRYWGhRz0uLk6SFBQU5HXe1eretGnTRna7XZmZmRo1alS956F54MQbAAAAABogIiJCDz30kFauXKmqqqp6z+vevbscDocuXbpUr/Fz5syRn5+fli1bdrOtwkcI3gAAAADQQK+//rpqamqUnJysnJwcnTx5UoWFhVq/fr1OnTqlVq1a1ZkzYcIERUVFaezYsdq7d6+Ki4v1zjvvaN++fV5/RmBgoDIzM7V8+XLT20EjI3gDAAAAQAPFx8fryJEjSklJkd1uV48ePZScnKwVK1Zo3rx5Wrx4cZ05AQEBys3NVWRkpEaOHKnExES99NJLXkP69yZNmuR+hB0/HBbX1b5QDmihUtvzNzGASX0sYb5uAWjx+lzhrcZAUxhamuPrFtBCcOINAAAAAIBBBG8AAAAAAAziUXMAzZrT6VRWVpbsdnu9vs4DwI3j9wwwj98z4NZG8AbQrH355ZcKCwtTRUWFQkNDfd0O0CLxewaYx+8ZcGvjUXMAAAAAAAwieAMAAAAAYBDBGwAAAAAAgwjeAJo1q9WqjIwMXkQDGMTvGWAev2fArY2XqwEAAAAAYBAn3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A2g2XrttdfUsWNHBQYGqnfv3jpw4ICvWwJalN27dys1NVXR0dGyWCzavHmzr1sCWpSsrCzdf//9uv322xUZGamxY8eqsLDQ120B8AGCN4BmKScnR+np6crIyNDhw4fVo0cPDR8+XJ999pmvWwNajKqqKvXo0UOvvfaar1sBWqRdu3Zp1qxZ+uijj7R9+3Z9++23GjZsmKqqqnzdGoAmxteJAWiWevfurfvvv18rV66UJNXW1qpdu3aaM2eOFi5c6OPugJbHYrHo3Xff1dixY33dCtBilZWVKTIyUrt27dLAgQN93Q6AJsSJN4Bmp7q6WocOHVJKSoq75ufnp5SUFO3bt8+HnQEAcPMqKiokSeHh4T7uBEBTI3gDaHY+//xzfffdd2rbtq1HvW3btrp48aKPugIA4ObV1tbq2WefVf/+/dWtWzdftwOgifn7ugEAAACgpZs1a5aOHz+uPXv2+LoVAD5A8AbQ7Nx5551q1aqVSktLPeqlpaWKioryUVcAANyc2bNn67333tPu3bsVGxvr63YA+ACPmgNodgICAtSrVy/l5eW5a7W1tcrLy1Pfvn192BkAAPXncrk0e/Zsvfvuu8rPz1enTp183RIAH+HEG0CzlJ6erkmTJik5OVkPPPCAfv/736uqqkpTpkzxdWtAi1FZWamioiL3508//VQOh0Ph4eFq3769DzsDWoZZs2Zpw4YN2rJli26//Xb3e0rCwsIUFBTk4+4ANCW+TgxAs7Vy5Uq98sorunjxonr27Knly5erd+/evm4LaDEKCgo0ZMiQOvVJkybpjTfeaPqGgBbGYrF4rWdnZ2vy5MlN2wwAnyJ4AwAAAABgEH/jDQAAAACAQQRvAAAAAAAMIngDAAAAAGAQwRsAAAAAAIMI3gAAAAAAGETwBgAAAADAIII3AAAAAAAGEbwBAAAAADCI4A0AAAAAgEEEbwAAAAAADCJ4AwAAAABgEMEbAAAAAACD/h/jkBnm3yjo0gAAAABJRU5ErkJggg==
"
>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_embeddings" class="doc_header"><code>plot_embeddings</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/analysis.py#L213" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_embeddings</code>(<strong><code>glycans</code></strong>, <strong><code>emb</code></strong>=<em><code>None</code></em>, <strong><code>label_list</code></strong>=<em><code>None</code></em>, <strong><code>shape_feature</code></strong>=<em><code>None</code></em>, <strong><code>filepath</code></strong>=<em><code>''</code></em>, <strong><code>alpha</code></strong>=<em><code>0.8</code></em>, <strong><code>palette</code></strong>=<em><code>'colorblind'</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>plots glycan representations for a list of glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr>
<td style="text-align:left">emb (dictionary): stored glycan representations; default takes them from trained species-level SweetNet model</td>
</tr>
<tr>
<td style="text-align:left">label_list (list): list of same length as glycans if coloring of the plot is desired</td>
</tr>
<tr>
<td style="text-align:left">shape_feature (string): monosaccharide/bond used to display alternative shapes for dots on the plot</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): absolute path including full filename allows for saving the plot</td>
</tr>
<tr>
<td style="text-align:left">alpha (float): transparency of points in plot; default:0.8</td>
</tr>
<tr>
<td style="text-align:left">palette (string): color palette to color different classes; default:'colorblind'</td>
</tr>
<tr>
<td style="text-align:left">**kwargs: keyword arguments that are directly passed on to matplotlib</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df_fabales</span> <span class="o">=</span> <span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Order</span> <span class="o">==</span> <span class="s1">&#39;Fabales&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">plot_embeddings</span><span class="p">(</span><span class="n">df_fabales</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">label_list</span> <span class="o">=</span> <span class="n">df_fabales</span><span class="o">.</span><span class="n">Family</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-intense-fg ansi-bold">---------------------------------------------------------------------------</span>
<span class="ansi-red-intense-fg ansi-bold">KeyError</span>                                  Traceback (most recent call last)
<span class="ansi-green-intense-fg ansi-bold">&lt;ipython-input-11-4564563a757b&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">      1</span> df_fabales <span class="ansi-yellow-intense-fg ansi-bold">=</span> df_species<span class="ansi-yellow-intense-fg ansi-bold">[</span>df_species<span class="ansi-yellow-intense-fg ansi-bold">.</span>Order <span class="ansi-yellow-intense-fg ansi-bold">==</span> <span class="ansi-blue-intense-fg ansi-bold">&#39;Fabales&#39;</span><span class="ansi-yellow-intense-fg ansi-bold">]</span><span class="ansi-yellow-intense-fg ansi-bold">.</span>reset_index<span class="ansi-yellow-intense-fg ansi-bold">(</span>drop <span class="ansi-yellow-intense-fg ansi-bold">=</span> <span class="ansi-green-intense-fg ansi-bold">True</span><span class="ansi-yellow-intense-fg ansi-bold">)</span>
<span class="ansi-green-intense-fg ansi-bold">----&gt; 2</span><span class="ansi-yellow-intense-fg ansi-bold"> </span>plot_embeddings<span class="ansi-yellow-intense-fg ansi-bold">(</span>df_fabales<span class="ansi-yellow-intense-fg ansi-bold">.</span>target<span class="ansi-yellow-intense-fg ansi-bold">.</span>values<span class="ansi-yellow-intense-fg ansi-bold">.</span>tolist<span class="ansi-yellow-intense-fg ansi-bold">(</span><span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">,</span> label_list <span class="ansi-yellow-intense-fg ansi-bold">=</span> df_fabales<span class="ansi-yellow-intense-fg ansi-bold">.</span>Family<span class="ansi-yellow-intense-fg ansi-bold">.</span>values<span class="ansi-yellow-intense-fg ansi-bold">.</span>tolist<span class="ansi-yellow-intense-fg ansi-bold">(</span><span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">)</span>

<span class="ansi-green-intense-fg ansi-bold">~\OneDrive\Documents\glycowork\glycowork\motif\analysis.py</span> in <span class="ansi-cyan-fg">plot_embeddings</span><span class="ansi-blue-intense-fg ansi-bold">(glycans, emb, label_list, shape_feature, filepath, alpha, palette, **kwargs)</span>
<span class="ansi-green-fg">    235</span>     <span class="ansi-green-intense-fg ansi-bold">if</span> isinstance<span class="ansi-yellow-intense-fg ansi-bold">(</span>emb<span class="ansi-yellow-intense-fg ansi-bold">,</span> pd<span class="ansi-yellow-intense-fg ansi-bold">.</span>DataFrame<span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">:</span>
<span class="ansi-green-fg">    236</span>         emb <span class="ansi-yellow-intense-fg ansi-bold">=</span> <span class="ansi-yellow-intense-fg ansi-bold">{</span>glycan<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">]</span><span class="ansi-yellow-intense-fg ansi-bold">:</span>emb<span class="ansi-yellow-intense-fg ansi-bold">.</span>iloc<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">,</span><span class="ansi-yellow-intense-fg ansi-bold">:</span><span class="ansi-yellow-intense-fg ansi-bold">]</span> <span class="ansi-green-intense-fg ansi-bold">for</span> k <span class="ansi-green-intense-fg ansi-bold">in</span> range<span class="ansi-yellow-intense-fg ansi-bold">(</span>len<span class="ansi-yellow-intense-fg ansi-bold">(</span>glycans<span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">}</span>
<span class="ansi-green-intense-fg ansi-bold">--&gt; 237</span><span class="ansi-yellow-intense-fg ansi-bold">     </span>embs <span class="ansi-yellow-intense-fg ansi-bold">=</span> np<span class="ansi-yellow-intense-fg ansi-bold">.</span>array<span class="ansi-yellow-intense-fg ansi-bold">(</span><span class="ansi-yellow-intense-fg ansi-bold">[</span>emb<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">]</span> <span class="ansi-green-intense-fg ansi-bold">for</span> k <span class="ansi-green-intense-fg ansi-bold">in</span> glycans<span class="ansi-yellow-intense-fg ansi-bold">]</span><span class="ansi-yellow-intense-fg ansi-bold">)</span>
<span class="ansi-green-fg">    238</span>     <span class="ansi-red-intense-fg ansi-bold">#calculate t-SNE of embeddings</span>
<span class="ansi-green-fg">    239</span>     embs = TSNE(random_state = 42,

<span class="ansi-green-intense-fg ansi-bold">~\OneDrive\Documents\glycowork\glycowork\motif\analysis.py</span> in <span class="ansi-cyan-fg">&lt;listcomp&gt;</span><span class="ansi-blue-intense-fg ansi-bold">(.0)</span>
<span class="ansi-green-fg">    235</span>     <span class="ansi-green-intense-fg ansi-bold">if</span> isinstance<span class="ansi-yellow-intense-fg ansi-bold">(</span>emb<span class="ansi-yellow-intense-fg ansi-bold">,</span> pd<span class="ansi-yellow-intense-fg ansi-bold">.</span>DataFrame<span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">:</span>
<span class="ansi-green-fg">    236</span>         emb <span class="ansi-yellow-intense-fg ansi-bold">=</span> <span class="ansi-yellow-intense-fg ansi-bold">{</span>glycan<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">]</span><span class="ansi-yellow-intense-fg ansi-bold">:</span>emb<span class="ansi-yellow-intense-fg ansi-bold">.</span>iloc<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">,</span><span class="ansi-yellow-intense-fg ansi-bold">:</span><span class="ansi-yellow-intense-fg ansi-bold">]</span> <span class="ansi-green-intense-fg ansi-bold">for</span> k <span class="ansi-green-intense-fg ansi-bold">in</span> range<span class="ansi-yellow-intense-fg ansi-bold">(</span>len<span class="ansi-yellow-intense-fg ansi-bold">(</span>glycans<span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">)</span><span class="ansi-yellow-intense-fg ansi-bold">}</span>
<span class="ansi-green-intense-fg ansi-bold">--&gt; 237</span><span class="ansi-yellow-intense-fg ansi-bold">     </span>embs <span class="ansi-yellow-intense-fg ansi-bold">=</span> np<span class="ansi-yellow-intense-fg ansi-bold">.</span>array<span class="ansi-yellow-intense-fg ansi-bold">(</span><span class="ansi-yellow-intense-fg ansi-bold">[</span>emb<span class="ansi-yellow-intense-fg ansi-bold">[</span>k<span class="ansi-yellow-intense-fg ansi-bold">]</span> <span class="ansi-green-intense-fg ansi-bold">for</span> k <span class="ansi-green-intense-fg ansi-bold">in</span> glycans<span class="ansi-yellow-intense-fg ansi-bold">]</span><span class="ansi-yellow-intense-fg ansi-bold">)</span>
<span class="ansi-green-fg">    238</span>     <span class="ansi-red-intense-fg ansi-bold">#calculate t-SNE of embeddings</span>
<span class="ansi-green-fg">    239</span>     embs = TSNE(random_state = 42,

<span class="ansi-red-intense-fg ansi-bold">KeyError</span>: &#39;Gal(?1-?)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Man(a1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="characterize_monosaccharide" class="doc_header"><code>characterize_monosaccharide</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/analysis.py#L260" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>characterize_monosaccharide</code>(<strong><code>sugar</code></strong>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>mode</code></strong>=<em><code>'sugar'</code></em>, <strong><code>glycan_col_name</code></strong>=<em><code>'target'</code></em>, <strong><code>rank</code></strong>=<em><code>None</code></em>, <strong><code>focus</code></strong>=<em><code>None</code></em>, <strong><code>modifications</code></strong>=<em><code>False</code></em>, <strong><code>filepath</code></strong>=<em><code>''</code></em>, <strong><code>thresh</code></strong>=<em><code>10</code></em>)</p>
</blockquote>
<p>for a given monosaccharide/linkage, return typical neighboring linkage/monosaccharide</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">sugar (string): monosaccharide or linkage</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): dataframe to use for analysis; default:df_species</td>
</tr>
<tr>
<td style="text-align:left">mode (string): either 'sugar' (connected monosaccharides), 'bond' (monosaccharides making a provided linkage), or 'sugarbond' (linkages that a provided monosaccharides makes); default:'sugar'</td>
</tr>
<tr>
<td style="text-align:left">glycan_col_name (string): column name under which glycans can be found; default:'target'</td>
</tr>
<tr>
<td style="text-align:left">rank (string): add column name as string if you want to filter for a group</td>
</tr>
<tr>
<td style="text-align:left">focus (string): add row value as string if you want to filter for a group</td>
</tr>
<tr>
<td style="text-align:left">modifications (bool): set to True if you want to consider modified versions of a monosaccharide; default:False</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): absolute path including full filename allows for saving the plot</td>
</tr>
<tr>
<td style="text-align:left">thresh (int): threshold count of when to include motifs in plot; default:10 occurrences</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Plots modification distribution and typical neighboring bond/monosaccharide</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">characterize_monosaccharide</span><span class="p">(</span><span class="s1">&#39;D-Rha&#39;</span><span class="p">,</span> <span class="n">rank</span> <span class="o">=</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">,</span> <span class="n">focus</span> <span class="o">=</span> <span class="s1">&#39;Bacteria&#39;</span><span class="p">,</span> <span class="n">modifications</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="annotate">annotate<a class="anchor-link" href="#annotate"> </a></h2><blockquote><p>extract curated motifs, graph features, and sequence features from glycan sequences</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="motif_matrix" class="doc_header"><code>motif_matrix</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/annotate.py#L117" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>motif_matrix</code>(<strong><code>df</code></strong>, <strong><code>glycan_col_name</code></strong>, <strong><code>label_col_name</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>generates dataframe with counted glycoletters and disaccharides in glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe containing glycan sequences and labels</td>
</tr>
<tr>
<td style="text-align:left">glycan_col_name (string): column name for glycan sequences</td>
</tr>
<tr>
<td style="text-align:left">label_col_name (string): column name for labels; string</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe with glycoletter + disaccharide counts (columns) for each glycan (rows)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P&#39;</span><span class="p">]</span>
<span class="n">label</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">test_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;glycan&#39;</span><span class="p">:</span><span class="n">glycans</span><span class="p">,</span> <span class="s1">&#39;eukaryotic&#39;</span><span class="p">:</span><span class="n">label</span><span class="p">})</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Glyco-Motif Test&quot;</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">motif_matrix</span><span class="p">(</span><span class="n">test_df</span><span class="p">,</span> <span class="s1">&#39;glycan&#39;</span><span class="p">,</span> <span class="s1">&#39;eukaryotic&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="annotate_glycan" class="doc_header"><code>annotate_glycan</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/annotate.py#L178" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>annotate_glycan</code>(<strong><code>glycan</code></strong>, <strong><code>motifs</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>extra</code></strong>=<em><code>'termini'</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>, <strong><code>termini_list</code></strong>=<em><code>[]</code></em>)</p>
</blockquote>
<p>searches for known motifs in glycan sequence</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): IUPAC-condensed glycan sequence</td>
</tr>
<tr>
<td style="text-align:left">motifs (dataframe): dataframe of glycan motifs (name + sequence); default:motif_list</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
<tr>
<td style="text-align:left">extra (string): 'ignore' skips this, 'wildcards' allows for wildcard matching', and 'termini' allows for positional matching; default:'termini'</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcard names (such as 'z1-z', 'Hex', 'HexNAc', 'Sia')</td>
</tr>
<tr>
<td style="text-align:left">termini_list (list): list of monosaccharide/linkage positions (from 'terminal','internal', and 'flexible')</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe with counts of motifs in glycan</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="annotate_dataset" class="doc_header"><code>annotate_dataset</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/annotate.py#L224" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>annotate_dataset</code>(<strong><code>glycans</code></strong>, <strong><code>motifs</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>feature_set</code></strong>=<em><code>['known']</code></em>, <strong><code>extra</code></strong>=<em><code>'termini'</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>, <strong><code>termini_list</code></strong>=<em><code>[]</code></em>, <strong><code>condense</code></strong>=<em><code>False</code></em>, <strong><code>estimate_speedup</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>wrapper function to annotate motifs in list of glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr>
<td style="text-align:left">motifs (dataframe): dataframe of glycan motifs (name + sequence); default:motif_list</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our data; default:lib</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default is 'known'; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans) and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
<tr>
<td style="text-align:left">extra (string): 'ignore' skips this, 'wildcards' allows for wildcard matching', and 'termini' allows for positional matching; default:'termini'</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcard names (such as 'z1-z', 'Hex', 'HexNAc', 'Sia')</td>
</tr>
<tr>
<td style="text-align:left">termini_list (list): list of monosaccharide/linkage positions (from 'terminal','internal', and 'flexible')</td>
</tr>
<tr>
<td style="text-align:left">condense (bool): if True, throws away columns with only zeroes; default:False</td>
</tr>
<tr>
<td style="text-align:left">estimate_speedup (bool): if True, pre-selects motifs for those which are present in glycans, not 100% exact; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe of glycans (rows) and presence/absence of known motifs (columns)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Annotate Test&quot;</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">annotate_dataset</span><span class="p">(</span><span class="n">glycans</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_k_saccharides" class="doc_header"><code>get_k_saccharides</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/annotate.py#L280" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_k_saccharides</code>(<strong><code>glycan</code></strong>, <strong><code>size</code></strong>=<em><code>3</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>function to retrieve k-saccharides (default:trisaccharides) occurring in a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string or networkx): glycan in IUPAC-condensed nomenclature or as networkx graph</td>
</tr>
<tr>
<td style="text-align:left">size (int): number of monosaccharides per -saccharide, default:3 (for trisaccharides)</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our data; default:lib</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of trisaccharides in glycan in IUPAC-condensed nomenclature</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_k_saccharides</span><span class="p">(</span><span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P&#39;</span><span class="p">,</span>
                 <span class="n">k</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="graph">graph<a class="anchor-link" href="#graph"> </a></h2><blockquote><p>convert glycan sequences to graphs and contains helper functions to search for motifs / check whether two sequences describe the same sequence, etc.</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="glycan_to_nxGraph" class="doc_header"><code>glycan_to_nxGraph</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L163" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>glycan_to_nxGraph</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>termini</code></strong>=<em><code>'ignore'</code></em>, <strong><code>termini_list</code></strong>=<em><code>None</code></em>, <strong><code>override_reducing_end</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>wrapper for converting glycans into networkx graphs; also works with floating substituents</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">termini (string): whether to encode terminal/internal position of monosaccharides, 'ignore' for skipping, 'calc' for automatic annotation, or 'provided' if this information is provided in termini_list; default:'ignore'</td>
</tr>
<tr>
<td style="text-align:left">termini_list (list): list of monosaccharide/linkage positions (from 'terminal','internal', and 'flexible')</td>
</tr>
<tr>
<td style="text-align:left">override_reducing_end (bool): if True, it allows graph generation for glycans ending in a linkage; though the output doesn't work with all downstream functions; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns networkx graph object of glycan</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Glycan to networkx Graph (only edges printed)&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">glycan_to_nxGraph</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">edges</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="graph_to_string" class="doc_header"><code>graph_to_string</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L528" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>graph_to_string</code>(<strong><code>graph</code></strong>, <strong><code>fallback</code></strong>=<em><code>False</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>converts glycan graph back to IUPAC-condensed format</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">graph (networkx object): glycan graph, works with most glycans. Will often not properly format repeat glycans, e.g., xyloglucan etc</td>
</tr>
<tr>
<td style="text-align:left">fallback (bool): just searches for the corresponding string in df_glycan; default:False</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns glycan in IUPAC-condensed format (string)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">graph_to_string</span><span class="p">(</span><span class="n">glycan_to_nxGraph</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="compare_glycans" class="doc_header"><code>compare_glycans</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L242" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>compare_glycans</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>wildcards</code></strong>=<em><code>False</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>)</p>
</blockquote>
<p>returns True if glycans are the same and False if not</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string or networkx object): glycan in IUPAC-condensed format or as a precomputed networkx object</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (stringor networkx object): glycan in IUPAC-condensed format or as a precomputed networkx object</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">wildcards (bool): set to True to allow wildcards (e.g., '?1-?', 'monosaccharide'); default is False</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcards to consider, in the form of '?1-?' etc.</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns True if two glycans are the same and False if not</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Graph Isomorphism Test&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">compare_glycans</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">,</span>
                      <span class="s1">&#39;Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="subgraph_isomorphism" class="doc_header"><code>subgraph_isomorphism</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L281" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>subgraph_isomorphism</code>(<strong><code>glycan</code></strong>, <strong><code>motif</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>extra</code></strong>=<em><code>'ignore'</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>[]</code></em>, <strong><code>termini_list</code></strong>=<em><code>[]</code></em>, <strong><code>count</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>returns True if motif is in glycan and False if not</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format or as graph in NetworkX format</td>
</tr>
<tr>
<td style="text-align:left">motif (string): glycan motif in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">extra (string): 'ignore' skips this, 'wildcards' allows for wildcard matching', and 'termini' allows for positional matching; default:'ignore'</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list of wildcard names (such as '?1-?', 'Hex', 'HexNAc', 'Sia')</td>
</tr>
<tr>
<td style="text-align:left">termini_list (list): list of monosaccharide/linkage positions (from 'terminal','internal', and 'flexible')</td>
</tr>
<tr>
<td style="text-align:left">count (bool): whether to return the number or absence/presence of motifs; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns True if motif is in glycan and False if not</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Subgraph Isomorphism Test&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subgraph_isomorphism</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;Fuc(a1-6)GlcNAc&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_graph_features" class="doc_header"><code>generate_graph_features</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L366" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_graph_features</code>(<strong><code>glycan</code></strong>, <strong><code>glycan_graph</code></strong>=<em><code>True</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>label</code></strong>=<em><code>'network'</code></em>)</p>
</blockquote>
<p>compute graph features of glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string or networkx object): glycan in IUPAC-condensed format (or glycan network if glycan_graph=False)</td>
</tr>
<tr>
<td style="text-align:left">glycan_graph (bool): True expects a glycan, False expects a network (from construct_network); default:True</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">label (string): Label to place in output dataframe if glycan_graph=False; default:'network'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a pandas dataframe with different graph features as columns and glycan as row</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="largest_subgraph" class="doc_header"><code>largest_subgraph</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/graph.py#L606" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>largest_subgraph</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>find the largest common subgraph of two glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string or networkx): glycan in IUPAC-condensed format or as networkx graph</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string or networkx): glycan in IUPAC-condensed format or as networkx graph</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the largest common subgraph as a string in IUPAC-condensed; returns empty string if there is no common subgraph</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycan1</span> <span class="o">=</span> <span class="s1">&#39;Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span>
<span class="n">glycan2</span> <span class="o">=</span> <span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span>
<span class="n">largest_subgraph</span><span class="p">(</span><span class="n">glycan1</span><span class="p">,</span> <span class="n">glycan2</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="processing">processing<a class="anchor-link" href="#processing"> </a></h2><blockquote><p>process IUPAC-condensed glycan sequences into glycoletters etc.</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="min_process_glycans" class="doc_header"><code>min_process_glycans</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/processing.py#L18" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>min_process_glycans</code>(<strong><code>glycan_list</code></strong>)</p>
</blockquote>
<p>converts list of glycans into a nested lists of glycoletters</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_list (list): list of glycans in IUPAC-condensed format as strings</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of glycoletter lists</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">min_process_glycans</span><span class="p">([</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_lib" class="doc_header"><code>get_lib</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/processing.py#L31" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_lib</code>(<strong><code>glycan_list</code></strong>)</p>
</blockquote>
<p>returns sorted list of unique glycoletters in list of glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_list (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns sorted list of unique glycoletters (strings) in glycan_list</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_lib</span><span class="p">([</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="expand_lib" class="doc_header"><code>expand_lib</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/processing.py#L48" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>expand_lib</code>(<strong><code>libr</code></strong>, <strong><code>glycan_list</code></strong>)</p>
</blockquote>
<p>updates libr with newly introduced glycoletters</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset</td>
</tr>
<tr>
<td style="text-align:left">glycan_list (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns new lib</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lib1</span> <span class="o">=</span> <span class="n">get_lib</span><span class="p">([</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">])</span>
<span class="n">lib2</span> <span class="o">=</span> <span class="n">expand_lib</span><span class="p">(</span><span class="n">lib1</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">])</span>
<span class="n">lib2</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="presence_to_matrix" class="doc_header"><code>presence_to_matrix</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/processing.py#L88" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>presence_to_matrix</code>(<strong><code>df</code></strong>, <strong><code>glycan_col_name</code></strong>=<em><code>'target'</code></em>, <strong><code>label_col_name</code></strong>=<em><code>'Species'</code></em>)</p>
</blockquote>
<p>converts a dataframe such as df_species to absence/presence matrix</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycan occurrence, rows are glycan-label pairs</td>
</tr>
<tr>
<td style="text-align:left">glycan_col_name (string): column name under which glycans are stored; default:target</td>
</tr>
<tr>
<td style="text-align:left">label_col_name (string): column name under which labels are stored; default:Species</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns pandas dataframe with labels as rows and glycan occurrences as columns</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">out</span> <span class="o">=</span> <span class="n">presence_to_matrix</span><span class="p">(</span><span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Order</span> <span class="o">==</span> <span class="s1">&#39;Fabales&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">),</span>
                         <span class="n">label_col_name</span> <span class="o">=</span> <span class="s1">&#39;Family&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="choose_correct_isoform" class="doc_header"><code>choose_correct_isoform</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/processing.py#L107" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>choose_correct_isoform</code>(<strong><code>glycans</code></strong>, <strong><code>reverse</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>given a list of glycan branch isomers, this function returns the correct isomer</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): glycans in IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">reverse (bool): whether to return the correct isomer (False) or everything except the correct isomer (True); default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the correct isomer as a string (if reverse=False, otherwise it returns a list of strings)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="query">query<a class="anchor-link" href="#query"> </a></h2><blockquote><p>for interacting with the databases contained in glycowork, delivering insights for sequences of interest</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_insight" class="doc_header"><code>get_insight</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/query.py#L9" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_insight</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>motifs</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>prints out meta-information about a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">motifs (dataframe): dataframe of glycan motifs (name + sequence); default:motif_list</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Test get_insight with &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;&quot;</span><span class="p">)</span>
<span class="n">get_insight</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="glytoucan_to_glycan" class="doc_header"><code>glytoucan_to_glycan</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/query.py#L63" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>glytoucan_to_glycan</code>(<strong><code>ids</code></strong>, <strong><code>revert</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>interconverts GlyTouCan IDs and glycans in IUPAC-condensed</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">ids (list): list of GlyTouCan IDs as strings (if using glycans instead, change 'revert' to True</td>
</tr>
<tr>
<td style="text-align:left">revert (bool): whether glycans should be mapped to GlyTouCan IDs or vice versa; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of either GlyTouCan IDs or glycans in IUPAC-condensed</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glytoucan_to_glycan</span><span class="p">([</span><span class="s1">&#39;G59003TN&#39;</span><span class="p">,</span> <span class="s1">&#39;G88185OE&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="tokenization">tokenization<a class="anchor-link" href="#tokenization"> </a></h2><blockquote><p>helper functions to map m/z--&gt;composition, composition--&gt;structure, structure--&gt;motif, and more</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="string_to_labels" class="doc_header"><code>string_to_labels</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L82" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>string_to_labels</code>(<strong><code>character_string</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>tokenizes word by indexing characters in passed library</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">character_string (string): string of characters to index</td>
</tr>
<tr>
<td style="text-align:left">libr (list): list of library items</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns indexes of characters in library</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">string_to_labels</span><span class="p">([</span><span class="s1">&#39;Man&#39;</span><span class="p">,</span><span class="s1">&#39;a1-3&#39;</span><span class="p">,</span><span class="s1">&#39;Man&#39;</span><span class="p">,</span><span class="s1">&#39;a1-6&#39;</span><span class="p">,</span><span class="s1">&#39;Man&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="pad_sequence" class="doc_header"><code>pad_sequence</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L96" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>pad_sequence</code>(<strong><code>seq</code></strong>, <strong><code>max_length</code></strong>, <strong><code>pad_label</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>brings all sequences to same length by adding padding token</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">seq (list): sequence to pad (from string_to_labels)</td>
</tr>
<tr>
<td style="text-align:left">max_length (int): sequence length to pad to</td>
</tr>
<tr>
<td style="text-align:left">pad_label (int): which padding label to use</td>
</tr>
<tr>
<td style="text-align:left">libr (list): list of library items</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns padded sequence</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pad_sequence</span><span class="p">(</span><span class="n">string_to_labels</span><span class="p">([</span><span class="s1">&#39;Man&#39;</span><span class="p">,</span><span class="s1">&#39;a1-3&#39;</span><span class="p">,</span><span class="s1">&#39;Man&#39;</span><span class="p">,</span><span class="s1">&#39;a1-6&#39;</span><span class="p">,</span><span class="s1">&#39;Man&#39;</span><span class="p">]),</span> <span class="mi">7</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="stemify_glycan" class="doc_header"><code>stemify_glycan</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L179" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>stemify_glycan</code>(<strong><code>glycan</code></strong>, <strong><code>stem_lib</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>removes modifications from all monosaccharides in a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">stem_lib (dictionary): dictionary of form modified_monosaccharide:core_monosaccharide; default:created from lib</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns stemmed glycan as string</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="stemify_dataset" class="doc_header"><code>stemify_dataset</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L242" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>stemify_dataset</code>(<strong><code>df</code></strong>, <strong><code>stem_lib</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>glycan_col_name</code></strong>=<em><code>'target'</code></em>, <strong><code>rarity_filter</code></strong>=<em><code>1</code></em>)</p>
</blockquote>
<p>stemifies all glycans in a dataset by removing monosaccharide modifications</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycans in IUPAC-condensed format in column glycan_col_name</td>
</tr>
<tr>
<td style="text-align:left">stem_lib (dictionary): dictionary of form modified_monosaccharide:core_monosaccharide; default:created from lib</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
<tr>
<td style="text-align:left">glycan_col_name (string): column name under which glycans are stored; default:target</td>
</tr>
<tr>
<td style="text-align:left">rarity_filter (int): how often monosaccharide modification has to occur to not get removed; default:1</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns df with glycans stemified</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="mask_rare_glycoletters" class="doc_header"><code>mask_rare_glycoletters</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L662" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>mask_rare_glycoletters</code>(<strong><code>glycans</code></strong>, <strong><code>thresh_monosaccharides</code></strong>=<em><code>None</code></em>, <strong><code>thresh_linkages</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>masks rare monosaccharides and linkages in a list of glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed form</td>
</tr>
<tr>
<td style="text-align:left">thresh_monosaccharides (int): threshold-value for monosaccharides seen as "rare"; default:(0.001*len(glycans))</td>
</tr>
<tr>
<td style="text-align:left">thresh_linkages (int): threshold-value for linkages seen as "rare"; default:(0.03*len(glycans))</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of glycans in IUPAC-condensed with masked rare monosaccharides and linkages</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="match_composition_relaxed" class="doc_header"><code>match_composition_relaxed</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L445" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>match_composition_relaxed</code>(<strong><code>composition</code></strong>, <strong><code>group</code></strong>, <strong><code>level</code></strong>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Given a coarse-grained monosaccharide composition (Hex, HexNAc, etc.), it returns all corresponding glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">composition (dict): a dictionary indicating the composition to match (for example {"Fuc":1, "Gal":1, "GlcNAc":1})</td>
</tr>
<tr>
<td style="text-align:left">group (string): name of the Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain used to filter</td>
</tr>
<tr>
<td style="text-align:left">level (string): Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): glycan dataframe for searching glycan structures; default:df_species</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (string): filters possible glycans by reducing end monosaccharide; default:None</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of glycans matching composition in IUPAC-condensed</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">match_composition_relaxed</span><span class="p">({</span><span class="s2">&quot;Hex&quot;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;HexNAc&quot;</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;dHex&quot;</span><span class="p">:</span><span class="mi">1</span><span class="p">},</span> <span class="s2">&quot;Homo_sapiens&quot;</span><span class="p">,</span> <span class="s2">&quot;Species&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="condense_composition_matching" class="doc_header"><code>condense_composition_matching</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L478" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>condense_composition_matching</code>(<strong><code>matched_composition</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Given a list of glycans matching a composition, find the minimum number of glycans characterizing this set</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">matched_composition (list): list of glycans matching to a composition</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns minimal list of glycans that match a composition</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">match_comp</span> <span class="o">=</span> <span class="n">match_composition_relaxed</span><span class="p">({</span><span class="s1">&#39;Hex&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;HexNAc&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Neu5Ac&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">},</span>
                                       <span class="s1">&#39;Homo_sapiens&#39;</span><span class="p">,</span> <span class="s1">&#39;Species&#39;</span><span class="p">,</span> <span class="n">reducing_end</span> <span class="o">=</span> <span class="s1">&#39;GalNAc&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">match_comp</span><span class="p">)</span>
<span class="n">condense_composition_matching</span><span class="p">(</span><span class="n">match_comp</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="mz_to_structures" class="doc_header"><code>mz_to_structures</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L561" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>mz_to_structures</code>(<strong><code>mz_list</code></strong>, <strong><code>reducing_end</code></strong>, <strong><code>group</code></strong>=<em><code>'Homo_sapiens'</code></em>, <strong><code>level</code></strong>=<em><code>'Species'</code></em>, <strong><code>abundances</code></strong>=<em><code>None</code></em>, <strong><code>mode</code></strong>=<em><code>'positive'</code></em>, <strong><code>mass_value</code></strong>=<em><code>'monoisotopic'</code></em>, <strong><code>sample_prep</code></strong>=<em><code>'underivatized'</code></em>, <strong><code>mass_tolerance</code></strong>=<em><code>0.2</code></em>, <strong><code>check_all_adducts</code></strong>=<em><code>False</code></em>, <strong><code>check_specific_adduct</code></strong>=<em><code>None</code></em>, <strong><code>ptm</code></strong>=<em><code>False</code></em>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>verbose</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>wrapper function to map precursor masses to structures, condense them, and match them with relative intensities</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">mz_list (list): list of precursor masses</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (string): filters possible glycans by reducing end monosaccharide</td>
</tr>
<tr>
<td style="text-align:left">group (string): name of the Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain used to filter; default:Homo_sapiens</td>
</tr>
<tr>
<td style="text-align:left">level (string): Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain; default:Species</td>
</tr>
<tr>
<td style="text-align:left">abundances (dataframe): every row one composition (matching mz_list in order), every column one sample; default:pd.DataFrame([range(len(mz_list))]*2).T</td>
</tr>
<tr>
<td style="text-align:left">mode (string): whether mz_value comes from MS in 'positive' or 'negative' mode; default:'positive'</td>
</tr>
<tr>
<td style="text-align:left">mass_value (string): whether the expected mass is 'monoisotopic' or 'average'; default:'monoisotopic'</td>
</tr>
<tr>
<td style="text-align:left">sample_prep (string): whether the glycans has been 'underivatized', 'permethylated', or 'peracetylated'; default:'underivatized'</td>
</tr>
<tr>
<td style="text-align:left">mass_tolerance (float): how much deviation to tolerate for a match; default:0.2</td>
</tr>
<tr>
<td style="text-align:left">check_all_adducts (bool): whether to also check for matches with ion adducts (depending on mode); default:False</td>
</tr>
<tr>
<td style="text-align:left">check_specific_adduct (string): choose adduct from 'H+', 'Na+', 'K+', 'H', 'Acetate', 'Trifluoroacetic acid'; default:None</td>
</tr>
<tr>
<td style="text-align:left">ptm (bool): whether to check for post-translational modification (sulfation, phosphorylation); default:False</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): glycan dataframe for searching glycan structures; default:df_species</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
<tr>
<td style="text-align:left">verbose (bool): whether to print any non-matching compositions; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe of (matched structures) x (relative intensities)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="compositions_to_structures" class="doc_header"><code>compositions_to_structures</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L515" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>compositions_to_structures</code>(<strong><code>composition_list</code></strong>, <strong><code>group</code></strong>=<em><code>'Homo_sapiens'</code></em>, <strong><code>level</code></strong>=<em><code>'Species'</code></em>, <strong><code>abundances</code></strong>=<em><code>None</code></em>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>None</code></em>, <strong><code>verbose</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>wrapper function to map compositions to structures, condense them, and match them with relative intensities</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">composition_list (list): list of composition dictionaries of the form {'Hex': 1, 'HexNAc': 1}</td>
</tr>
<tr>
<td style="text-align:left">group (string): name of the Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain used to filter; default:Homo_sapiens</td>
</tr>
<tr>
<td style="text-align:left">level (string): Species, Genus, Family, Order, Class, Phylum, Kingdom, or Domain; default:Species</td>
</tr>
<tr>
<td style="text-align:left">abundances (dataframe): every row one composition (matching composition_list in order), every column one sample;default:pd.DataFrame([range(len(composition_list))]*2).T</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): glycan dataframe for searching glycan structures; default:df_species</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (string): filters possible glycans by reducing end monosaccharide; default:None</td>
</tr>
<tr>
<td style="text-align:left">verbose (bool): whether to print any non-matching compositions; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe of (matched structures) x (relative intensities)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="structures_to_motifs" class="doc_header"><code>structures_to_motifs</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L624" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>structures_to_motifs</code>(<strong><code>df</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>feature_set</code></strong>=<em><code>['exhaustive']</code></em>, <strong><code>form</code></strong>=<em><code>'wide'</code></em>)</p>
</blockquote>
<p>function to convert relative intensities of glycan structures to those of glycan motifs</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): function expects glycans in the first column and rel. intensities of each sample in a new column</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our dataset; default:lib</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default is 'exhaustive'; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans) and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
<tr>
<td style="text-align:left">form (string): whether to return 'wide' or 'long' dataframe; default:'wide'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe of motifs, relative intensities, and sample IDs</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="canonicalize_iupac" class="doc_header"><code>canonicalize_iupac</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L708" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>canonicalize_iupac</code>(<strong><code>glycan</code></strong>)</p>
</blockquote>
<p>converts a glycan from any IUPAC flavor into the exact IUPAC-condensed version that is optimized for glycowork</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan sequence in IUPAC; post-biosynthetic modifications can still be an issue</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns glycan as a string in canonicalized IUPAC-condensed</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="structure_to_basic" class="doc_header"><code>structure_to_basic</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L810" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>structure_to_basic</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>converts a monosaccharide- and linkage-defined glycan structure to the base topology</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the glycan topology as a string</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="glycan_to_composition" class="doc_header"><code>glycan_to_composition</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L832" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>glycan_to_composition</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>go_fast</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>maps glycan to its composition</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">go_fast (bool): if True, it will only do stemification if necessary (2-10x speed-up), <em>will fail if non-allowed modifications occur</em>; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a dictionary of form "Monosaccharide" : count</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="calculate_theoretical_mass" class="doc_header"><code>calculate_theoretical_mass</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/motif/tokenization.py#L896" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>calculate_theoretical_mass</code>(<strong><code>glycan</code></strong>, <strong><code>mass_value</code></strong>=<em><code>'monoisotopic'</code></em>, <strong><code>sample_prep</code></strong>=<em><code>'underivatized'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>go_fast</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>given a glycan, calculates its theoretical mass; only allowed extra-modifications are methylation, sulfation, phosphorylation</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">mass_value (string): whether the expected mass is 'monoisotopic' or 'average'; default:'monoisotopic'</td>
</tr>
<tr>
<td style="text-align:left">sample_prep (string): whether the glycans has been 'underivatized', 'permethylated', or 'peracetylated'; default:'underivatized'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">go_fast (bool): if True, it will only do stemification if necessary (2-10x speed-up), <em>will fail if non-allowed modifications occur</em>; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the theoretical mass of input glycan</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

