---

title: network


keywords: fastai
sidebar: home_sidebar

summary: "Analyzing groups of glycans as networks (e.g., biosynthesis networks)"
description: "Analyzing groups of glycans as networks (e.g., biosynthesis networks)"
nb_path: "06_network.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 06_network.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/glycowork/network.html"><code>network</code></a> contains functions to arrange and analyze glycans in the context of networks. In such a network, each node represents a glycan and edges represent, for instance, their connection via a biosynthetic step. It should be noted, since <code>glycowork</code> treats glycans as molecular graphs, that these networks represent hierarchical graphs, with the network being one graph and each node within the network also a graph. <a href="/glycowork/network.html"><code>network</code></a> contains the following modules:</p>
<ul>
<li><code>biosynthesis</code> contains functions to construct and analyze biosynthetic glycan networks</li>
<li><code>evolution</code> contains functions to compare (taxonomic) groups as to their glycan repertoires</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="biosynthesis">biosynthesis<a class="anchor-link" href="#biosynthesis"> </a></h1><blockquote><p>constructing and analyzing biosynthetic glycan networks</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="subgraph_to_string" class="doc_header"><code>subgraph_to_string</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L45" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>subgraph_to_string</code>(<strong><code>subgraph</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>converts glycan subgraph back to IUPAC-condensed format</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">subgraph (networkx object): subgraph of one monosaccharide and its linkage</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns glycan motif in IUPAC-condensed format (string)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_neighbors" class="doc_header"><code>get_neighbors</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L61" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_neighbors</code>(<strong><code>glycan</code></strong>, <strong><code>glycans</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>graphs</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>find (observed) biosynthetic precursors of a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">graphs (list): list of glycans in df as graphs; optional if you call get_neighbors often with the same df and want to provide it precomputed</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) a list of direct glycan precursors in IUPAC-condensed</td>
</tr>
<tr>
<td style="text-align:left">(2) a list of indices where each precursor from (1) can be found in glycans</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_neighbors</span><span class="p">(</span><span class="s1">&#39;Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc&#39;</span><span class="p">,</span>
             <span class="p">[</span><span class="s1">&#39;Gal(b1-3)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc&#39;</span><span class="p">,</span> <span class="s1">&#39;Gal(b1-4)Glc&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([&#39;Gal(b1-3)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc&#39;], [[0], [], [], []])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="create_adjacency_matrix" class="doc_header"><code>create_adjacency_matrix</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L86" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>create_adjacency_matrix</code>(<strong><code>glycans</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>virtual_nodes</code></strong>=<em><code>False</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>)</p>
</blockquote>
<p>creates a biosynthetic adjacency matrix from a list of glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">virtual_nodes (bool): whether to include virtual nodes in network; default:False</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) adjacency matrix (glycan X glycan) denoting whether two glycans are connected by one biosynthetic step</td>
</tr>
<tr>
<td style="text-align:left">(2) list of which nodes are virtual nodes (empty list if virtual_nodes is False)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_diff" class="doc_header"><code>find_diff</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L145" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_diff</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>finds the subgraph that differs between glycans and returns it, will only work if the differing subgraph is connected</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns difference between glycan_a and glycan_b in IUPAC-condensed</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_diff</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;Fuca1-6&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="construct_network" class="doc_header"><code>construct_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L171" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>construct_network</code>(<strong><code>glycans</code></strong>, <strong><code>add_virtual_nodes</code></strong>=<em><code>'none'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>, <strong><code>ptm</code></strong>=<em><code>False</code></em>, <strong><code>allowed_ptms</code></strong>=<em><code>['OS', '3S', '6S', '1P', '6P', 'OAc', '4Ac']</code></em>)</p>
</blockquote>
<p>visualize biosynthetic network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">add_virtual_nodes (string): indicates whether no ('none'), proximal ('simple'), or all ('exhaustive') virtual nodes should be added; default:'none'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
<tr>
<td style="text-align:left">ptm (bool): whether to consider post-translational modifications in the network construction; default:False</td>
</tr>
<tr>
<td style="text-align:left">allowed_ptms (list): list of PTMs to consider</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a networkx object of the network</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_network" class="doc_header"><code>plot_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L233" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_network</code>(<strong><code>network</code></strong>, <strong><code>plot_format</code></strong>=<em><code>'kamada_kawai'</code></em>, <strong><code>edge_label_draw</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>visualizes biosynthetic network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network, returned from construct_network</td>
</tr>
<tr>
<td style="text-align:left">plot_format (string): how to layout network, either 'kamada_kawai' or 'spring'; default:'kamada_kawai'</td>
</tr>
<tr>
<td style="text-align:left">edge_label_draw (bool): draws edge labels if True; default:True</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_virtual_nodes" class="doc_header"><code>get_virtual_nodes</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L355" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_virtual_nodes</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>)</p>
</blockquote>
<p>find unobserved biosynthetic precursors of a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of virtual node graphs</td>
</tr>
<tr>
<td style="text-align:left">(2) list of virtual nodes in IUPAC-condensed format</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_virtual_nodes</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([&lt;networkx.classes.graph.Graph at 0x226d77d9a90&gt;,
  &lt;networkx.classes.graph.Graph at 0x226d77d9220&gt;,
  &lt;networkx.classes.graph.Graph at 0x226d77d9670&gt;],
 [&#39;Man(a1-3)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;,
  &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-6)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_shared_virtuals" class="doc_header"><code>find_shared_virtuals</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L276" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_shared_virtuals</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>)</p>
</blockquote>
<p>finds virtual nodes that are shared between two glycans (i.e., that connect these two glycans)</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of edges between glycan and virtual node (if virtual node connects the two glycans)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_shared_virtuals</span><span class="p">(</span><span class="s1">&#39;GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;),
 (&#39;GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fill_with_virtuals" class="doc_header"><code>fill_with_virtuals</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L310" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fill_with_virtuals</code>(<strong><code>glycans</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>)</p>
</blockquote>
<p>for a list of glycans, identify virtual nodes connecting observed glycans and return their edges</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of edges that connect observed glycans to virtual nodes</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_path" class="doc_header"><code>find_path</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L434" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_path</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>find virtual node path between two glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of edges to connect glycan_a and glycan_b via virtual nodes</td>
</tr>
<tr>
<td style="text-align:left">(2) dictionary of edge labels detailing difference between two connected nodes</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span><span class="s1">&#39;Gal(b1-4)Glc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;),
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;)],
 {(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;,
   &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;): &#39;Galb1-4&#39;,
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;): &#39;GlcNAcb1-3&#39;})</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_shortest_path" class="doc_header"><code>find_shortest_path</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L492" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_shortest_path</code>(<strong><code>goal_glycan</code></strong>, <strong><code>glycan_list</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>finds the glycan with the shortest path via virtual nodes to the goal glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">goal_glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_list (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of edges of shortest path to connect goal_glycan and glycan via virtual nodes</td>
</tr>
<tr>
<td style="text-align:left">(2) dictionary of edge labels detailing difference between two connected nodes in shortest path</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_shortest_path</span><span class="p">(</span><span class="s1">&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span>
                  <span class="p">[</span><span class="s1">&#39;Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span> <span class="s1">&#39;Gal(b1-4)GlcNAc-ol&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;),
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;)],
 {(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;,
   &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;): &#39;Galb1-4&#39;,
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;): &#39;GlcNAcb1-3&#39;})</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="network_alignment" class="doc_header"><code>network_alignment</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L545" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>network_alignment</code>(<strong><code>network_a</code></strong>, <strong><code>network_b</code></strong>)</p>
</blockquote>
<p>combines two networks into a new network
| Arguments:
| :-
| network_a (networkx object): biosynthetic network from construct_network
| network_b (networkx object): biosynthetic network from construct_network</p>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns combined network as a networkx object</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="infer_virtual_nodes" class="doc_header"><code>infer_virtual_nodes</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L566" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>infer_virtual_nodes</code>(<strong><code>network_a</code></strong>, <strong><code>network_b</code></strong>, <strong><code>combined</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>identifies virtual nodes that have been observed in other species</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network_a (networkx object): biosynthetic network from construct_network</td>
</tr>
<tr>
<td style="text-align:left">network_b (networkx object): biosynthetic network from construct_network</td>
</tr>
<tr>
<td style="text-align:left">combined (networkx object): merged network of network_a and network_b from network_alignment; default:None</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) tuple of (virtual nodes of network_a observed in network_b, virtual nodes occurring in both network_a and network_b)</td>
</tr>
<tr>
<td style="text-align:left">(2) tuple of (virtual nodes of network_b observed in network_a, virtual nodes occurring in both network_a and network_b)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="infer_network" class="doc_header"><code>infer_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L604" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>infer_network</code>(<strong><code>network</code></strong>, <strong><code>network_species</code></strong>, <strong><code>species_list</code></strong>, <strong><code>filepath</code></strong>=<em><code>None</code></em>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>add_virtual_nodes</code></strong>=<em><code>'exhaustive'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>replaces virtual nodes if they are observed in other species</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network that should be inferred</td>
</tr>
<tr>
<td style="text-align:left">network_species (string): species from which the network stems</td>
</tr>
<tr>
<td style="text-align:left">species_list (list): list of species to compare network to</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): filepath to load biosynthetic networks from other species, if precalculated (def. recommended, as calculation will take ~1.5 hours); default:None</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): dataframe containing species-specific glycans, only needed if filepath=None;default:None</td>
</tr>
<tr>
<td style="text-align:left">add_virtual_nodes (string): indicates whether no ('None'), proximal ('simple'), or all ('exhaustive') virtual nodes should be added;only needed if filepath=None;default:'exhaustive'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used;only needed if filepath=None</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed;only needed if filepath=None;default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes;only needed if filepath=None;default:5</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns network with filled in virtual nodes</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="process_ptm" class="doc_header"><code>process_ptm</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L709" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>process_ptm</code>(<strong><code>glycans</code></strong>, <strong><code>allowed_ptms</code></strong>=<em><code>['OS', '3S', '6S', '1P', '6P', 'OAc', '4Ac']</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>suffix</code></strong>=<em><code>'-ol'</code></em>)</p>
</blockquote>
<p>identifies glycans that contain post-translational modifications and their biosynthetic precursor</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">allowed_ptms (list): list of PTMs to consider</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">suffix (string): optional suffix to be added to the stemified glycan; default:'-ol'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of edge tuples between input glycans and their biosynthetic precusor without PTM</td>
</tr>
<tr>
<td style="text-align:left">(2) list of the PTMs that are contained in the input glycans</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="update_network" class="doc_header"><code>update_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L736" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>update_network</code>(<strong><code>network_in</code></strong>, <strong><code>edge_list</code></strong>, <strong><code>edge_labels</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>updates a network with new edges and their labels</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): network that should be modified</td>
</tr>
<tr>
<td style="text-align:left">edge_list (list): list of edges as node tuples</td>
</tr>
<tr>
<td style="text-align:left">edge_labels (list): list of edge labels as strings</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns network with added edges</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="evolution">evolution<a class="anchor-link" href="#evolution"> </a></h1><blockquote><p>constructing evolutionary networks based on glycans</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="distance_from_embeddings" class="doc_header"><code>distance_from_embeddings</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L7" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>distance_from_embeddings</code>(<strong><code>df</code></strong>, <strong><code>embeddings</code></strong>, <strong><code>cut_off</code></strong>=<em><code>10</code></em>, <strong><code>rank</code></strong>=<em><code>'Species'</code></em>)</p>
</blockquote>
<p>calculates a cosine distance matrix from learned embeddings</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycans as rows and taxonomic information as columns</td>
</tr>
<tr>
<td style="text-align:left">embeddings (dataframe): dataframe with glycans as rows and learned embeddings as columns (e.g., from glycans_to_emb)</td>
</tr>
<tr>
<td style="text-align:left">cut_off (int): how many glycans a rank (e.g., species) needs to have at least to be included; default:10</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which taxonomic rank to use for grouping organisms; default:'Species'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a rank x rank distance matrix</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="dendrogram_from_distance" class="doc_header"><code>dendrogram_from_distance</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L31" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>dendrogram_from_distance</code>(<strong><code>dm</code></strong>, <strong><code>ylabel</code></strong>=<em><code>'Mammalia'</code></em>)</p>
</blockquote>
<p>plots a dendrogram from distance matrix</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">dm (dataframe): a rank x rank distance matrix (e.g., from distance_from_embeddings)</td>
</tr>
<tr>
<td style="text-align:left">ylabel (string): how to label the y-axis of the dendrogram; default:'Mammalia'</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

