---

title: network


keywords: fastai
sidebar: home_sidebar

summary: "Analyzing groups of glycans as networks (e.g., biosynthesis networks)"
description: "Analyzing groups of glycans as networks (e.g., biosynthesis networks)"
nb_path: "06_network.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 06_network.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/glycowork/network.html"><code>network</code></a> contains functions to arrange and analyze glycans in the context of networks. In such a network, each node represents a glycan and edges represent, for instance, their connection via a biosynthetic step. It should be noted, since <code>glycowork</code> treats glycans as molecular graphs, that these networks represent hierarchical graphs, with the network being one graph and each node within the network also a graph. <a href="/glycowork/network.html"><code>network</code></a> contains the following modules:</p>
<ul>
<li><code>biosynthesis</code> contains functions to construct and analyze biosynthetic glycan networks</li>
<li><code>evolution</code> contains functions to compare (taxonomic) groups as to their glycan repertoires</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="biosynthesis">biosynthesis<a class="anchor-link" href="#biosynthesis"> </a></h2><blockquote><p>constructing and analyzing biosynthetic glycan networks</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="construct_network" class="doc_header"><code>construct_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L190" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>construct_network</code>(<strong><code>glycans</code></strong>, <strong><code>add_virtual_nodes</code></strong>=<em><code>'exhaustive'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>, <strong><code>ptm</code></strong>=<em><code>True</code></em>, <strong><code>allowed_ptms</code></strong>=<em><code>['OS', '3S', '6S', '1P', '6P', 'OAc', '4Ac']</code></em>, <strong><code>permitted_roots</code></strong>=<em><code>['Gal(b1-4)Glc-ol', 'Gal(b1-4)GlcNAc-ol']</code></em>, <strong><code>directed</code></strong>=<em><code>False</code></em>, <strong><code>edge_type</code></strong>=<em><code>'monolink'</code></em>)</p>
</blockquote>
<p>visualize biosynthetic network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">add_virtual_nodes (string): indicates whether no ('none'), proximal ('simple'), or all ('exhaustive') virtual nodes should be added; default:'exhaustive'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
<tr>
<td style="text-align:left">ptm (bool): whether to consider post-translational modifications in the network construction; default:True</td>
</tr>
<tr>
<td style="text-align:left">allowed_ptms (list): list of PTMs to consider</td>
</tr>
<tr>
<td style="text-align:left">permitted_roots (list): which nodes should be considered as roots; default:["Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc-ol"]</td>
</tr>
<tr>
<td style="text-align:left">directed (bool): whether to return a network with directed edges in the direction of biosynthesis; default:False</td>
</tr>
<tr>
<td style="text-align:left">edge_type (string): indicates whether edges represent monosaccharides ('monosaccharide'), monosaccharide(linkage) ('monolink'), or enzyme catalyzing the reaction ('enzyme'); default:'monolink'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a networkx object of the network</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_network" class="doc_header"><code>plot_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L302" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_network</code>(<strong><code>network</code></strong>, <strong><code>plot_format</code></strong>=<em><code>'pydot2'</code></em>, <strong><code>edge_label_draw</code></strong>=<em><code>True</code></em>, <strong><code>node_size</code></strong>=<em><code>False</code></em>, <strong><code>lfc_dict</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>visualizes biosynthetic network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network, returned from construct_network</td>
</tr>
<tr>
<td style="text-align:left">plot_format (string): how to layout network, either 'pydot2', 'kamada_kawai', or 'spring'; default:'pydot2'</td>
</tr>
<tr>
<td style="text-align:left">edge_label_draw (bool): draws edge labels if True; default:True</td>
</tr>
<tr>
<td style="text-align:left">node_size (bool): whether nodes should be sized by an "abundance" attribute in network; default:False</td>
</tr>
<tr>
<td style="text-align:left">lfc_dict (dict): dictionary of enzyme:log2-fold-change to scale edge width; default:None</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Gal(b1-4)Glc-ol&quot;</span><span class="p">,</span> <span class="s2">&quot;GlcNAc(b1-3)Gal(b1-4)Glc-ol&quot;</span><span class="p">,</span>
           <span class="s2">&quot;GlcNAc6S(b1-3)Gal(b1-4)Glc-ol&quot;</span><span class="p">,</span>
           <span class="s2">&quot;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&quot;</span><span class="p">,</span> <span class="s2">&quot;Fuc(a1-2)Gal(b1-4)Glc-ol&quot;</span><span class="p">,</span>
          <span class="s2">&quot;Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol&quot;</span><span class="p">]</span>
<span class="n">network</span> <span class="o">=</span> <span class="n">construct_network</span><span class="p">(</span><span class="n">glycans</span><span class="p">)</span>
<span class="n">plot_network</span><span class="p">(</span><span class="n">network</span><span class="p">,</span> <span class="n">plot_format</span> <span class="o">=</span> <span class="s1">&#39;kamada_kawai&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea ">


<style>

</style>

<div id="fig_el692818441821527204632460426"></div>
<script>
function mpld3_load_lib(url, callback){
  var s = document.createElement('script');
  s.src = url;
  s.async = true;
  s.onreadystatechange = s.onload = callback;
  s.onerror = function(){console.warn("failed to load library " + url);};
  document.getElementsByTagName("head")[0].appendChild(s);
}

if(typeof(mpld3) !== "undefined" && mpld3._mpld3IsLoaded){
   // already loaded: just create the figure
   !function(mpld3){
       
       mpld3.draw_figure("fig_el692818441821527204632460426", {"width": 1152.0, "height": 1152.0, "axes": [{"bbox": [0.125, 0.125, 0.775, 0.755], "xlim": [-0.5437595571613685, 0.7410549713470251], "ylim": [-0.6423211038686082, 1.1560576614535782], "xdomain": [-0.5437595571613685, 0.7410549713470251], "ydomain": [-0.6423211038686082, 1.1560576614535782], "xscale": "linear", "yscale": "linear", "axes": [{"position": "top", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}, {"position": "right", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}], "axesbg": "#FFFFFF", "axesbgalpha": null, "zoomable": true, "id": "el69281844181876256", "lines": [], "paths": [], "markers": [], "texts": [{"text": "GlcNAc(b1-3)", "position": [0.2328715107102744, 0.3482450999532405], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -59.850033396209426, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181658784"}, {"text": "GlcNAc6S(b1-3)", "position": [0.3510179627407464, 0.39375435765240996], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -276.3151421794561, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181690976"}, {"text": "Fuc(a1-2)", "position": [0.4823085895266378, 0.8005038972483789], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -43.31698278380035, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181629392"}, {"text": "Gal(b1-4)", "position": [-0.006394680564630648, 0.08573144439004701], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.834232746500115, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181627904"}, {"text": "6S", "position": [0.2488350959155296, 0.14099166310889277], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -15.007625149328206, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181593440"}, {"text": "GlcNAc(b1-6)", "position": [0.12738593650565183, -0.043200266000297034], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -88.04026837107536, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181591232"}, {"text": "GlcNAc(b1-6)", "position": [-0.14898468077551705, -0.06403019188993336], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -86.76566635823725, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181564816"}, {"text": "Neu5Ac(a2-3)", "position": [-0.27964068684216287, 0.06865051841717573], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.1456943753537625, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181563472"}, {"text": "Neu5Ac(a2-3)", "position": [-0.28354598296493305, -0.484247886366637], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -0.5873641290989782, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181561984"}, {"text": "Gal(b1-3)", "position": [-0.4263245322357605, -0.3495212041817968], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -273.5731154691977, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181535712"}, {"text": "Gal(b1-3)", "position": [-0.15060082266359398, -0.3431365987342407], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -272.3013351534296, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181533504"}, {"text": "Gal(b1-4)", "position": [-0.009974402659009382, -0.46766297231237314], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -4.241220129144065, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181502992"}, {"text": "GlcNAc(b1-6)", "position": [-0.4240353780010673, -0.07572920624229135], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -85.06745648096113, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181501648"}, {"text": "Neu5Ac(a2-3)", "position": [-0.29337937193442143, -0.20840991654940044], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -1.2542380120031849, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181500112"}, {"text": "Gal(b1-4)", "position": [-0.015204063705234555, -0.1929619022802774], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -3.1686317765884975, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181473888"}, {"text": "Gal(b1-3)", "position": [0.12542235629935006, -0.31748827585840983], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -270.8128941631433, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181471680"}], "collections": [{"offsets": "data01", "xindex": 0, "yindex": 1, "paths": [[[[0.0, -0.5], [0.13260155, -0.5], [0.25978993539242673, -0.44731684579412084], [0.3535533905932738, -0.3535533905932738], [0.44731684579412084, -0.25978993539242673], [0.5, -0.13260155], [0.5, 0.0], [0.5, 0.13260155], [0.44731684579412084, 0.25978993539242673], [0.3535533905932738, 0.3535533905932738], [0.25978993539242673, 0.44731684579412084], [0.13260155, 0.5], [0.0, 0.5], [-0.13260155, 0.5], [-0.25978993539242673, 0.44731684579412084], [-0.3535533905932738, 0.3535533905932738], [-0.44731684579412084, 0.25978993539242673], [-0.5, 0.13260155], [-0.5, 0.0], [-0.5, -0.13260155], [-0.44731684579412084, -0.25978993539242673], [-0.3535533905932738, -0.3535533905932738], [-0.25978993539242673, -0.44731684579412084], [-0.13260155, -0.5], [0.0, -0.5]], ["M", "C", "C", "C", "C", "C", "C", "C", "C", "Z"]]], "pathtransforms": [[7.0710678118654755, 0.0, 0.0, 7.0710678118654755, 0.0, 0.0]], "alphas": [0.7], "edgecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "facecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "edgewidths": [1.0], "offsetcoordinates": "data", "pathcoordinates": "display", "zorder": 2, "id": "el69281844181690160"}, {"offsets": "data02", "xindex": 0, "yindex": 1, "paths": [[[[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.6295628015177843, 1.0]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [-0.14347800501431884, 0.07598048337037076]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.43226738733212766, -0.2127789659485634]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [0.12676148347245397, -0.45309361430650236]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.15449135653671525, -0.20404086715023748], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[0.12676148347245397, -0.45309361430650236], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]]], "pathtransforms": [], "alphas": [null], "edgecolors": ["#6495ED"], "facecolors": [], "edgewidths": [1.0], "offsetcoordinates": "display", "pathcoordinates": "data", "zorder": 1, "id": "el69281844181977600"}], "images": [], "sharex": [], "sharey": []}], "data": {"data01": [[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227], [-0.14347800501431884, 0.07598048337037076], [0.6295628015177843, 1.0], [-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439], [-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748], [0.12676148347245397, -0.45309361430650236], [-0.4158033686700069, 0.0613205534639807], [0.12408322912624614, -0.18188293741031733]], "data02": [[0.0, 0.0]]}, "id": "el69281844182152720", "plugins": [{"type": "reset"}, {"type": "zoom", "button": true, "enabled": false}, {"type": "boxzoom", "button": true, "enabled": false}, {"type": "tooltip", "id": "el69281844181690160", "labels": ["Gal(b1-4)Glc-ol", "GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc6S(b1-3)Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "Fuc(a1-2)Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol"], "hoffset": 0, "voffset": 10, "location": "mouse"}]});
   }(mpld3);
}else if(typeof define === "function" && define.amd){
   // require.js is available: use it to load d3/mpld3
   require.config({paths: {d3: "https://d3js.org/d3.v5"}});
   require(["d3"], function(d3){
      window.d3 = d3;
      mpld3_load_lib("https://mpld3.github.io/js/mpld3.v0.5.2.js", function(){
         
         mpld3.draw_figure("fig_el692818441821527204632460426", {"width": 1152.0, "height": 1152.0, "axes": [{"bbox": [0.125, 0.125, 0.775, 0.755], "xlim": [-0.5437595571613685, 0.7410549713470251], "ylim": [-0.6423211038686082, 1.1560576614535782], "xdomain": [-0.5437595571613685, 0.7410549713470251], "ydomain": [-0.6423211038686082, 1.1560576614535782], "xscale": "linear", "yscale": "linear", "axes": [{"position": "top", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}, {"position": "right", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}], "axesbg": "#FFFFFF", "axesbgalpha": null, "zoomable": true, "id": "el69281844181876256", "lines": [], "paths": [], "markers": [], "texts": [{"text": "GlcNAc(b1-3)", "position": [0.2328715107102744, 0.3482450999532405], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -59.850033396209426, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181658784"}, {"text": "GlcNAc6S(b1-3)", "position": [0.3510179627407464, 0.39375435765240996], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -276.3151421794561, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181690976"}, {"text": "Fuc(a1-2)", "position": [0.4823085895266378, 0.8005038972483789], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -43.31698278380035, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181629392"}, {"text": "Gal(b1-4)", "position": [-0.006394680564630648, 0.08573144439004701], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.834232746500115, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181627904"}, {"text": "6S", "position": [0.2488350959155296, 0.14099166310889277], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -15.007625149328206, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181593440"}, {"text": "GlcNAc(b1-6)", "position": [0.12738593650565183, -0.043200266000297034], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -88.04026837107536, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181591232"}, {"text": "GlcNAc(b1-6)", "position": [-0.14898468077551705, -0.06403019188993336], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -86.76566635823725, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181564816"}, {"text": "Neu5Ac(a2-3)", "position": [-0.27964068684216287, 0.06865051841717573], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.1456943753537625, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181563472"}, {"text": "Neu5Ac(a2-3)", "position": [-0.28354598296493305, -0.484247886366637], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -0.5873641290989782, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181561984"}, {"text": "Gal(b1-3)", "position": [-0.4263245322357605, -0.3495212041817968], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -273.5731154691977, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181535712"}, {"text": "Gal(b1-3)", "position": [-0.15060082266359398, -0.3431365987342407], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -272.3013351534296, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181533504"}, {"text": "Gal(b1-4)", "position": [-0.009974402659009382, -0.46766297231237314], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -4.241220129144065, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181502992"}, {"text": "GlcNAc(b1-6)", "position": [-0.4240353780010673, -0.07572920624229135], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -85.06745648096113, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181501648"}, {"text": "Neu5Ac(a2-3)", "position": [-0.29337937193442143, -0.20840991654940044], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -1.2542380120031849, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181500112"}, {"text": "Gal(b1-4)", "position": [-0.015204063705234555, -0.1929619022802774], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -3.1686317765884975, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181473888"}, {"text": "Gal(b1-3)", "position": [0.12542235629935006, -0.31748827585840983], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -270.8128941631433, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181471680"}], "collections": [{"offsets": "data01", "xindex": 0, "yindex": 1, "paths": [[[[0.0, -0.5], [0.13260155, -0.5], [0.25978993539242673, -0.44731684579412084], [0.3535533905932738, -0.3535533905932738], [0.44731684579412084, -0.25978993539242673], [0.5, -0.13260155], [0.5, 0.0], [0.5, 0.13260155], [0.44731684579412084, 0.25978993539242673], [0.3535533905932738, 0.3535533905932738], [0.25978993539242673, 0.44731684579412084], [0.13260155, 0.5], [0.0, 0.5], [-0.13260155, 0.5], [-0.25978993539242673, 0.44731684579412084], [-0.3535533905932738, 0.3535533905932738], [-0.44731684579412084, 0.25978993539242673], [-0.5, 0.13260155], [-0.5, 0.0], [-0.5, -0.13260155], [-0.44731684579412084, -0.25978993539242673], [-0.3535533905932738, -0.3535533905932738], [-0.25978993539242673, -0.44731684579412084], [-0.13260155, -0.5], [0.0, -0.5]], ["M", "C", "C", "C", "C", "C", "C", "C", "C", "Z"]]], "pathtransforms": [[7.0710678118654755, 0.0, 0.0, 7.0710678118654755, 0.0, 0.0]], "alphas": [0.7], "edgecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "facecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "edgewidths": [1.0], "offsetcoordinates": "data", "pathcoordinates": "display", "zorder": 2, "id": "el69281844181690160"}, {"offsets": "data02", "xindex": 0, "yindex": 1, "paths": [[[[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.6295628015177843, 1.0]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [-0.14347800501431884, 0.07598048337037076]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.43226738733212766, -0.2127789659485634]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [0.12676148347245397, -0.45309361430650236]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.15449135653671525, -0.20404086715023748], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[0.12676148347245397, -0.45309361430650236], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]]], "pathtransforms": [], "alphas": [null], "edgecolors": ["#6495ED"], "facecolors": [], "edgewidths": [1.0], "offsetcoordinates": "display", "pathcoordinates": "data", "zorder": 1, "id": "el69281844181977600"}], "images": [], "sharex": [], "sharey": []}], "data": {"data01": [[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227], [-0.14347800501431884, 0.07598048337037076], [0.6295628015177843, 1.0], [-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439], [-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748], [0.12676148347245397, -0.45309361430650236], [-0.4158033686700069, 0.0613205534639807], [0.12408322912624614, -0.18188293741031733]], "data02": [[0.0, 0.0]]}, "id": "el69281844182152720", "plugins": [{"type": "reset"}, {"type": "zoom", "button": true, "enabled": false}, {"type": "boxzoom", "button": true, "enabled": false}, {"type": "tooltip", "id": "el69281844181690160", "labels": ["Gal(b1-4)Glc-ol", "GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc6S(b1-3)Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "Fuc(a1-2)Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol"], "hoffset": 0, "voffset": 10, "location": "mouse"}]});
      });
    });
}else{
    // require.js not available: dynamically load d3 & mpld3
    mpld3_load_lib("https://d3js.org/d3.v5.js", function(){
         mpld3_load_lib("https://mpld3.github.io/js/mpld3.v0.5.2.js", function(){
                 
                 mpld3.draw_figure("fig_el692818441821527204632460426", {"width": 1152.0, "height": 1152.0, "axes": [{"bbox": [0.125, 0.125, 0.775, 0.755], "xlim": [-0.5437595571613685, 0.7410549713470251], "ylim": [-0.6423211038686082, 1.1560576614535782], "xdomain": [-0.5437595571613685, 0.7410549713470251], "ydomain": [-0.6423211038686082, 1.1560576614535782], "xscale": "linear", "yscale": "linear", "axes": [{"position": "top", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}, {"position": "right", "nticks": 0, "tickvalues": [], "tickformat_formatter": "", "tickformat": "", "scale": "linear", "fontsize": null, "grid": {"gridOn": false}, "visible": true}], "axesbg": "#FFFFFF", "axesbgalpha": null, "zoomable": true, "id": "el69281844181876256", "lines": [], "paths": [], "markers": [], "texts": [{"text": "GlcNAc(b1-3)", "position": [0.2328715107102744, 0.3482450999532405], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -59.850033396209426, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181658784"}, {"text": "GlcNAc6S(b1-3)", "position": [0.3510179627407464, 0.39375435765240996], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -276.3151421794561, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181690976"}, {"text": "Fuc(a1-2)", "position": [0.4823085895266378, 0.8005038972483789], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -43.31698278380035, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181629392"}, {"text": "Gal(b1-4)", "position": [-0.006394680564630648, 0.08573144439004701], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.834232746500115, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181627904"}, {"text": "6S", "position": [0.2488350959155296, 0.14099166310889277], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -15.007625149328206, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181593440"}, {"text": "GlcNAc(b1-6)", "position": [0.12738593650565183, -0.043200266000297034], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -88.04026837107536, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181591232"}, {"text": "GlcNAc(b1-6)", "position": [-0.14898468077551705, -0.06403019188993336], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -86.76566635823725, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181564816"}, {"text": "Neu5Ac(a2-3)", "position": [-0.27964068684216287, 0.06865051841717573], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -2.1456943753537625, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181563472"}, {"text": "Neu5Ac(a2-3)", "position": [-0.28354598296493305, -0.484247886366637], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -0.5873641290989782, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181561984"}, {"text": "Gal(b1-3)", "position": [-0.4263245322357605, -0.3495212041817968], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -273.5731154691977, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181535712"}, {"text": "Gal(b1-3)", "position": [-0.15060082266359398, -0.3431365987342407], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -272.3013351534296, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181533504"}, {"text": "Gal(b1-4)", "position": [-0.009974402659009382, -0.46766297231237314], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -4.241220129144065, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181502992"}, {"text": "GlcNAc(b1-6)", "position": [-0.4240353780010673, -0.07572920624229135], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -85.06745648096113, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181501648"}, {"text": "Neu5Ac(a2-3)", "position": [-0.29337937193442143, -0.20840991654940044], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -1.2542380120031849, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181500112"}, {"text": "Gal(b1-4)", "position": [-0.015204063705234555, -0.1929619022802774], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -3.1686317765884975, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181473888"}, {"text": "Gal(b1-3)", "position": [0.12542235629935006, -0.31748827585840983], "coordinates": "data", "h_anchor": "middle", "v_baseline": "auto", "rotation": -270.8128941631433, "fontsize": 10.0, "color": "#000000", "alpha": 1, "zorder": 1, "id": "el69281844181471680"}], "collections": [{"offsets": "data01", "xindex": 0, "yindex": 1, "paths": [[[[0.0, -0.5], [0.13260155, -0.5], [0.25978993539242673, -0.44731684579412084], [0.3535533905932738, -0.3535533905932738], [0.44731684579412084, -0.25978993539242673], [0.5, -0.13260155], [0.5, 0.0], [0.5, 0.13260155], [0.44731684579412084, 0.25978993539242673], [0.3535533905932738, 0.3535533905932738], [0.25978993539242673, 0.44731684579412084], [0.13260155, 0.5], [0.0, 0.5], [-0.13260155, 0.5], [-0.25978993539242673, 0.44731684579412084], [-0.3535533905932738, 0.3535533905932738], [-0.44731684579412084, 0.25978993539242673], [-0.5, 0.13260155], [-0.5, 0.0], [-0.5, -0.13260155], [-0.44731684579412084, -0.25978993539242673], [-0.3535533905932738, -0.3535533905932738], [-0.25978993539242673, -0.44731684579412084], [-0.13260155, -0.5], [0.0, -0.5]], ["M", "C", "C", "C", "C", "C", "C", "C", "C", "Z"]]], "pathtransforms": [[7.0710678118654755, 0.0, 0.0, 7.0710678118654755, 0.0, 0.0]], "alphas": [0.7], "edgecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "facecolors": ["rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(100, 149, 237, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)", "rgba(255, 140, 0, 0.7)"], "edgewidths": [1.0], "offsetcoordinates": "data", "pathcoordinates": "display", "zorder": 2, "id": "el69281844181690160"}, {"offsets": "data02", "xindex": 0, "yindex": 1, "paths": [[[[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.3350543775354912, 0.6010077944967577], [0.6295628015177843, 1.0]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [-0.14347800501431884, 0.07598048337037076]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227]], ["M", "L"]], [[[0.13068864388505755, 0.09548240540972326], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14347800501431884, 0.07598048337037076], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439]], ["M", "L"]], [[[-0.4203816771393934, -0.4862634424150301], [-0.43226738733212766, -0.2127789659485634]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.14671028879047274, -0.4822323303182439], [0.12676148347245397, -0.45309361430650236]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.4158033686700069, 0.0613205534639807]], ["M", "L"]], [[[-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748]], ["M", "L"]], [[[-0.15449135653671525, -0.20404086715023748], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]], [[[0.12676148347245397, -0.45309361430650236], [0.12408322912624614, -0.18188293741031733]], ["M", "L"]]], "pathtransforms": [], "alphas": [null], "edgecolors": ["#6495ED"], "facecolors": [], "edgewidths": [1.0], "offsetcoordinates": "display", "pathcoordinates": "data", "zorder": 1, "id": "el69281844181977600"}], "images": [], "sharex": [], "sharey": []}], "data": {"data01": [[0.3350543775354912, 0.6010077944967577], [0.13068864388505755, 0.09548240540972326], [0.36698154794600163, 0.18650092080806227], [-0.14347800501431884, 0.07598048337037076], [0.6295628015177843, 1.0], [-0.4203816771393934, -0.4862634424150301], [-0.14671028879047274, -0.4822323303182439], [-0.43226738733212766, -0.2127789659485634], [-0.15449135653671525, -0.20404086715023748], [0.12676148347245397, -0.45309361430650236], [-0.4158033686700069, 0.0613205534639807], [0.12408322912624614, -0.18188293741031733]], "data02": [[0.0, 0.0]]}, "id": "el69281844182152720", "plugins": [{"type": "reset"}, {"type": "zoom", "button": true, "enabled": false}, {"type": "boxzoom", "button": true, "enabled": false}, {"type": "tooltip", "id": "el69281844181690160", "labels": ["Gal(b1-4)Glc-ol", "GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc6S(b1-3)Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "Fuc(a1-2)Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[Gal(b1-3)GlcNAc(b1-6)]Gal(b1-4)Glc-ol", "Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol", "GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)Glc-ol"], "hoffset": 0, "voffset": 10, "location": "mouse"}]});
            })
         });
}
</script>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_diff" class="doc_header"><code>find_diff</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L157" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_diff</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>finds the subgraph that differs between glycans and returns it, will only work if the differing subgraph is connected</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns difference between glycan_a and glycan_b in IUPAC-condensed</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_diff</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;Fuc(a1-6)&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_virtual_nodes" class="doc_header"><code>get_virtual_nodes</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L459" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_virtual_nodes</code>(<strong><code>glycan</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>)</p>
</blockquote>
<p>find unobserved biosynthetic precursors of a glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of virtual node graphs</td>
</tr>
<tr>
<td style="text-align:left">(2) list of virtual nodes in IUPAC-condensed format</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_virtual_nodes</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([&lt;networkx.classes.graph.Graph at 0x1ad61bf4430&gt;,
  &lt;networkx.classes.graph.Graph at 0x1ad61bf4a60&gt;,
  &lt;networkx.classes.graph.Graph at 0x1ad61bf4760&gt;],
 [&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-3)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;,
  &#39;Man(a1-6)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc-ol&#39;])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_shared_virtuals" class="doc_header"><code>find_shared_virtuals</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L368" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_shared_virtuals</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>min_size</code></strong>=<em><code>1</code></em>)</p>
</blockquote>
<p>finds virtual nodes that are shared between two glycans (i.e., that connect these two glycans)</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">min_size (int): length of smallest root in biosynthetic network; default:1</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns list of edges between glycan and virtual node (if virtual node connects the two glycans)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_shared_virtuals</span><span class="p">(</span><span class="s1">&#39;GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;),
 (&#39;GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;,
  &#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol&#39;)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_path" class="doc_header"><code>find_path</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L540" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_path</code>(<strong><code>glycan_a</code></strong>, <strong><code>glycan_b</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>find virtual node path between two glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan_a (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_b (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of edges to connect glycan_a and glycan_b via virtual nodes</td>
</tr>
<tr>
<td style="text-align:left">(2) dictionary of edge labels detailing difference between two connected nodes</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span><span class="s1">&#39;Gal(b1-4)Glc-ol&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;),
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;)],
 {(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;,
   &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;): &#39;Gal(b1-4)&#39;,
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;): &#39;GlcNAc(b1-3)&#39;})</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_shortest_path" class="doc_header"><code>find_shortest_path</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L598" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_shortest_path</code>(<strong><code>goal_glycan</code></strong>, <strong><code>glycan_list</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>finds the glycan with the shortest path via virtual nodes to the goal glycan</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">goal_glycan (string): glycan in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">glycan_list (list): list of glycans in IUPAC-condensed format</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed; default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes; default:5</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of edges of shortest path to connect goal_glycan and glycan via virtual nodes</td>
</tr>
<tr>
<td style="text-align:left">(2) dictionary of edge labels detailing difference between two connected nodes in shortest path</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">find_shortest_path</span><span class="p">(</span><span class="s1">&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span>
                  <span class="p">[</span><span class="s1">&#39;Gal(b1-4)Glc-ol&#39;</span><span class="p">,</span> <span class="s1">&#39;Gal(b1-4)GlcNAc-ol&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;),
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;)],
 {(&#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;,
   &#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;): &#39;Gal(b1-4)&#39;,
  (&#39;GlcNAc(b1-3)Gal(b1-4)Glc-ol&#39;, &#39;Gal(b1-4)Glc-ol&#39;): &#39;GlcNAc(b1-3)&#39;})</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="network_alignment" class="doc_header"><code>network_alignment</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L654" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>network_alignment</code>(<strong><code>network_a</code></strong>, <strong><code>network_b</code></strong>)</p>
</blockquote>
<p>combines two networks into a new network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network_a (networkx object): biosynthetic network from construct_network</td>
</tr>
<tr>
<td style="text-align:left">network_b (networkx object): biosynthetic network from construct_network</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns combined network as a networkx object</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="infer_network" class="doc_header"><code>infer_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L715" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>infer_network</code>(<strong><code>network</code></strong>, <strong><code>network_species</code></strong>, <strong><code>species_list</code></strong>, <strong><code>filepath</code></strong>=<em><code>None</code></em>, <strong><code>df</code></strong>=<em><code>None</code></em>, <strong><code>add_virtual_nodes</code></strong>=<em><code>'exhaustive'</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>reducing_end</code></strong>=<em><code>['Glc-ol', 'GlcNAc-ol', 'Glc3S-ol', 'GlcNAc6S-ol', 'GlcNAc6P-ol', 'GlcNAc1P-ol', 'Glc3P-ol', 'Glc6S-ol', 'GlcOS-ol']</code></em>, <strong><code>limit</code></strong>=<em><code>5</code></em>, <strong><code>ptm</code></strong>=<em><code>False</code></em>, <strong><code>allowed_ptms</code></strong>=<em><code>['OS', '3S', '6S', '1P', '6P', 'OAc', '4Ac']</code></em>, <strong><code>permitted_roots</code></strong>=<em><code>['Gal(b1-4)Glc-ol', 'Gal(b1-4)GlcNAc-ol']</code></em>, <strong><code>directed</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>replaces virtual nodes if they are observed in other species</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network that should be inferred</td>
</tr>
<tr>
<td style="text-align:left">network_species (string): species from which the network stems</td>
</tr>
<tr>
<td style="text-align:left">species_list (list): list of species to compare network to</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): filepath to load biosynthetic networks from other species, if precalculated (def. recommended, as calculation will take ~1.5 hours); default:None</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): dataframe containing species-specific glycans, only needed if filepath=None;default:None</td>
</tr>
<tr>
<td style="text-align:left">add_virtual_nodes (string): indicates whether no ('None'), proximal ('simple'), or all ('exhaustive') virtual nodes should be added;only needed if filepath=None;default:'exhaustive'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used;only needed if filepath=None</td>
</tr>
<tr>
<td style="text-align:left">reducing_end (list): monosaccharides at the reducing end that are allowed;only needed if filepath=None;default:milk glycan reducing ends</td>
</tr>
<tr>
<td style="text-align:left">limit (int): maximum number of virtual nodes between observed nodes;only needed if filepath=None;default:5</td>
</tr>
<tr>
<td style="text-align:left">ptm (bool): whether to consider post-translational modifications in the network construction; default:False</td>
</tr>
<tr>
<td style="text-align:left">allowed_ptms (list): list of PTMs to consider</td>
</tr>
<tr>
<td style="text-align:left">permitted_roots (list): which nodes should be considered as roots; default:["Gal(b1-4)Glc-ol", "Gal(b1-4)GlcNAc-ol"]</td>
</tr>
<tr>
<td style="text-align:left">directed (bool): whether to return a network with directed edges in the direction of biosynthesis; default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns network with filled in virtual nodes</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="retrieve_inferred_nodes" class="doc_header"><code>retrieve_inferred_nodes</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L761" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>retrieve_inferred_nodes</code>(<strong><code>network</code></strong>, <strong><code>species</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>returns the inferred virtual nodes of a network that has been used with infer_network</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network with inferred virtual nodes</td>
</tr>
<tr>
<td style="text-align:left">species (string): species from which the network stems (only relevant if multiple species in network); default:None</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns inferred nodes as list or dictionary (if species argument is used)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="update_network" class="doc_header"><code>update_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L859" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>update_network</code>(<strong><code>network_in</code></strong>, <strong><code>edge_list</code></strong>, <strong><code>edge_labels</code></strong>=<em><code>None</code></em>, <strong><code>node_labels</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>updates a network with new edges and their labels</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): network that should be modified</td>
</tr>
<tr>
<td style="text-align:left">edge_list (list): list of edges as node tuples</td>
</tr>
<tr>
<td style="text-align:left">edge_labels (list): list of edge labels as strings</td>
</tr>
<tr>
<td style="text-align:left">node_labels (dict): dictionary of form node:0 or 1 depending on whether the node is observed or virtual</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns network with added edges</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="trace_diamonds" class="doc_header"><code>trace_diamonds</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L1124" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>trace_diamonds</code>(<strong><code>network</code></strong>, <strong><code>species_list</code></strong>, <strong><code>filepath</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>extracts diamond-shape motifs from biosynthetic networks (A-&gt;B,A-&gt;C,B-&gt;D,C-&gt;D) and uses evolutionary information to determine which path is taken from A to D</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network, returned from construct_network</td>
</tr>
<tr>
<td style="text-align:left">species_list (list): list of species to compare network to</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): filepath to load biosynthetic networks from other species; files need to be species name + '_graph_exhaustive.pkl'</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns dataframe of each intermediary glycan and its proportion (0-1) of how often it has been experimentally observed in this path</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="infuse_network" class="doc_header"><code>infuse_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L1042" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>infuse_network</code>(<strong><code>network</code></strong>, <strong><code>node_df</code></strong>, <strong><code>node_abundance</code></strong>=<em><code>True</code></em>, <strong><code>glycan_col</code></strong>=<em><code>'target'</code></em>, <strong><code>intensity_col</code></strong>=<em><code>'rel_intensity'</code></em>)</p>
</blockquote>
<p>add new data to an existing network, such as glycomics abundance data</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network, returned from construct_network</td>
</tr>
<tr>
<td style="text-align:left">node_df (dataframe): dataframe containing glycans and their relative intensity</td>
</tr>
<tr>
<td style="text-align:left">node_abundance (bool): whether to add node abundance to the network as the node attribute "abundance"; default:True</td>
</tr>
<tr>
<td style="text-align:left">glycan_col (string): column name of the glycans in node_df</td>
</tr>
<tr>
<td style="text-align:left">intensity_col (string): column name of the relative intensities in node_df</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a network with added information</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="export_network" class="doc_header"><code>export_network</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/biosynthesis.py#L995" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>export_network</code>(<strong><code>network</code></strong>, <strong><code>filepath</code></strong>)</p>
</blockquote>
<p>converts NetworkX network into files usable by Gephi</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">network (networkx object): biosynthetic network, returned from construct_network</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): should describe a valid path + file name prefix, will be appended by file description and type</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) saves a .csv dataframe containing the edge list and edge labels</td>
</tr>
<tr>
<td style="text-align:left">(2) saves a .csv dataframe containing node IDs and labels</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="evolution">evolution<a class="anchor-link" href="#evolution"> </a></h2><blockquote><p>investigating evolutionary relationships of glycans</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="distance_from_embeddings" class="doc_header"><code>distance_from_embeddings</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L9" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>distance_from_embeddings</code>(<strong><code>df</code></strong>, <strong><code>embeddings</code></strong>, <strong><code>cut_off</code></strong>=<em><code>10</code></em>, <strong><code>rank</code></strong>=<em><code>'Species'</code></em>, <strong><code>averaging</code></strong>=<em><code>'median'</code></em>)</p>
</blockquote>
<p>calculates a cosine distance matrix from learned embeddings</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycans as rows and taxonomic information as columns</td>
</tr>
<tr>
<td style="text-align:left">embeddings (dataframe): dataframe with glycans as rows and learned embeddings as columns (e.g., from glycans_to_emb)</td>
</tr>
<tr>
<td style="text-align:left">cut_off (int): how many glycans a rank (e.g., species) needs to have at least to be included; default:10</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which taxonomic rank to use for grouping organisms; default:'Species'</td>
</tr>
<tr>
<td style="text-align:left">averaging (string): how to average embeddings, by 'median' or 'mean'; default:'median'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a rank x rank distance matrix</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="distance_from_metric" class="doc_header"><code>distance_from_metric</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L54" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>distance_from_metric</code>(<strong><code>df</code></strong>, <strong><code>networks</code></strong>, <strong><code>metric</code></strong>=<em><code>'Jaccard'</code></em>, <strong><code>cut_off</code></strong>=<em><code>10</code></em>, <strong><code>rank</code></strong>=<em><code>'Species'</code></em>)</p>
</blockquote>
<p>calculates a distance matrix of generated networks based on provided metric</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe with glycans as rows and taxonomic information as columns</td>
</tr>
<tr>
<td style="text-align:left">networks (list): list of networks in networkx format</td>
</tr>
<tr>
<td style="text-align:left">metric (string): which metric to use, available: 'Jaccard'; default:'Jaccard'</td>
</tr>
<tr>
<td style="text-align:left">cut_off (int): how many glycans a rank (e.g., species) needs to have at least to be included; default:10</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which taxonomic rank to use for grouping organisms; default:'Species'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a rank x rank distance matrix</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="dendrogram_from_distance" class="doc_header"><code>dendrogram_from_distance</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L82" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>dendrogram_from_distance</code>(<strong><code>dm</code></strong>, <strong><code>ylabel</code></strong>=<em><code>'Mammalia'</code></em>, <strong><code>filepath</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>plots a dendrogram from distance matrix</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">dm (dataframe): a rank x rank distance matrix (e.g., from distance_from_embeddings)</td>
</tr>
<tr>
<td style="text-align:left">ylabel (string): how to label the y-axis of the dendrogram; default:'Mammalia'</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): absolute path including full filename allows for saving the plot</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_conservation" class="doc_header"><code>check_conservation</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/network/evolution.py#L110" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_conservation</code>(<strong><code>glycan</code></strong>, <strong><code>df</code></strong>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>rank</code></strong>=<em><code>'Order'</code></em>, <strong><code>filepath</code></strong>=<em><code>None</code></em>, <strong><code>threshold</code></strong>=<em><code>5</code></em>, <strong><code>motif</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>estimates evolutionary conservation of glycans and glycan motifs via biosynthetic networks</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycan (string): full glycan or glycan motif in IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">df (dataframe): dataframe in the style of df_species, each row one glycan and columns are the taxonomic levels</td>
</tr>
<tr>
<td style="text-align:left">libr (list): library of monosaccharides; if you have one use it, otherwise a comprehensive lib will be used</td>
</tr>
<tr>
<td style="text-align:left">rank (string): at which taxonomic level to assess conservation; default:Order</td>
</tr>
<tr>
<td style="text-align:left">filepath (string): filepath to load biosynthetic networks from other species; files need to be species name + '_graph_exhaustive.pkl'</td>
</tr>
<tr>
<td style="text-align:left">threshold (int): threshold of how many glycans a species needs to have to consider the species;default:5</td>
</tr>
<tr>
<td style="text-align:left">motif (bool): whether glycan is a motif (True) or a full sequence (False); default:False</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns a dictionary of taxonomic group : degree of conservation</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

